// rclnodejs/generated/rcl_interfaces/rcl_interfaces__srv__GetParameterTypes_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const GetParameterTypes_ResponseRefStruct = StructType({
  types: UInt8Wrapper.refObjectArrayType,
});
const GetParameterTypes_ResponseRefArray = ArrayType(GetParameterTypes_ResponseRefStruct);
const GetParameterTypes_ResponseRefStructArray = StructType({
  data: GetParameterTypes_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetParameterTypes_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetParameterTypes_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.types = UInt8Wrapper.createArray();
      this._wrapperFields.types.copy(other._wrapperFields.types);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetParameterTypes_ResponseRefStruct();
    this._wrapperFields.types = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetParameterTypes_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetParameterTypes_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetParameterTypes_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetParameterTypes_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetParameterTypes_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.types.freeze(own, checkConsistency);
    this._refObject.types = this._wrapperFields.types.refObject;
    if (own) {
      this._wrapperFields.types.fill(Uint8Array.from([]));
      this._wrapperFields.types.freeze(own, checkConsistency);
      this._refObject.types = this._wrapperFields.types.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.types.copyRefObject(refObject.types);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.types.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.types);
      if (UInt8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.types, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GetParameterTypes_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'srv', interfaceName: 'GetParameterTypes_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get types() {
    return this._wrapperFields['types'].data;
  }
  set types(value) {
    this._wrapperFields['types'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetParameterTypes_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.types.copyRefObject(this._refObject.types);
  }
  copy(other) {
    this._refObject = new GetParameterTypes_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.types.copy(other._wrapperFields.types);
  }
  static get classType() {
    return GetParameterTypes_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"types","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"GetParameterTypes_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetParameterTypes_Response"};
  }
  hasMember(name) {
    let memberNames = ["types"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetParameterTypes_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetParameterTypes_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetParameterTypes_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetParameterTypes_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetParameterTypes_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetParameterTypes_ResponseRefArray(size);
    this._refObject = new GetParameterTypes_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetParameterTypes_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetParameterTypes_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetParameterTypes_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetParameterTypes_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetParameterTypes_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetParameterTypes_ResponseArrayWrapper;
  }
}
module.exports = GetParameterTypes_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "types",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "GetParameterTypes_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetParameterTypes_Response"
}
*/
