// rclnodejs/generated/rcl_interfaces/rcl_interfaces__srv__GetParameterTypes_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetParameterTypes_RequestRefStruct = StructType({
  names: StringWrapper.refObjectArrayType,
});
const GetParameterTypes_RequestRefArray = ArrayType(GetParameterTypes_RequestRefStruct);
const GetParameterTypes_RequestRefStructArray = StructType({
  data: GetParameterTypes_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetParameterTypes_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetParameterTypes_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.names = StringWrapper.createArray();
      this._wrapperFields.names.copy(other._wrapperFields.names);
      this.names = other.names;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetParameterTypes_RequestRefStruct();
    this._wrapperFields.names = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetParameterTypes_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetParameterTypes_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetParameterTypes_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetParameterTypes_RequestRefStructArray
  }
  static get refObjectType() {
    return GetParameterTypes_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    if (!own) {
      this._wrapperFields.names.fill(this._namesArray);
      this._wrapperFields.names.freeze(own, checkConsistency);
      this._refObject.names = this._wrapperFields.names.refObject;
    } else {
      this._wrapperFields.names.fill([]);
      this._wrapperFields.names.freeze(own, checkConsistency);
      this._refObject.names = this._wrapperFields.names.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.names.data.length = refObject.names.size;
    for (let index = 0; index < refObject.names.size; index++) {
      this._namesArray[index] = refObject.names.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.names, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GetParameterTypes_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'srv', interfaceName: 'GetParameterTypes_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get names() {
    return this._namesArray;
  }
  set names(value) {
    this._namesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetParameterTypes_RequestRefStruct(refObject.toObject());
    refObject.names.data.length = refObject.names.size;
    for (let index = 0; index < refObject.names.size; index++) {
      this._namesArray[index] = refObject.names.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new GetParameterTypes_RequestRefStruct(other._refObject.toObject());
    this._namesArray = other._namesArray.slice();
  }
  static get classType() {
    return GetParameterTypes_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"GetParameterTypes_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetParameterTypes_Request"};
  }
  hasMember(name) {
    let memberNames = ["names"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetParameterTypes_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetParameterTypes_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetParameterTypes_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetParameterTypes_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetParameterTypes_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetParameterTypes_RequestRefArray(size);
    this._refObject = new GetParameterTypes_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetParameterTypes_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetParameterTypes_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetParameterTypes_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetParameterTypes_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetParameterTypes_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetParameterTypes_RequestArrayWrapper;
  }
}
module.exports = GetParameterTypes_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "GetParameterTypes_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetParameterTypes_Request"
}
*/
