// rclnodejs/generated/rcl_interfaces/rcl_interfaces__srv__DescribeParameters_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ParameterDescriptorWrapper = require('../../generated/rcl_interfaces/rcl_interfaces__msg__ParameterDescriptor.js');
const DescribeParameters_ResponseRefStruct = StructType({
  descriptors: ParameterDescriptorWrapper.refObjectArrayType,
});
const DescribeParameters_ResponseRefArray = ArrayType(DescribeParameters_ResponseRefStruct);
const DescribeParameters_ResponseRefStructArray = StructType({
  data: DescribeParameters_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DescribeParameters_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DescribeParameters_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.descriptors = ParameterDescriptorWrapper.createArray();
      this._wrapperFields.descriptors.copy(other._wrapperFields.descriptors);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DescribeParameters_ResponseRefStruct();
    this._wrapperFields.descriptors = ParameterDescriptorWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new DescribeParameters_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DescribeParameters_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return DescribeParameters_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return DescribeParameters_ResponseRefStructArray
  }
  static get refObjectType() {
    return DescribeParameters_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.descriptors.freeze(own, checkConsistency);
    this._refObject.descriptors = this._wrapperFields.descriptors.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.descriptors.copyRefObject(refObject.descriptors);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.descriptors.size != 0) {
      ParameterDescriptorWrapper.ArrayType.freeArray(refObject.descriptors);
      if (ParameterDescriptorWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.descriptors, ParameterDescriptorWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    DescribeParameters_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'srv', interfaceName: 'DescribeParameters_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get descriptors() {
    return this._wrapperFields.descriptors;
  }
  set descriptors(value) {
    this._wrapperFields.descriptors.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new DescribeParameters_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.descriptors.copyRefObject(this._refObject.descriptors);
  }
  copy(other) {
    this._refObject = new DescribeParameters_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.descriptors.copy(other._wrapperFields.descriptors);
  }
  static get classType() {
    return DescribeParameters_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"descriptors","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rcl_interfaces","type":"ParameterDescriptor","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"DescribeParameters_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DescribeParameters_Response"};
  }
  hasMember(name) {
    let memberNames = ["descriptors"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DescribeParameters_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DescribeParameters_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DescribeParameters_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DescribeParameters_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DescribeParameters_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DescribeParameters_ResponseRefArray(size);
    this._refObject = new DescribeParameters_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DescribeParameters_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DescribeParameters_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DescribeParameters_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DescribeParameters_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DescribeParameters_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DescribeParameters_ResponseArrayWrapper;
  }
}
module.exports = DescribeParameters_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "descriptors",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rcl_interfaces",
        "type": "ParameterDescriptor",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "DescribeParameters_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DescribeParameters_Response"
}
*/
