// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__IntegerRange.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const IntegerRangeRefStruct = StructType({
  from_value: primitiveTypes.int64,
  to_value: primitiveTypes.int64,
  step: primitiveTypes.uint64,
});
const IntegerRangeRefArray = ArrayType(IntegerRangeRefStruct);
const IntegerRangeRefStructArray = StructType({
  data: IntegerRangeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class IntegerRangeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new IntegerRangeRefStruct(other._refObject.toObject());
      this._from_valueIntialized = true;
      this._to_valueIntialized = true;
      this._stepIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new IntegerRangeRefStruct();
    this._from_valueIntialized = false;
    this._to_valueIntialized = false;
    this._stepIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new IntegerRangeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IntegerRangeArrayWrapper;
  }
  static get ArrayType() {
    return IntegerRangeArrayWrapper;
  }
  static get refObjectArrayType() {
    return IntegerRangeRefStructArray
  }
  static get refObjectType() {
    return IntegerRangeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._from_valueIntialized) {
        throw new TypeError('Invalid argument: from_value in IntegerRange');
      }
      if (!this._to_valueIntialized) {
        throw new TypeError('Invalid argument: to_value in IntegerRange');
      }
      if (!this._stepIntialized) {
        throw new TypeError('Invalid argument: step in IntegerRange');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._from_valueIntialized = true;
    this._refObject.from_value = refObject.from_value;
    this._to_valueIntialized = true;
    this._refObject.to_value = refObject.to_value;
    this._stepIntialized = true;
    this._refObject.step = refObject.step;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    IntegerRangeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'IntegerRange'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get from_value() {
    if (!this._from_valueIntialized) {
      return undefined;
    }
    return this._refObject.from_value;
  }
  set from_value(value) {
    this._from_valueIntialized = true;
    this._refObject.from_value = value;
  }
  get to_value() {
    if (!this._to_valueIntialized) {
      return undefined;
    }
    return this._refObject.to_value;
  }
  set to_value(value) {
    this._to_valueIntialized = true;
    this._refObject.to_value = value;
  }
  get step() {
    if (!this._stepIntialized) {
      return undefined;
    }
    return this._refObject.step;
  }
  set step(value) {
    this._stepIntialized = true;
    this._refObject.step = value;
  }
  copyRefObject(refObject) {
    this._refObject = new IntegerRangeRefStruct(refObject.toObject());
    this._from_valueIntialized = true;
    this._to_valueIntialized = true;
    this._stepIntialized = true;
  }
  copy(other) {
    this._refObject = new IntegerRangeRefStruct(other._refObject.toObject());
    this._from_valueIntialized = true;
    this._to_valueIntialized = true;
    this._stepIntialized = true;
  }
  static get classType() {
    return IntegerRangeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"from_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"to_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"step","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"IntegerRange","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IntegerRange"};
  }
  hasMember(name) {
    let memberNames = ["from_value","to_value","step"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class IntegerRangeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IntegerRangeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IntegerRangeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IntegerRangeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IntegerRangeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new IntegerRangeRefArray(size);
    this._refObject = new IntegerRangeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IntegerRangeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IntegerRangeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IntegerRangeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IntegerRangeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IntegerRangeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IntegerRangeArrayWrapper;
  }
}
module.exports = IntegerRangeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "from_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "to_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "step",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "IntegerRange",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "IntegerRange"
}
*/
