// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__SetParametersResult.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SetParametersResultRefStruct = StructType({
  successful: primitiveTypes.bool,
  reason: primitiveTypes.string,
});
const SetParametersResultRefArray = ArrayType(SetParametersResultRefStruct);
const SetParametersResultRefStructArray = StructType({
  data: SetParametersResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetParametersResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetParametersResultRefStruct(other._refObject.toObject());
      this._successfulIntialized = true;
      this._reasonIntialized = true;
      this._wrapperFields.reason =  new StringWrapper(other._wrapperFields.reason);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetParametersResultRefStruct();
    this._successfulIntialized = false;
    this._reasonIntialized = false;
    this._wrapperFields.reason = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetParametersResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetParametersResultArrayWrapper;
  }
  static get ArrayType() {
    return SetParametersResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetParametersResultRefStructArray
  }
  static get refObjectType() {
    return SetParametersResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successfulIntialized) {
        throw new TypeError('Invalid argument: successful in SetParametersResult');
      }
      if (!this._reasonIntialized) {
        throw new TypeError('Invalid argument: reason in SetParametersResult');
      }
    }
    if (own) {
      this._wrapperFields.reason.freeze(own, checkConsistency);
    }
    this._refObject.reason = this._wrapperFields.reason.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successfulIntialized = true;
    this._refObject.successful = refObject.successful;
    this._reasonIntialized = true;
    this._wrapperFields.reason.data = refObject.reason.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.reason);
  }
  static destoryRawROS(msg) {
    SetParametersResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'SetParametersResult'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get successful() {
    if (!this._successfulIntialized) {
      return undefined;
    }
    return this._refObject.successful;
  }
  set successful(value) {
    this._successfulIntialized = true;
    this._refObject.successful = value;
  }
  get reason() {
    if (!this._reasonIntialized) {
      return undefined;
    }
    return this._wrapperFields.reason.data;
  }
  set reason(value) {
    this._reasonIntialized = true;
    this._wrapperFields.reason.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetParametersResultRefStruct(refObject.toObject());
    this._successfulIntialized = true;
    this._reasonIntialized = true;
    this._wrapperFields.reason.copyRefObject(this._refObject.reason);
  }
  copy(other) {
    this._refObject = new SetParametersResultRefStruct(other._refObject.toObject());
    this._successfulIntialized = true;
    this._reasonIntialized = true;
    this._wrapperFields.reason.copy(other._wrapperFields.reason);
  }
  static get classType() {
    return SetParametersResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"successful","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reason","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"SetParametersResult","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetParametersResult"};
  }
  hasMember(name) {
    let memberNames = ["successful","reason"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetParametersResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetParametersResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetParametersResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetParametersResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetParametersResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetParametersResultRefArray(size);
    this._refObject = new SetParametersResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetParametersResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetParametersResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetParametersResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetParametersResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetParametersResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetParametersResultArrayWrapper;
  }
}
module.exports = SetParametersResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "successful",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reason",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "SetParametersResult",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetParametersResult"
}
*/
