// rclnodejs/generated/nav_msgs/nav_msgs__msg__MapMetaData.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const MapMetaDataRefStruct = StructType({
  map_load_time: TimeWrapper.refObjectType,
  resolution: primitiveTypes.float32,
  width: primitiveTypes.uint32,
  height: primitiveTypes.uint32,
  origin: PoseWrapper.refObjectType,
});
const MapMetaDataRefArray = ArrayType(MapMetaDataRefStruct);
const MapMetaDataRefStructArray = StructType({
  data: MapMetaDataRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MapMetaDataWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MapMetaDataRefStruct(other._refObject.toObject());
      this._wrapperFields.map_load_time =  new TimeWrapper(other._wrapperFields.map_load_time);
      this._resolutionIntialized = true;
      this._widthIntialized = true;
      this._heightIntialized = true;
      this._wrapperFields.origin =  new PoseWrapper(other._wrapperFields.origin);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MapMetaDataRefStruct();
    this._wrapperFields.map_load_time = new TimeWrapper();
    this._resolutionIntialized = false;
    this._widthIntialized = false;
    this._heightIntialized = false;
    this._wrapperFields.origin = new PoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MapMetaDataWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MapMetaDataArrayWrapper;
  }
  static get ArrayType() {
    return MapMetaDataArrayWrapper;
  }
  static get refObjectArrayType() {
    return MapMetaDataRefStructArray
  }
  static get refObjectType() {
    return MapMetaDataRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._resolutionIntialized) {
        throw new TypeError('Invalid argument: resolution in MapMetaData');
      }
      if (!this._widthIntialized) {
        throw new TypeError('Invalid argument: width in MapMetaData');
      }
      if (!this._heightIntialized) {
        throw new TypeError('Invalid argument: height in MapMetaData');
      }
    }
    this._wrapperFields.map_load_time.freeze(own, checkConsistency);
    this._refObject.map_load_time = this._wrapperFields.map_load_time.refObject;
    this._wrapperFields.origin.freeze(own, checkConsistency);
    this._refObject.origin = this._wrapperFields.origin.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map_load_time.copyRefObject(refObject.map_load_time);
    this._resolutionIntialized = true;
    this._refObject.resolution = refObject.resolution;
    this._widthIntialized = true;
    this._refObject.width = refObject.width;
    this._heightIntialized = true;
    this._refObject.height = refObject.height;
    this._wrapperFields.origin.copyRefObject(refObject.origin);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.map_load_time);
    PoseWrapper.freeStruct(refObject.origin);
  }
  static destoryRawROS(msg) {
    MapMetaDataWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'msg', interfaceName: 'MapMetaData'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map_load_time() {
    return this._wrapperFields.map_load_time;
  }
  set map_load_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.map_load_time.copy(value);
    } else {
      this._wrapperFields.map_load_time.copy(new TimeWrapper(value));
    }
  }
  get resolution() {
    if (!this._resolutionIntialized) {
      return undefined;
    }
    return this._refObject.resolution;
  }
  set resolution(value) {
    this._resolutionIntialized = true;
    this._refObject.resolution = value;
  }
  get width() {
    if (!this._widthIntialized) {
      return undefined;
    }
    return this._refObject.width;
  }
  set width(value) {
    this._widthIntialized = true;
    this._refObject.width = value;
  }
  get height() {
    if (!this._heightIntialized) {
      return undefined;
    }
    return this._refObject.height;
  }
  set height(value) {
    this._heightIntialized = true;
    this._refObject.height = value;
  }
  get origin() {
    return this._wrapperFields.origin;
  }
  set origin(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.origin.copy(value);
    } else {
      this._wrapperFields.origin.copy(new PoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new MapMetaDataRefStruct(refObject.toObject());
    this._wrapperFields.map_load_time.copyRefObject(this._refObject.map_load_time);
    this._resolutionIntialized = true;
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._wrapperFields.origin.copyRefObject(this._refObject.origin);
  }
  copy(other) {
    this._refObject = new MapMetaDataRefStruct(other._refObject.toObject());
    this._wrapperFields.map_load_time.copy(other._wrapperFields.map_load_time);
    this._resolutionIntialized = true;
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._wrapperFields.origin.copy(other._wrapperFields.origin);
  }
  static get classType() {
    return MapMetaDataWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map_load_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"resolution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"origin","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"MapMetaData","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MapMetaData"};
  }
  hasMember(name) {
    let memberNames = ["map_load_time","resolution","width","height","origin"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MapMetaDataArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MapMetaDataWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MapMetaDataWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MapMetaDataArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MapMetaDataArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MapMetaDataRefArray(size);
    this._refObject = new MapMetaDataRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MapMetaDataWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MapMetaDataArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MapMetaDataArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MapMetaDataWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MapMetaDataWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MapMetaDataArrayWrapper;
  }
}
module.exports = MapMetaDataWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "map_load_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "resolution",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "origin",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav_msgs",
    "type": "MapMetaData",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MapMetaData"
}
*/
