// rclnodejs/generated/nav_msgs/nav_msgs__srv__GetPlan_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PathWrapper = require('../../generated/nav_msgs/nav_msgs__msg__Path.js');
const GetPlan_ResponseRefStruct = StructType({
  plan: PathWrapper.refObjectType,
});
const GetPlan_ResponseRefArray = ArrayType(GetPlan_ResponseRefStruct);
const GetPlan_ResponseRefStructArray = StructType({
  data: GetPlan_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetPlan_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPlan_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.plan =  new PathWrapper(other._wrapperFields.plan);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetPlan_ResponseRefStruct();
    this._wrapperFields.plan = new PathWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetPlan_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPlan_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetPlan_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPlan_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetPlan_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.plan.freeze(own, checkConsistency);
    this._refObject.plan = this._wrapperFields.plan.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.plan.copyRefObject(refObject.plan);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PathWrapper.freeStruct(refObject.plan);
  }
  static destoryRawROS(msg) {
    GetPlan_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'srv', interfaceName: 'GetPlan_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get plan() {
    return this._wrapperFields.plan;
  }
  set plan(value) {
    if (value instanceof PathWrapper) {
      this._wrapperFields.plan.copy(value);
    } else {
      this._wrapperFields.plan.copy(new PathWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetPlan_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.plan.copyRefObject(this._refObject.plan);
  }
  copy(other) {
    this._refObject = new GetPlan_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.plan.copy(other._wrapperFields.plan);
  }
  static get classType() {
    return GetPlan_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"plan","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"Path","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"GetPlan_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPlan_Response"};
  }
  hasMember(name) {
    let memberNames = ["plan"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetPlan_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPlan_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPlan_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPlan_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPlan_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetPlan_ResponseRefArray(size);
    this._refObject = new GetPlan_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPlan_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPlan_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPlan_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPlan_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPlan_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPlan_ResponseArrayWrapper;
  }
}
module.exports = GetPlan_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "plan",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_msgs",
        "type": "Path",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav_msgs",
    "type": "GetPlan_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetPlan_Response"
}
*/
