// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__ParameterEvent.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ParameterWrapper = require('../../generated/rcl_interfaces/rcl_interfaces__msg__Parameter.js');
const ParameterEventRefStruct = StructType({
  stamp: TimeWrapper.refObjectType,
  node: primitiveTypes.string,
  new_parameters: ParameterWrapper.refObjectArrayType,
  changed_parameters: ParameterWrapper.refObjectArrayType,
  deleted_parameters: ParameterWrapper.refObjectArrayType,
});
const ParameterEventRefArray = ArrayType(ParameterEventRefStruct);
const ParameterEventRefStructArray = StructType({
  data: ParameterEventRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ParameterEventWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParameterEventRefStruct(other._refObject.toObject());
      this._wrapperFields.stamp =  new TimeWrapper(other._wrapperFields.stamp);
      this._nodeIntialized = true;
      this._wrapperFields.node =  new StringWrapper(other._wrapperFields.node);
      this._wrapperFields.new_parameters = ParameterWrapper.createArray();
      this._wrapperFields.new_parameters.copy(other._wrapperFields.new_parameters);
      this._wrapperFields.changed_parameters = ParameterWrapper.createArray();
      this._wrapperFields.changed_parameters.copy(other._wrapperFields.changed_parameters);
      this._wrapperFields.deleted_parameters = ParameterWrapper.createArray();
      this._wrapperFields.deleted_parameters.copy(other._wrapperFields.deleted_parameters);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ParameterEventRefStruct();
    this._wrapperFields.stamp = new TimeWrapper();
    this._nodeIntialized = false;
    this._wrapperFields.node = new StringWrapper();
    this._wrapperFields.new_parameters = ParameterWrapper.createArray();
    this._wrapperFields.changed_parameters = ParameterWrapper.createArray();
    this._wrapperFields.deleted_parameters = ParameterWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ParameterEventWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParameterEventArrayWrapper;
  }
  static get ArrayType() {
    return ParameterEventArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParameterEventRefStructArray
  }
  static get refObjectType() {
    return ParameterEventRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._nodeIntialized) {
        throw new TypeError('Invalid argument: node in ParameterEvent');
      }
    }
    this._wrapperFields.stamp.freeze(own, checkConsistency);
    this._refObject.stamp = this._wrapperFields.stamp.refObject;
    if (own) {
      this._wrapperFields.node.freeze(own, checkConsistency);
    }
    this._refObject.node = this._wrapperFields.node.refObject;
    this._wrapperFields.new_parameters.freeze(own, checkConsistency);
    this._refObject.new_parameters = this._wrapperFields.new_parameters.refObject;
    this._wrapperFields.changed_parameters.freeze(own, checkConsistency);
    this._refObject.changed_parameters = this._wrapperFields.changed_parameters.refObject;
    this._wrapperFields.deleted_parameters.freeze(own, checkConsistency);
    this._refObject.deleted_parameters = this._wrapperFields.deleted_parameters.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.stamp.copyRefObject(refObject.stamp);
    this._nodeIntialized = true;
    this._wrapperFields.node.data = refObject.node.data;
    this._wrapperFields.new_parameters.copyRefObject(refObject.new_parameters);
    this._wrapperFields.changed_parameters.copyRefObject(refObject.changed_parameters);
    this._wrapperFields.deleted_parameters.copyRefObject(refObject.deleted_parameters);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.stamp);
    StringWrapper.freeStruct(refObject.node);
    if (refObject.new_parameters.size != 0) {
      ParameterWrapper.ArrayType.freeArray(refObject.new_parameters);
      if (ParameterWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.new_parameters, ParameterWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.changed_parameters.size != 0) {
      ParameterWrapper.ArrayType.freeArray(refObject.changed_parameters);
      if (ParameterWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.changed_parameters, ParameterWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.deleted_parameters.size != 0) {
      ParameterWrapper.ArrayType.freeArray(refObject.deleted_parameters);
      if (ParameterWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.deleted_parameters, ParameterWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ParameterEventWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'ParameterEvent'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get stamp() {
    return this._wrapperFields.stamp;
  }
  set stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.stamp.copy(value);
    } else {
      this._wrapperFields.stamp.copy(new TimeWrapper(value));
    }
  }
  get node() {
    if (!this._nodeIntialized) {
      return undefined;
    }
    return this._wrapperFields.node.data;
  }
  set node(value) {
    this._nodeIntialized = true;
    this._wrapperFields.node.data = value;
  }
  get new_parameters() {
    return this._wrapperFields.new_parameters;
  }
  set new_parameters(value) {
    this._wrapperFields.new_parameters.fill(value);
  }
  get changed_parameters() {
    return this._wrapperFields.changed_parameters;
  }
  set changed_parameters(value) {
    this._wrapperFields.changed_parameters.fill(value);
  }
  get deleted_parameters() {
    return this._wrapperFields.deleted_parameters;
  }
  set deleted_parameters(value) {
    this._wrapperFields.deleted_parameters.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ParameterEventRefStruct(refObject.toObject());
    this._wrapperFields.stamp.copyRefObject(this._refObject.stamp);
    this._nodeIntialized = true;
    this._wrapperFields.node.copyRefObject(this._refObject.node);
    this._wrapperFields.new_parameters.copyRefObject(this._refObject.new_parameters);
    this._wrapperFields.changed_parameters.copyRefObject(this._refObject.changed_parameters);
    this._wrapperFields.deleted_parameters.copyRefObject(this._refObject.deleted_parameters);
  }
  copy(other) {
    this._refObject = new ParameterEventRefStruct(other._refObject.toObject());
    this._wrapperFields.stamp.copy(other._wrapperFields.stamp);
    this._nodeIntialized = true;
    this._wrapperFields.node.copy(other._wrapperFields.node);
    this._wrapperFields.new_parameters.copy(other._wrapperFields.new_parameters);
    this._wrapperFields.changed_parameters.copy(other._wrapperFields.changed_parameters);
    this._wrapperFields.deleted_parameters.copy(other._wrapperFields.deleted_parameters);
  }
  static get classType() {
    return ParameterEventWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"node","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"new_parameters","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rcl_interfaces","type":"Parameter","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"changed_parameters","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rcl_interfaces","type":"Parameter","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"deleted_parameters","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rcl_interfaces","type":"Parameter","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"ParameterEvent","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParameterEvent"};
  }
  hasMember(name) {
    let memberNames = ["stamp","node","new_parameters","changed_parameters","deleted_parameters"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ParameterEventArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParameterEventWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParameterEventWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParameterEventArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParameterEventArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ParameterEventRefArray(size);
    this._refObject = new ParameterEventRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParameterEventWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParameterEventArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParameterEventArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParameterEventWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParameterEventWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParameterEventArrayWrapper;
  }
}
module.exports = ParameterEventWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "stamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "node",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "new_parameters",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rcl_interfaces",
        "type": "Parameter",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "changed_parameters",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rcl_interfaces",
        "type": "Parameter",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "deleted_parameters",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rcl_interfaces",
        "type": "Parameter",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "ParameterEvent",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ParameterEvent"
}
*/
