// rclnodejs/generated/rcl_interfaces/rcl_interfaces__srv__SetParameters_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ParameterWrapper = require('../../generated/rcl_interfaces/rcl_interfaces__msg__Parameter.js');
const SetParameters_RequestRefStruct = StructType({
  parameters: ParameterWrapper.refObjectArrayType,
});
const SetParameters_RequestRefArray = ArrayType(SetParameters_RequestRefStruct);
const SetParameters_RequestRefStructArray = StructType({
  data: SetParameters_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetParameters_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetParameters_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.parameters = ParameterWrapper.createArray();
      this._wrapperFields.parameters.copy(other._wrapperFields.parameters);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetParameters_RequestRefStruct();
    this._wrapperFields.parameters = ParameterWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SetParameters_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetParameters_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetParameters_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetParameters_RequestRefStructArray
  }
  static get refObjectType() {
    return SetParameters_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.parameters.freeze(own, checkConsistency);
    this._refObject.parameters = this._wrapperFields.parameters.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.parameters.copyRefObject(refObject.parameters);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.parameters.size != 0) {
      ParameterWrapper.ArrayType.freeArray(refObject.parameters);
      if (ParameterWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.parameters, ParameterWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    SetParameters_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'srv', interfaceName: 'SetParameters_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get parameters() {
    return this._wrapperFields.parameters;
  }
  set parameters(value) {
    this._wrapperFields.parameters.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SetParameters_RequestRefStruct(refObject.toObject());
    this._wrapperFields.parameters.copyRefObject(this._refObject.parameters);
  }
  copy(other) {
    this._refObject = new SetParameters_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.parameters.copy(other._wrapperFields.parameters);
  }
  static get classType() {
    return SetParameters_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"parameters","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rcl_interfaces","type":"Parameter","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"SetParameters_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetParameters_Request"};
  }
  hasMember(name) {
    let memberNames = ["parameters"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetParameters_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetParameters_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetParameters_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetParameters_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetParameters_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetParameters_RequestRefArray(size);
    this._refObject = new SetParameters_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetParameters_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetParameters_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetParameters_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetParameters_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetParameters_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetParameters_RequestArrayWrapper;
  }
}
module.exports = SetParameters_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "parameters",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rcl_interfaces",
        "type": "Parameter",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "SetParameters_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetParameters_Request"
}
*/
