// rclnodejs/generated/rcl_interfaces/rcl_interfaces__srv__SetParametersAtomically_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetParametersResultWrapper = require('../../generated/rcl_interfaces/rcl_interfaces__msg__SetParametersResult.js');
const SetParametersAtomically_ResponseRefStruct = StructType({
  result: SetParametersResultWrapper.refObjectType,
});
const SetParametersAtomically_ResponseRefArray = ArrayType(SetParametersAtomically_ResponseRefStruct);
const SetParametersAtomically_ResponseRefStructArray = StructType({
  data: SetParametersAtomically_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetParametersAtomically_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetParametersAtomically_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.result =  new SetParametersResultWrapper(other._wrapperFields.result);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetParametersAtomically_ResponseRefStruct();
    this._wrapperFields.result = new SetParametersResultWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetParametersAtomically_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetParametersAtomically_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return SetParametersAtomically_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetParametersAtomically_ResponseRefStructArray
  }
  static get refObjectType() {
    return SetParametersAtomically_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.result.freeze(own, checkConsistency);
    this._refObject.result = this._wrapperFields.result.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.result.copyRefObject(refObject.result);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    SetParametersResultWrapper.freeStruct(refObject.result);
  }
  static destoryRawROS(msg) {
    SetParametersAtomically_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'srv', interfaceName: 'SetParametersAtomically_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get result() {
    return this._wrapperFields.result;
  }
  set result(value) {
    if (value instanceof SetParametersResultWrapper) {
      this._wrapperFields.result.copy(value);
    } else {
      this._wrapperFields.result.copy(new SetParametersResultWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetParametersAtomically_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.result.copyRefObject(this._refObject.result);
  }
  copy(other) {
    this._refObject = new SetParametersAtomically_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.result.copy(other._wrapperFields.result);
  }
  static get classType() {
    return SetParametersAtomically_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"result","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"rcl_interfaces","type":"SetParametersResult","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"SetParametersAtomically_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetParametersAtomically_Response"};
  }
  hasMember(name) {
    let memberNames = ["result"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetParametersAtomically_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetParametersAtomically_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetParametersAtomically_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetParametersAtomically_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetParametersAtomically_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetParametersAtomically_ResponseRefArray(size);
    this._refObject = new SetParametersAtomically_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetParametersAtomically_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetParametersAtomically_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetParametersAtomically_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetParametersAtomically_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetParametersAtomically_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetParametersAtomically_ResponseArrayWrapper;
  }
}
module.exports = SetParametersAtomically_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "result",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "rcl_interfaces",
        "type": "SetParametersResult",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rcl_interfaces",
    "type": "SetParametersAtomically_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetParametersAtomically_Response"
}
*/
