// rclnodejs/generated/std_msgs/std_msgs__msg__MultiArrayLayout.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MultiArrayDimensionWrapper = require('../../generated/std_msgs/std_msgs__msg__MultiArrayDimension.js');
const MultiArrayLayoutRefStruct = StructType({
  dim: MultiArrayDimensionWrapper.refObjectArrayType,
  data_offset: primitiveTypes.uint32,
});
const MultiArrayLayoutRefArray = ArrayType(MultiArrayLayoutRefStruct);
const MultiArrayLayoutRefStructArray = StructType({
  data: MultiArrayLayoutRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MultiArrayLayoutWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MultiArrayLayoutRefStruct(other._refObject.toObject());
      this._wrapperFields.dim = MultiArrayDimensionWrapper.createArray();
      this._wrapperFields.dim.copy(other._wrapperFields.dim);
      this._data_offsetIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MultiArrayLayoutRefStruct();
    this._wrapperFields.dim = MultiArrayDimensionWrapper.createArray();
    this._data_offsetIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MultiArrayLayoutWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MultiArrayLayoutArrayWrapper;
  }
  static get ArrayType() {
    return MultiArrayLayoutArrayWrapper;
  }
  static get refObjectArrayType() {
    return MultiArrayLayoutRefStructArray
  }
  static get refObjectType() {
    return MultiArrayLayoutRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._data_offsetIntialized) {
        throw new TypeError('Invalid argument: data_offset in MultiArrayLayout');
      }
    }
    this._wrapperFields.dim.freeze(own, checkConsistency);
    this._refObject.dim = this._wrapperFields.dim.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.dim.copyRefObject(refObject.dim);
    this._data_offsetIntialized = true;
    this._refObject.data_offset = refObject.data_offset;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.dim.size != 0) {
      MultiArrayDimensionWrapper.ArrayType.freeArray(refObject.dim);
      if (MultiArrayDimensionWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.dim, MultiArrayDimensionWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MultiArrayLayoutWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'std_msgs', subFolder: 'msg', interfaceName: 'MultiArrayLayout'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get dim() {
    return this._wrapperFields.dim;
  }
  set dim(value) {
    this._wrapperFields.dim.fill(value);
  }
  get data_offset() {
    if (!this._data_offsetIntialized) {
      return undefined;
    }
    return this._refObject.data_offset;
  }
  set data_offset(value) {
    this._data_offsetIntialized = true;
    this._refObject.data_offset = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MultiArrayLayoutRefStruct(refObject.toObject());
    this._wrapperFields.dim.copyRefObject(this._refObject.dim);
    this._data_offsetIntialized = true;
  }
  copy(other) {
    this._refObject = new MultiArrayLayoutRefStruct(other._refObject.toObject());
    this._wrapperFields.dim.copy(other._wrapperFields.dim);
    this._data_offsetIntialized = true;
  }
  static get classType() {
    return MultiArrayLayoutWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"dim","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"std_msgs","type":"MultiArrayDimension","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"data_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"std_msgs","type":"MultiArrayLayout","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MultiArrayLayout"};
  }
  hasMember(name) {
    let memberNames = ["dim","data_offset"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MultiArrayLayoutArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MultiArrayLayoutWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MultiArrayLayoutWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiArrayLayoutArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiArrayLayoutArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MultiArrayLayoutRefArray(size);
    this._refObject = new MultiArrayLayoutRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MultiArrayLayoutWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MultiArrayLayoutArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MultiArrayLayoutArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MultiArrayLayoutWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MultiArrayLayoutWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MultiArrayLayoutArrayWrapper;
  }
}
module.exports = MultiArrayLayoutWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "dim",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "std_msgs",
        "type": "MultiArrayDimension",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "data_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "std_msgs",
    "type": "MultiArrayLayout",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MultiArrayLayout"
}
*/
