// rclnodejs/generated/std_msgs/std_msgs__msg__ColorRGBA.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ColorRGBARefStruct = StructType({
  r: primitiveTypes.float32,
  g: primitiveTypes.float32,
  b: primitiveTypes.float32,
  a: primitiveTypes.float32,
});
const ColorRGBARefArray = ArrayType(ColorRGBARefStruct);
const ColorRGBARefStructArray = StructType({
  data: ColorRGBARefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ColorRGBAWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ColorRGBARefStruct(other._refObject.toObject());
      this._rIntialized = true;
      this._gIntialized = true;
      this._bIntialized = true;
      this._aIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ColorRGBARefStruct();
    this._rIntialized = false;
    this._gIntialized = false;
    this._bIntialized = false;
    this._aIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ColorRGBAWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ColorRGBAArrayWrapper;
  }
  static get ArrayType() {
    return ColorRGBAArrayWrapper;
  }
  static get refObjectArrayType() {
    return ColorRGBARefStructArray
  }
  static get refObjectType() {
    return ColorRGBARefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._rIntialized) {
        throw new TypeError('Invalid argument: r in ColorRGBA');
      }
      if (!this._gIntialized) {
        throw new TypeError('Invalid argument: g in ColorRGBA');
      }
      if (!this._bIntialized) {
        throw new TypeError('Invalid argument: b in ColorRGBA');
      }
      if (!this._aIntialized) {
        throw new TypeError('Invalid argument: a in ColorRGBA');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._rIntialized = true;
    this._refObject.r = refObject.r;
    this._gIntialized = true;
    this._refObject.g = refObject.g;
    this._bIntialized = true;
    this._refObject.b = refObject.b;
    this._aIntialized = true;
    this._refObject.a = refObject.a;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ColorRGBAWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'std_msgs', subFolder: 'msg', interfaceName: 'ColorRGBA'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get r() {
    if (!this._rIntialized) {
      return undefined;
    }
    return this._refObject.r;
  }
  set r(value) {
    this._rIntialized = true;
    this._refObject.r = value;
  }
  get g() {
    if (!this._gIntialized) {
      return undefined;
    }
    return this._refObject.g;
  }
  set g(value) {
    this._gIntialized = true;
    this._refObject.g = value;
  }
  get b() {
    if (!this._bIntialized) {
      return undefined;
    }
    return this._refObject.b;
  }
  set b(value) {
    this._bIntialized = true;
    this._refObject.b = value;
  }
  get a() {
    if (!this._aIntialized) {
      return undefined;
    }
    return this._refObject.a;
  }
  set a(value) {
    this._aIntialized = true;
    this._refObject.a = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ColorRGBARefStruct(refObject.toObject());
    this._rIntialized = true;
    this._gIntialized = true;
    this._bIntialized = true;
    this._aIntialized = true;
  }
  copy(other) {
    this._refObject = new ColorRGBARefStruct(other._refObject.toObject());
    this._rIntialized = true;
    this._gIntialized = true;
    this._bIntialized = true;
    this._aIntialized = true;
  }
  static get classType() {
    return ColorRGBAWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"r","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"g","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"b","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"a","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"std_msgs","type":"ColorRGBA","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ColorRGBA"};
  }
  hasMember(name) {
    let memberNames = ["r","g","b","a"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ColorRGBAArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ColorRGBAWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ColorRGBAWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ColorRGBAArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ColorRGBAArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ColorRGBARefArray(size);
    this._refObject = new ColorRGBARefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ColorRGBAWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ColorRGBAArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ColorRGBAArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ColorRGBAWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ColorRGBAWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ColorRGBAArrayWrapper;
  }
}
module.exports = ColorRGBAWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "r",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "g",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "b",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "a",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "std_msgs",
    "type": "ColorRGBA",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ColorRGBA"
}
*/
