// rclnodejs/generated/tf2_msgs/tf2_msgs__action__LookupTransform_SendGoal_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const LookupTransform_GoalWrapper = require('../../generated/tf2_msgs/tf2_msgs__action__LookupTransform_Goal.js');
const LookupTransform_SendGoal_RequestRefStruct = StructType({
  goal_id: UUIDWrapper.refObjectType,
  goal: LookupTransform_GoalWrapper.refObjectType,
});
const LookupTransform_SendGoal_RequestRefArray = ArrayType(LookupTransform_SendGoal_RequestRefStruct);
const LookupTransform_SendGoal_RequestRefStructArray = StructType({
  data: LookupTransform_SendGoal_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LookupTransform_SendGoal_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LookupTransform_SendGoal_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.goal_id =  new UUIDWrapper(other._wrapperFields.goal_id);
      this._wrapperFields.goal =  new LookupTransform_GoalWrapper(other._wrapperFields.goal);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LookupTransform_SendGoal_RequestRefStruct();
    this._wrapperFields.goal_id = new UUIDWrapper();
    this._wrapperFields.goal = new LookupTransform_GoalWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LookupTransform_SendGoal_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LookupTransform_SendGoal_RequestArrayWrapper;
  }
  static get ArrayType() {
    return LookupTransform_SendGoal_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return LookupTransform_SendGoal_RequestRefStructArray
  }
  static get refObjectType() {
    return LookupTransform_SendGoal_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.goal_id.freeze(own, checkConsistency);
    this._refObject.goal_id = this._wrapperFields.goal_id.refObject;
    this._wrapperFields.goal.freeze(own, checkConsistency);
    this._refObject.goal = this._wrapperFields.goal.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.goal_id.copyRefObject(refObject.goal_id);
    this._wrapperFields.goal.copyRefObject(refObject.goal);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    UUIDWrapper.freeStruct(refObject.goal_id);
    LookupTransform_GoalWrapper.freeStruct(refObject.goal);
  }
  static destoryRawROS(msg) {
    LookupTransform_SendGoal_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'tf2_msgs', subFolder: 'action', interfaceName: 'LookupTransform_SendGoal_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get goal_id() {
    return this._wrapperFields.goal_id;
  }
  set goal_id(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.goal_id.copy(value);
    } else {
      this._wrapperFields.goal_id.copy(new UUIDWrapper(value));
    }
  }
  get goal() {
    return this._wrapperFields.goal;
  }
  set goal(value) {
    if (value instanceof LookupTransform_GoalWrapper) {
      this._wrapperFields.goal.copy(value);
    } else {
      this._wrapperFields.goal.copy(new LookupTransform_GoalWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new LookupTransform_SendGoal_RequestRefStruct(refObject.toObject());
    this._wrapperFields.goal_id.copyRefObject(this._refObject.goal_id);
    this._wrapperFields.goal.copyRefObject(this._refObject.goal);
  }
  copy(other) {
    this._refObject = new LookupTransform_SendGoal_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.goal_id.copy(other._wrapperFields.goal_id);
    this._wrapperFields.goal.copy(other._wrapperFields.goal);
  }
  static get classType() {
    return LookupTransform_SendGoal_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"goal_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":null,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":null,"pkgName":"tf2_msgs","type":"LookupTransform_Goal","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"tf2_msgs","type":"LookupTransform_SendGoal_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LookupTransform_SendGoal_Request"};
  }
  hasMember(name) {
    let memberNames = ["goal_id","goal"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LookupTransform_SendGoal_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LookupTransform_SendGoal_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LookupTransform_SendGoal_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LookupTransform_SendGoal_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LookupTransform_SendGoal_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LookupTransform_SendGoal_RequestRefArray(size);
    this._refObject = new LookupTransform_SendGoal_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LookupTransform_SendGoal_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LookupTransform_SendGoal_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LookupTransform_SendGoal_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LookupTransform_SendGoal_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LookupTransform_SendGoal_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LookupTransform_SendGoal_RequestArrayWrapper;
  }
}
module.exports = LookupTransform_SendGoal_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "goal_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": null,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "goal",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": null,
        "pkgName": "tf2_msgs",
        "type": "LookupTransform_Goal",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "tf2_msgs",
    "type": "LookupTransform_SendGoal_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LookupTransform_SendGoal_Request"
}
*/
