// rclnodejs/generated/statistics_msgs/statistics_msgs__msg__MetricsMessage.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StatisticDataPointWrapper = require('../../generated/statistics_msgs/statistics_msgs__msg__StatisticDataPoint.js');
const MetricsMessageRefStruct = StructType({
  measurement_source_name: primitiveTypes.string,
  metrics_source: primitiveTypes.string,
  unit: primitiveTypes.string,
  window_start: TimeWrapper.refObjectType,
  window_stop: TimeWrapper.refObjectType,
  statistics: StatisticDataPointWrapper.refObjectArrayType,
});
const MetricsMessageRefArray = ArrayType(MetricsMessageRefStruct);
const MetricsMessageRefStructArray = StructType({
  data: MetricsMessageRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MetricsMessageWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MetricsMessageRefStruct(other._refObject.toObject());
      this._measurement_source_nameIntialized = true;
      this._wrapperFields.measurement_source_name =  new StringWrapper(other._wrapperFields.measurement_source_name);
      this._metrics_sourceIntialized = true;
      this._wrapperFields.metrics_source =  new StringWrapper(other._wrapperFields.metrics_source);
      this._unitIntialized = true;
      this._wrapperFields.unit =  new StringWrapper(other._wrapperFields.unit);
      this._wrapperFields.window_start =  new TimeWrapper(other._wrapperFields.window_start);
      this._wrapperFields.window_stop =  new TimeWrapper(other._wrapperFields.window_stop);
      this._wrapperFields.statistics = StatisticDataPointWrapper.createArray();
      this._wrapperFields.statistics.copy(other._wrapperFields.statistics);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MetricsMessageRefStruct();
    this._measurement_source_nameIntialized = false;
    this._wrapperFields.measurement_source_name = new StringWrapper();
    this._metrics_sourceIntialized = false;
    this._wrapperFields.metrics_source = new StringWrapper();
    this._unitIntialized = false;
    this._wrapperFields.unit = new StringWrapper();
    this._wrapperFields.window_start = new TimeWrapper();
    this._wrapperFields.window_stop = new TimeWrapper();
    this._wrapperFields.statistics = StatisticDataPointWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MetricsMessageWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MetricsMessageArrayWrapper;
  }
  static get ArrayType() {
    return MetricsMessageArrayWrapper;
  }
  static get refObjectArrayType() {
    return MetricsMessageRefStructArray
  }
  static get refObjectType() {
    return MetricsMessageRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._measurement_source_nameIntialized) {
        throw new TypeError('Invalid argument: measurement_source_name in MetricsMessage');
      }
      if (!this._metrics_sourceIntialized) {
        throw new TypeError('Invalid argument: metrics_source in MetricsMessage');
      }
      if (!this._unitIntialized) {
        throw new TypeError('Invalid argument: unit in MetricsMessage');
      }
    }
    if (own) {
      this._wrapperFields.measurement_source_name.freeze(own, checkConsistency);
    }
    this._refObject.measurement_source_name = this._wrapperFields.measurement_source_name.refObject;
    if (own) {
      this._wrapperFields.metrics_source.freeze(own, checkConsistency);
    }
    this._refObject.metrics_source = this._wrapperFields.metrics_source.refObject;
    if (own) {
      this._wrapperFields.unit.freeze(own, checkConsistency);
    }
    this._refObject.unit = this._wrapperFields.unit.refObject;
    this._wrapperFields.window_start.freeze(own, checkConsistency);
    this._refObject.window_start = this._wrapperFields.window_start.refObject;
    this._wrapperFields.window_stop.freeze(own, checkConsistency);
    this._refObject.window_stop = this._wrapperFields.window_stop.refObject;
    this._wrapperFields.statistics.freeze(own, checkConsistency);
    this._refObject.statistics = this._wrapperFields.statistics.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._measurement_source_nameIntialized = true;
    this._wrapperFields.measurement_source_name.data = refObject.measurement_source_name.data;
    this._metrics_sourceIntialized = true;
    this._wrapperFields.metrics_source.data = refObject.metrics_source.data;
    this._unitIntialized = true;
    this._wrapperFields.unit.data = refObject.unit.data;
    this._wrapperFields.window_start.copyRefObject(refObject.window_start);
    this._wrapperFields.window_stop.copyRefObject(refObject.window_stop);
    this._wrapperFields.statistics.copyRefObject(refObject.statistics);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.measurement_source_name);
    StringWrapper.freeStruct(refObject.metrics_source);
    StringWrapper.freeStruct(refObject.unit);
    TimeWrapper.freeStruct(refObject.window_start);
    TimeWrapper.freeStruct(refObject.window_stop);
    if (refObject.statistics.size != 0) {
      StatisticDataPointWrapper.ArrayType.freeArray(refObject.statistics);
      if (StatisticDataPointWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.statistics, StatisticDataPointWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MetricsMessageWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'statistics_msgs', subFolder: 'msg', interfaceName: 'MetricsMessage'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get measurement_source_name() {
    if (!this._measurement_source_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.measurement_source_name.data;
  }
  set measurement_source_name(value) {
    this._measurement_source_nameIntialized = true;
    this._wrapperFields.measurement_source_name.data = value;
  }
  get metrics_source() {
    if (!this._metrics_sourceIntialized) {
      return undefined;
    }
    return this._wrapperFields.metrics_source.data;
  }
  set metrics_source(value) {
    this._metrics_sourceIntialized = true;
    this._wrapperFields.metrics_source.data = value;
  }
  get unit() {
    if (!this._unitIntialized) {
      return undefined;
    }
    return this._wrapperFields.unit.data;
  }
  set unit(value) {
    this._unitIntialized = true;
    this._wrapperFields.unit.data = value;
  }
  get window_start() {
    return this._wrapperFields.window_start;
  }
  set window_start(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.window_start.copy(value);
    } else {
      this._wrapperFields.window_start.copy(new TimeWrapper(value));
    }
  }
  get window_stop() {
    return this._wrapperFields.window_stop;
  }
  set window_stop(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.window_stop.copy(value);
    } else {
      this._wrapperFields.window_stop.copy(new TimeWrapper(value));
    }
  }
  get statistics() {
    return this._wrapperFields.statistics;
  }
  set statistics(value) {
    this._wrapperFields.statistics.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MetricsMessageRefStruct(refObject.toObject());
    this._measurement_source_nameIntialized = true;
    this._wrapperFields.measurement_source_name.copyRefObject(this._refObject.measurement_source_name);
    this._metrics_sourceIntialized = true;
    this._wrapperFields.metrics_source.copyRefObject(this._refObject.metrics_source);
    this._unitIntialized = true;
    this._wrapperFields.unit.copyRefObject(this._refObject.unit);
    this._wrapperFields.window_start.copyRefObject(this._refObject.window_start);
    this._wrapperFields.window_stop.copyRefObject(this._refObject.window_stop);
    this._wrapperFields.statistics.copyRefObject(this._refObject.statistics);
  }
  copy(other) {
    this._refObject = new MetricsMessageRefStruct(other._refObject.toObject());
    this._measurement_source_nameIntialized = true;
    this._wrapperFields.measurement_source_name.copy(other._wrapperFields.measurement_source_name);
    this._metrics_sourceIntialized = true;
    this._wrapperFields.metrics_source.copy(other._wrapperFields.metrics_source);
    this._unitIntialized = true;
    this._wrapperFields.unit.copy(other._wrapperFields.unit);
    this._wrapperFields.window_start.copy(other._wrapperFields.window_start);
    this._wrapperFields.window_stop.copy(other._wrapperFields.window_stop);
    this._wrapperFields.statistics.copy(other._wrapperFields.statistics);
  }
  static get classType() {
    return MetricsMessageWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"measurement_source_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"metrics_source","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"unit","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"window_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"window_stop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"statistics","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"statistics_msgs","type":"StatisticDataPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"statistics_msgs","type":"MetricsMessage","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MetricsMessage"};
  }
  hasMember(name) {
    let memberNames = ["measurement_source_name","metrics_source","unit","window_start","window_stop","statistics"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MetricsMessageArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MetricsMessageWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MetricsMessageWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MetricsMessageArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MetricsMessageArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MetricsMessageRefArray(size);
    this._refObject = new MetricsMessageRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MetricsMessageWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MetricsMessageArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MetricsMessageArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MetricsMessageWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MetricsMessageWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MetricsMessageArrayWrapper;
  }
}
module.exports = MetricsMessageWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "measurement_source_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "metrics_source",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "unit",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "window_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "window_stop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "statistics",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "statistics_msgs",
        "type": "StatisticDataPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "statistics_msgs",
    "type": "MetricsMessage",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MetricsMessage"
}
*/
