// rclnodejs/generated/rosgraph_msgs/rosgraph_msgs__msg__Clock.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const ClockRefStruct = StructType({
  clock: TimeWrapper.refObjectType,
});
const ClockRefArray = ArrayType(ClockRefStruct);
const ClockRefStructArray = StructType({
  data: ClockRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ClockWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ClockRefStruct(other._refObject.toObject());
      this._wrapperFields.clock =  new TimeWrapper(other._wrapperFields.clock);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ClockRefStruct();
    this._wrapperFields.clock = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ClockWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ClockArrayWrapper;
  }
  static get ArrayType() {
    return ClockArrayWrapper;
  }
  static get refObjectArrayType() {
    return ClockRefStructArray
  }
  static get refObjectType() {
    return ClockRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.clock.freeze(own, checkConsistency);
    this._refObject.clock = this._wrapperFields.clock.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.clock.copyRefObject(refObject.clock);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.clock);
  }
  static destoryRawROS(msg) {
    ClockWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosgraph_msgs', subFolder: 'msg', interfaceName: 'Clock'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get clock() {
    return this._wrapperFields.clock;
  }
  set clock(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.clock.copy(value);
    } else {
      this._wrapperFields.clock.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ClockRefStruct(refObject.toObject());
    this._wrapperFields.clock.copyRefObject(this._refObject.clock);
  }
  copy(other) {
    this._refObject = new ClockRefStruct(other._refObject.toObject());
    this._wrapperFields.clock.copy(other._wrapperFields.clock);
  }
  static get classType() {
    return ClockWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"clock","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rosgraph_msgs","type":"Clock","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Clock"};
  }
  hasMember(name) {
    let memberNames = ["clock"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ClockArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ClockWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ClockWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ClockArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ClockArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ClockRefArray(size);
    this._refObject = new ClockRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ClockWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ClockArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ClockArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ClockWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ClockWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ClockArrayWrapper;
  }
}
module.exports = ClockWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "clock",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosgraph_msgs",
    "type": "Clock",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Clock"
}
*/
