// rclnodejs/generated/shape_msgs/shape_msgs__msg__MeshTriangle.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const MeshTriangleRefStruct = StructType({
  vertex_indices: ArrayType(primitiveTypes.uint32, 3),
});
const MeshTriangleRefArray = ArrayType(MeshTriangleRefStruct);
const MeshTriangleRefStructArray = StructType({
  data: MeshTriangleRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MeshTriangleWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MeshTriangleRefStruct(other._refObject.toObject());
      this._wrapperFields.vertex_indices = UInt32Wrapper.createArray();
      this._wrapperFields.vertex_indices.copy(other._wrapperFields.vertex_indices);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MeshTriangleRefStruct();
    this._wrapperFields.vertex_indices = UInt32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MeshTriangleWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MeshTriangleArrayWrapper;
  }
  static get ArrayType() {
    return MeshTriangleArrayWrapper;
  }
  static get refObjectArrayType() {
    return MeshTriangleRefStructArray
  }
  static get refObjectType() {
    return MeshTriangleRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    for (let i = 0; i < this._wrapperFields.vertex_indices.data.length; i++) {
      this._refObject.vertex_indices[i] = this._wrapperFields.vertex_indices.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.vertex_indices = refObject.vertex_indices;
    this._wrapperFields.vertex_indices.fill(refObject.vertex_indices.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    MeshTriangleWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'shape_msgs', subFolder: 'msg', interfaceName: 'MeshTriangle'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get vertex_indices() {
    return this._wrapperFields['vertex_indices'].data;
  }
  set vertex_indices(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['vertex_indices'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MeshTriangleRefStruct(refObject.toObject());
    this._wrapperFields.vertex_indices.fill(refObject.vertex_indices.toArray());
  }
  copy(other) {
    this._refObject = new MeshTriangleRefStruct(other._refObject.toObject());
    this._wrapperFields.vertex_indices.copy(other._wrapperFields.vertex_indices);
  }
  static get classType() {
    return MeshTriangleWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"vertex_indices","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"shape_msgs","type":"MeshTriangle","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MeshTriangle"};
  }
  hasMember(name) {
    let memberNames = ["vertex_indices"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MeshTriangleArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MeshTriangleWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MeshTriangleWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MeshTriangleArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MeshTriangleArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MeshTriangleRefArray(size);
    this._refObject = new MeshTriangleRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MeshTriangleWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MeshTriangleArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MeshTriangleArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MeshTriangleWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MeshTriangleWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MeshTriangleArrayWrapper;
  }
}
module.exports = MeshTriangleWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "vertex_indices",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "shape_msgs",
    "type": "MeshTriangle",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MeshTriangle"
}
*/
