// rclnodejs/generated/composition_interfaces/composition_interfaces__srv__ListNodes_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const UInt64Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt64.js');
const ListNodes_ResponseRefStruct = StructType({
  full_node_names: StringWrapper.refObjectArrayType,
  unique_ids: UInt64Wrapper.refObjectArrayType,
});
const ListNodes_ResponseRefArray = ArrayType(ListNodes_ResponseRefStruct);
const ListNodes_ResponseRefStructArray = StructType({
  data: ListNodes_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ListNodes_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._full_node_namesArray = [];
    this._unique_idsArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListNodes_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.full_node_names = StringWrapper.createArray();
      this._wrapperFields.full_node_names.copy(other._wrapperFields.full_node_names);
      this.full_node_names = other.full_node_names;
      this._wrapperFields.unique_ids = UInt64Wrapper.createArray();
      this._wrapperFields.unique_ids.copy(other._wrapperFields.unique_ids);
      this.unique_ids = other.unique_ids;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ListNodes_ResponseRefStruct();
    this._wrapperFields.full_node_names = StringWrapper.createArray();
    this._wrapperFields.unique_ids = UInt64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ListNodes_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListNodes_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ListNodes_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListNodes_ResponseRefStructArray
  }
  static get refObjectType() {
    return ListNodes_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    if (!own) {
      this._wrapperFields.full_node_names.fill(this._full_node_namesArray);
      this._wrapperFields.full_node_names.freeze(own, checkConsistency);
      this._refObject.full_node_names = this._wrapperFields.full_node_names.refObject;
    } else {
      this._wrapperFields.full_node_names.fill([]);
      this._wrapperFields.full_node_names.freeze(own, checkConsistency);
      this._refObject.full_node_names = this._wrapperFields.full_node_names.refObject;
    }
    if (!own) {
      this._wrapperFields.unique_ids.fill(this._unique_idsArray);
      this._wrapperFields.unique_ids.freeze(own, checkConsistency);
      this._refObject.unique_ids = this._wrapperFields.unique_ids.refObject;
    } else {
      this._wrapperFields.unique_ids.fill([]);
      this._wrapperFields.unique_ids.freeze(own, checkConsistency);
      this._refObject.unique_ids = this._wrapperFields.unique_ids.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.full_node_names.data.length = refObject.full_node_names.size;
    for (let index = 0; index < refObject.full_node_names.size; index++) {
      this._full_node_namesArray[index] = refObject.full_node_names.data[index].data;
    }
    refObject.unique_ids.data.length = refObject.unique_ids.size;
    for (let index = 0; index < refObject.unique_ids.size; index++) {
      this._unique_idsArray[index] = refObject.unique_ids.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.full_node_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.full_node_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.full_node_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.unique_ids.size != 0) {
      UInt64Wrapper.ArrayType.freeArray(refObject.unique_ids);
      if (UInt64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.unique_ids, UInt64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ListNodes_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'composition_interfaces', subFolder: 'srv', interfaceName: 'ListNodes_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get full_node_names() {
    return this._full_node_namesArray;
  }
  set full_node_names(value) {
    this._full_node_namesArray = value;
  }
  get unique_ids() {
    return this._unique_idsArray;
  }
  set unique_ids(value) {
    this._unique_idsArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ListNodes_ResponseRefStruct(refObject.toObject());
    refObject.full_node_names.data.length = refObject.full_node_names.size;
    for (let index = 0; index < refObject.full_node_names.size; index++) {
      this._full_node_namesArray[index] = refObject.full_node_names.data[index].data;
    }
    refObject.unique_ids.data.length = refObject.unique_ids.size;
    for (let index = 0; index < refObject.unique_ids.size; index++) {
      this._unique_idsArray[index] = refObject.unique_ids.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ListNodes_ResponseRefStruct(other._refObject.toObject());
    this._full_node_namesArray = other._full_node_namesArray.slice();
    this._unique_idsArray = other._unique_idsArray.slice();
  }
  static get classType() {
    return ListNodes_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"full_node_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"unique_ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"composition_interfaces","type":"ListNodes_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListNodes_Response"};
  }
  hasMember(name) {
    let memberNames = ["full_node_names","unique_ids"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ListNodes_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListNodes_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListNodes_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListNodes_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListNodes_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ListNodes_ResponseRefArray(size);
    this._refObject = new ListNodes_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListNodes_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListNodes_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListNodes_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListNodes_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListNodes_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListNodes_ResponseArrayWrapper;
  }
}
module.exports = ListNodes_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "full_node_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "unique_ids",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "composition_interfaces",
    "type": "ListNodes_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ListNodes_Response"
}
*/
