// rclnodejs/generated/composition_interfaces/composition_interfaces__srv__UnloadNode_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const UnloadNode_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  error_message: primitiveTypes.string,
});
const UnloadNode_ResponseRefArray = ArrayType(UnloadNode_ResponseRefStruct);
const UnloadNode_ResponseRefStructArray = StructType({
  data: UnloadNode_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UnloadNode_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UnloadNode_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._error_messageIntialized = true;
      this._wrapperFields.error_message =  new StringWrapper(other._wrapperFields.error_message);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UnloadNode_ResponseRefStruct();
    this._successIntialized = false;
    this._error_messageIntialized = false;
    this._wrapperFields.error_message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new UnloadNode_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UnloadNode_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return UnloadNode_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return UnloadNode_ResponseRefStructArray
  }
  static get refObjectType() {
    return UnloadNode_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in UnloadNode_Response');
      }
      if (!this._error_messageIntialized) {
        throw new TypeError('Invalid argument: error_message in UnloadNode_Response');
      }
    }
    if (own) {
      this._wrapperFields.error_message.freeze(own, checkConsistency);
    }
    this._refObject.error_message = this._wrapperFields.error_message.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._error_messageIntialized = true;
    this._wrapperFields.error_message.data = refObject.error_message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.error_message);
  }
  static destoryRawROS(msg) {
    UnloadNode_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'composition_interfaces', subFolder: 'srv', interfaceName: 'UnloadNode_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get error_message() {
    if (!this._error_messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.error_message.data;
  }
  set error_message(value) {
    this._error_messageIntialized = true;
    this._wrapperFields.error_message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new UnloadNode_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._error_messageIntialized = true;
    this._wrapperFields.error_message.copyRefObject(this._refObject.error_message);
  }
  copy(other) {
    this._refObject = new UnloadNode_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._error_messageIntialized = true;
    this._wrapperFields.error_message.copy(other._wrapperFields.error_message);
  }
  static get classType() {
    return UnloadNode_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"composition_interfaces","type":"UnloadNode_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UnloadNode_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","error_message"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UnloadNode_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UnloadNode_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UnloadNode_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UnloadNode_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UnloadNode_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UnloadNode_ResponseRefArray(size);
    this._refObject = new UnloadNode_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UnloadNode_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UnloadNode_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UnloadNode_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UnloadNode_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UnloadNode_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UnloadNode_ResponseArrayWrapper;
  }
}
module.exports = UnloadNode_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "composition_interfaces",
    "type": "UnloadNode_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UnloadNode_Response"
}
*/
