// rclnodejs/generated/composition_interfaces/composition_interfaces__srv__UnloadNode_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UnloadNode_RequestRefStruct = StructType({
  unique_id: primitiveTypes.uint64,
});
const UnloadNode_RequestRefArray = ArrayType(UnloadNode_RequestRefStruct);
const UnloadNode_RequestRefStructArray = StructType({
  data: UnloadNode_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UnloadNode_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UnloadNode_RequestRefStruct(other._refObject.toObject());
      this._unique_idIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UnloadNode_RequestRefStruct();
    this._unique_idIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new UnloadNode_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UnloadNode_RequestArrayWrapper;
  }
  static get ArrayType() {
    return UnloadNode_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return UnloadNode_RequestRefStructArray
  }
  static get refObjectType() {
    return UnloadNode_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._unique_idIntialized) {
        throw new TypeError('Invalid argument: unique_id in UnloadNode_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._unique_idIntialized = true;
    this._refObject.unique_id = refObject.unique_id;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    UnloadNode_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'composition_interfaces', subFolder: 'srv', interfaceName: 'UnloadNode_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get unique_id() {
    if (!this._unique_idIntialized) {
      return undefined;
    }
    return this._refObject.unique_id;
  }
  set unique_id(value) {
    this._unique_idIntialized = true;
    this._refObject.unique_id = value;
  }
  copyRefObject(refObject) {
    this._refObject = new UnloadNode_RequestRefStruct(refObject.toObject());
    this._unique_idIntialized = true;
  }
  copy(other) {
    this._refObject = new UnloadNode_RequestRefStruct(other._refObject.toObject());
    this._unique_idIntialized = true;
  }
  static get classType() {
    return UnloadNode_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"unique_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"composition_interfaces","type":"UnloadNode_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UnloadNode_Request"};
  }
  hasMember(name) {
    let memberNames = ["unique_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UnloadNode_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UnloadNode_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UnloadNode_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UnloadNode_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UnloadNode_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UnloadNode_RequestRefArray(size);
    this._refObject = new UnloadNode_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UnloadNode_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UnloadNode_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UnloadNode_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UnloadNode_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UnloadNode_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UnloadNode_RequestArrayWrapper;
  }
}
module.exports = UnloadNode_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "unique_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "composition_interfaces",
    "type": "UnloadNode_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UnloadNode_Request"
}
*/
