// rclnodejs/generated/actionlib_msgs/actionlib_msgs__msg__GoalID.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GoalIDRefStruct = StructType({
  stamp: TimeWrapper.refObjectType,
  id: primitiveTypes.string,
});
const GoalIDRefArray = ArrayType(GoalIDRefStruct);
const GoalIDRefStructArray = StructType({
  data: GoalIDRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GoalIDWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GoalIDRefStruct(other._refObject.toObject());
      this._wrapperFields.stamp =  new TimeWrapper(other._wrapperFields.stamp);
      this._idIntialized = true;
      this._wrapperFields.id =  new StringWrapper(other._wrapperFields.id);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GoalIDRefStruct();
    this._wrapperFields.stamp = new TimeWrapper();
    this._idIntialized = false;
    this._wrapperFields.id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GoalIDWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GoalIDArrayWrapper;
  }
  static get ArrayType() {
    return GoalIDArrayWrapper;
  }
  static get refObjectArrayType() {
    return GoalIDRefStructArray
  }
  static get refObjectType() {
    return GoalIDRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in GoalID');
      }
    }
    this._wrapperFields.stamp.freeze(own, checkConsistency);
    this._refObject.stamp = this._wrapperFields.stamp.refObject;
    if (own) {
      this._wrapperFields.id.freeze(own, checkConsistency);
    }
    this._refObject.id = this._wrapperFields.id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.stamp.copyRefObject(refObject.stamp);
    this._idIntialized = true;
    this._wrapperFields.id.data = refObject.id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.stamp);
    StringWrapper.freeStruct(refObject.id);
  }
  static destoryRawROS(msg) {
    GoalIDWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'actionlib_msgs', subFolder: 'msg', interfaceName: 'GoalID'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get stamp() {
    return this._wrapperFields.stamp;
  }
  set stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.stamp.copy(value);
    } else {
      this._wrapperFields.stamp.copy(new TimeWrapper(value));
    }
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._wrapperFields.id.data;
  }
  set id(value) {
    this._idIntialized = true;
    this._wrapperFields.id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GoalIDRefStruct(refObject.toObject());
    this._wrapperFields.stamp.copyRefObject(this._refObject.stamp);
    this._idIntialized = true;
    this._wrapperFields.id.copyRefObject(this._refObject.id);
  }
  copy(other) {
    this._refObject = new GoalIDRefStruct(other._refObject.toObject());
    this._wrapperFields.stamp.copy(other._wrapperFields.stamp);
    this._idIntialized = true;
    this._wrapperFields.id.copy(other._wrapperFields.id);
  }
  static get classType() {
    return GoalIDWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"actionlib_msgs","type":"GoalID","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GoalID"};
  }
  hasMember(name) {
    let memberNames = ["stamp","id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GoalIDArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GoalIDWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GoalIDWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalIDArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalIDArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GoalIDRefArray(size);
    this._refObject = new GoalIDRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GoalIDWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GoalIDArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GoalIDArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GoalIDWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GoalIDWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GoalIDArrayWrapper;
  }
}
module.exports = GoalIDWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "stamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "actionlib_msgs",
    "type": "GoalID",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GoalID"
}
*/
