// rclnodejs/generated/tf2_msgs/tf2_msgs__action__LookupTransform_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const LookupTransform_GoalRefStruct = StructType({
  target_frame: primitiveTypes.string,
  source_frame: primitiveTypes.string,
  source_time: TimeWrapper.refObjectType,
  timeout: DurationWrapper.refObjectType,
  target_time: TimeWrapper.refObjectType,
  fixed_frame: primitiveTypes.string,
  advanced: primitiveTypes.bool,
});
const LookupTransform_GoalRefArray = ArrayType(LookupTransform_GoalRefStruct);
const LookupTransform_GoalRefStructArray = StructType({
  data: LookupTransform_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LookupTransform_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LookupTransform_GoalRefStruct(other._refObject.toObject());
      this._target_frameIntialized = true;
      this._wrapperFields.target_frame =  new StringWrapper(other._wrapperFields.target_frame);
      this._source_frameIntialized = true;
      this._wrapperFields.source_frame =  new StringWrapper(other._wrapperFields.source_frame);
      this._wrapperFields.source_time =  new TimeWrapper(other._wrapperFields.source_time);
      this._wrapperFields.timeout =  new DurationWrapper(other._wrapperFields.timeout);
      this._wrapperFields.target_time =  new TimeWrapper(other._wrapperFields.target_time);
      this._fixed_frameIntialized = true;
      this._wrapperFields.fixed_frame =  new StringWrapper(other._wrapperFields.fixed_frame);
      this._advancedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LookupTransform_GoalRefStruct();
    this._target_frameIntialized = false;
    this._wrapperFields.target_frame = new StringWrapper();
    this._source_frameIntialized = false;
    this._wrapperFields.source_frame = new StringWrapper();
    this._wrapperFields.source_time = new TimeWrapper();
    this._wrapperFields.timeout = new DurationWrapper();
    this._wrapperFields.target_time = new TimeWrapper();
    this._fixed_frameIntialized = false;
    this._wrapperFields.fixed_frame = new StringWrapper();
    this._advancedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LookupTransform_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LookupTransform_GoalArrayWrapper;
  }
  static get ArrayType() {
    return LookupTransform_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return LookupTransform_GoalRefStructArray
  }
  static get refObjectType() {
    return LookupTransform_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._target_frameIntialized) {
        throw new TypeError('Invalid argument: target_frame in LookupTransform_Goal');
      }
      if (!this._source_frameIntialized) {
        throw new TypeError('Invalid argument: source_frame in LookupTransform_Goal');
      }
      if (!this._fixed_frameIntialized) {
        throw new TypeError('Invalid argument: fixed_frame in LookupTransform_Goal');
      }
      if (!this._advancedIntialized) {
        throw new TypeError('Invalid argument: advanced in LookupTransform_Goal');
      }
    }
    if (own) {
      this._wrapperFields.target_frame.freeze(own, checkConsistency);
    }
    this._refObject.target_frame = this._wrapperFields.target_frame.refObject;
    if (own) {
      this._wrapperFields.source_frame.freeze(own, checkConsistency);
    }
    this._refObject.source_frame = this._wrapperFields.source_frame.refObject;
    this._wrapperFields.source_time.freeze(own, checkConsistency);
    this._refObject.source_time = this._wrapperFields.source_time.refObject;
    this._wrapperFields.timeout.freeze(own, checkConsistency);
    this._refObject.timeout = this._wrapperFields.timeout.refObject;
    this._wrapperFields.target_time.freeze(own, checkConsistency);
    this._refObject.target_time = this._wrapperFields.target_time.refObject;
    if (own) {
      this._wrapperFields.fixed_frame.freeze(own, checkConsistency);
    }
    this._refObject.fixed_frame = this._wrapperFields.fixed_frame.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._target_frameIntialized = true;
    this._wrapperFields.target_frame.data = refObject.target_frame.data;
    this._source_frameIntialized = true;
    this._wrapperFields.source_frame.data = refObject.source_frame.data;
    this._wrapperFields.source_time.copyRefObject(refObject.source_time);
    this._wrapperFields.timeout.copyRefObject(refObject.timeout);
    this._wrapperFields.target_time.copyRefObject(refObject.target_time);
    this._fixed_frameIntialized = true;
    this._wrapperFields.fixed_frame.data = refObject.fixed_frame.data;
    this._advancedIntialized = true;
    this._refObject.advanced = refObject.advanced;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.target_frame);
    StringWrapper.freeStruct(refObject.source_frame);
    TimeWrapper.freeStruct(refObject.source_time);
    DurationWrapper.freeStruct(refObject.timeout);
    TimeWrapper.freeStruct(refObject.target_time);
    StringWrapper.freeStruct(refObject.fixed_frame);
  }
  static destoryRawROS(msg) {
    LookupTransform_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'tf2_msgs', subFolder: 'action', interfaceName: 'LookupTransform_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get target_frame() {
    if (!this._target_frameIntialized) {
      return undefined;
    }
    return this._wrapperFields.target_frame.data;
  }
  set target_frame(value) {
    this._target_frameIntialized = true;
    this._wrapperFields.target_frame.data = value;
  }
  get source_frame() {
    if (!this._source_frameIntialized) {
      return undefined;
    }
    return this._wrapperFields.source_frame.data;
  }
  set source_frame(value) {
    this._source_frameIntialized = true;
    this._wrapperFields.source_frame.data = value;
  }
  get source_time() {
    return this._wrapperFields.source_time;
  }
  set source_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.source_time.copy(value);
    } else {
      this._wrapperFields.source_time.copy(new TimeWrapper(value));
    }
  }
  get timeout() {
    return this._wrapperFields.timeout;
  }
  set timeout(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.timeout.copy(value);
    } else {
      this._wrapperFields.timeout.copy(new DurationWrapper(value));
    }
  }
  get target_time() {
    return this._wrapperFields.target_time;
  }
  set target_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.target_time.copy(value);
    } else {
      this._wrapperFields.target_time.copy(new TimeWrapper(value));
    }
  }
  get fixed_frame() {
    if (!this._fixed_frameIntialized) {
      return undefined;
    }
    return this._wrapperFields.fixed_frame.data;
  }
  set fixed_frame(value) {
    this._fixed_frameIntialized = true;
    this._wrapperFields.fixed_frame.data = value;
  }
  get advanced() {
    if (!this._advancedIntialized) {
      return undefined;
    }
    return this._refObject.advanced;
  }
  set advanced(value) {
    this._advancedIntialized = true;
    this._refObject.advanced = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LookupTransform_GoalRefStruct(refObject.toObject());
    this._target_frameIntialized = true;
    this._wrapperFields.target_frame.copyRefObject(this._refObject.target_frame);
    this._source_frameIntialized = true;
    this._wrapperFields.source_frame.copyRefObject(this._refObject.source_frame);
    this._wrapperFields.source_time.copyRefObject(this._refObject.source_time);
    this._wrapperFields.timeout.copyRefObject(this._refObject.timeout);
    this._wrapperFields.target_time.copyRefObject(this._refObject.target_time);
    this._fixed_frameIntialized = true;
    this._wrapperFields.fixed_frame.copyRefObject(this._refObject.fixed_frame);
    this._advancedIntialized = true;
  }
  copy(other) {
    this._refObject = new LookupTransform_GoalRefStruct(other._refObject.toObject());
    this._target_frameIntialized = true;
    this._wrapperFields.target_frame.copy(other._wrapperFields.target_frame);
    this._source_frameIntialized = true;
    this._wrapperFields.source_frame.copy(other._wrapperFields.source_frame);
    this._wrapperFields.source_time.copy(other._wrapperFields.source_time);
    this._wrapperFields.timeout.copy(other._wrapperFields.timeout);
    this._wrapperFields.target_time.copy(other._wrapperFields.target_time);
    this._fixed_frameIntialized = true;
    this._wrapperFields.fixed_frame.copy(other._wrapperFields.fixed_frame);
    this._advancedIntialized = true;
  }
  static get classType() {
    return LookupTransform_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"target_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"source_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"source_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"timeout","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"target_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"fixed_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"advanced","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"tf2_msgs","type":"LookupTransform_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LookupTransform_Goal"};
  }
  hasMember(name) {
    let memberNames = ["target_frame","source_frame","source_time","timeout","target_time","fixed_frame","advanced"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LookupTransform_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LookupTransform_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LookupTransform_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LookupTransform_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LookupTransform_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LookupTransform_GoalRefArray(size);
    this._refObject = new LookupTransform_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LookupTransform_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LookupTransform_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LookupTransform_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LookupTransform_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LookupTransform_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LookupTransform_GoalArrayWrapper;
  }
}
module.exports = LookupTransform_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "target_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "source_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "source_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "timeout",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "target_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "fixed_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "advanced",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "tf2_msgs",
    "type": "LookupTransform_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LookupTransform_Goal"
}
*/
