// rclnodejs/generated/tf2_msgs/tf2_msgs__action__LookupTransform.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class tf2_msgs__action__LookupTransform {
  static get Goal() {
    return require('./tf2_msgs__action__LookupTransform_Goal.js');
  }
  static get Result() {
    return require('./tf2_msgs__action__LookupTransform_Result.js');
  }
  static get Feedback() {
    return require('./tf2_msgs__action__LookupTransform_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./tf2_msgs__action__LookupTransform_SendGoal.js');
      },
      get GetResultService() {
        return require('./tf2_msgs__action__LookupTransform_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./tf2_msgs__action__LookupTransform_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'tf2_msgs', subFolder: 'action', interfaceName: 'LookupTransform'};
  }
}
module.exports = tf2_msgs__action__LookupTransform;
