// rclnodejs/generated/tf2_msgs/tf2_msgs__action__LookupTransform_SendGoal_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const LookupTransform_SendGoal_ResponseRefStruct = StructType({
  accepted: primitiveTypes.bool,
  stamp: TimeWrapper.refObjectType,
});
const LookupTransform_SendGoal_ResponseRefArray = ArrayType(LookupTransform_SendGoal_ResponseRefStruct);
const LookupTransform_SendGoal_ResponseRefStructArray = StructType({
  data: LookupTransform_SendGoal_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LookupTransform_SendGoal_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LookupTransform_SendGoal_ResponseRefStruct(other._refObject.toObject());
      this._acceptedIntialized = true;
      this._wrapperFields.stamp =  new TimeWrapper(other._wrapperFields.stamp);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LookupTransform_SendGoal_ResponseRefStruct();
    this._acceptedIntialized = false;
    this._wrapperFields.stamp = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LookupTransform_SendGoal_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LookupTransform_SendGoal_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return LookupTransform_SendGoal_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return LookupTransform_SendGoal_ResponseRefStructArray
  }
  static get refObjectType() {
    return LookupTransform_SendGoal_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._acceptedIntialized) {
        throw new TypeError('Invalid argument: accepted in LookupTransform_SendGoal_Response');
      }
    }
    this._wrapperFields.stamp.freeze(own, checkConsistency);
    this._refObject.stamp = this._wrapperFields.stamp.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._acceptedIntialized = true;
    this._refObject.accepted = refObject.accepted;
    this._wrapperFields.stamp.copyRefObject(refObject.stamp);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.stamp);
  }
  static destoryRawROS(msg) {
    LookupTransform_SendGoal_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'tf2_msgs', subFolder: 'action', interfaceName: 'LookupTransform_SendGoal_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get accepted() {
    if (!this._acceptedIntialized) {
      return undefined;
    }
    return this._refObject.accepted;
  }
  set accepted(value) {
    this._acceptedIntialized = true;
    this._refObject.accepted = value;
  }
  get stamp() {
    return this._wrapperFields.stamp;
  }
  set stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.stamp.copy(value);
    } else {
      this._wrapperFields.stamp.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new LookupTransform_SendGoal_ResponseRefStruct(refObject.toObject());
    this._acceptedIntialized = true;
    this._wrapperFields.stamp.copyRefObject(this._refObject.stamp);
  }
  copy(other) {
    this._refObject = new LookupTransform_SendGoal_ResponseRefStruct(other._refObject.toObject());
    this._acceptedIntialized = true;
    this._wrapperFields.stamp.copy(other._wrapperFields.stamp);
  }
  static get classType() {
    return LookupTransform_SendGoal_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"accepted","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":null,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"tf2_msgs","type":"LookupTransform_SendGoal_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LookupTransform_SendGoal_Response"};
  }
  hasMember(name) {
    let memberNames = ["accepted","stamp"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LookupTransform_SendGoal_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LookupTransform_SendGoal_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LookupTransform_SendGoal_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LookupTransform_SendGoal_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LookupTransform_SendGoal_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LookupTransform_SendGoal_ResponseRefArray(size);
    this._refObject = new LookupTransform_SendGoal_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LookupTransform_SendGoal_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LookupTransform_SendGoal_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LookupTransform_SendGoal_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LookupTransform_SendGoal_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LookupTransform_SendGoal_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LookupTransform_SendGoal_ResponseArrayWrapper;
  }
}
module.exports = LookupTransform_SendGoal_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "accepted",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": null,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "tf2_msgs",
    "type": "LookupTransform_SendGoal_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LookupTransform_SendGoal_Response"
}
*/
