// rclnodejs/generated/tf2_msgs/tf2_msgs__action__LookupTransform_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TransformStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__TransformStamped.js');
const TF2ErrorWrapper = require('../../generated/tf2_msgs/tf2_msgs__msg__TF2Error.js');
const LookupTransform_ResultRefStruct = StructType({
  transform: TransformStampedWrapper.refObjectType,
  error: TF2ErrorWrapper.refObjectType,
});
const LookupTransform_ResultRefArray = ArrayType(LookupTransform_ResultRefStruct);
const LookupTransform_ResultRefStructArray = StructType({
  data: LookupTransform_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LookupTransform_ResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LookupTransform_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.transform =  new TransformStampedWrapper(other._wrapperFields.transform);
      this._wrapperFields.error =  new TF2ErrorWrapper(other._wrapperFields.error);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LookupTransform_ResultRefStruct();
    this._wrapperFields.transform = new TransformStampedWrapper();
    this._wrapperFields.error = new TF2ErrorWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LookupTransform_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LookupTransform_ResultArrayWrapper;
  }
  static get ArrayType() {
    return LookupTransform_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return LookupTransform_ResultRefStructArray
  }
  static get refObjectType() {
    return LookupTransform_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.transform.freeze(own, checkConsistency);
    this._refObject.transform = this._wrapperFields.transform.refObject;
    this._wrapperFields.error.freeze(own, checkConsistency);
    this._refObject.error = this._wrapperFields.error.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.transform.copyRefObject(refObject.transform);
    this._wrapperFields.error.copyRefObject(refObject.error);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TransformStampedWrapper.freeStruct(refObject.transform);
    TF2ErrorWrapper.freeStruct(refObject.error);
  }
  static destoryRawROS(msg) {
    LookupTransform_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'tf2_msgs', subFolder: 'action', interfaceName: 'LookupTransform_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get transform() {
    return this._wrapperFields.transform;
  }
  set transform(value) {
    if (value instanceof TransformStampedWrapper) {
      this._wrapperFields.transform.copy(value);
    } else {
      this._wrapperFields.transform.copy(new TransformStampedWrapper(value));
    }
  }
  get error() {
    return this._wrapperFields.error;
  }
  set error(value) {
    if (value instanceof TF2ErrorWrapper) {
      this._wrapperFields.error.copy(value);
    } else {
      this._wrapperFields.error.copy(new TF2ErrorWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new LookupTransform_ResultRefStruct(refObject.toObject());
    this._wrapperFields.transform.copyRefObject(this._refObject.transform);
    this._wrapperFields.error.copyRefObject(this._refObject.error);
  }
  copy(other) {
    this._refObject = new LookupTransform_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.transform.copy(other._wrapperFields.transform);
    this._wrapperFields.error.copy(other._wrapperFields.error);
  }
  static get classType() {
    return LookupTransform_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"transform","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"TransformStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"tf2_msgs","type":"TF2Error","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"tf2_msgs","type":"LookupTransform_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LookupTransform_Result"};
  }
  hasMember(name) {
    let memberNames = ["transform","error"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LookupTransform_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LookupTransform_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LookupTransform_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LookupTransform_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LookupTransform_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LookupTransform_ResultRefArray(size);
    this._refObject = new LookupTransform_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LookupTransform_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LookupTransform_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LookupTransform_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LookupTransform_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LookupTransform_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LookupTransform_ResultArrayWrapper;
  }
}
module.exports = LookupTransform_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "transform",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "TransformStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "tf2_msgs",
        "type": "TF2Error",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "tf2_msgs",
    "type": "LookupTransform_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LookupTransform_Result"
}
*/
