// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__LaserEcho.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const LaserEchoRefStruct = StructType({
  echoes: Float32Wrapper.refObjectArrayType,
});
const LaserEchoRefArray = ArrayType(LaserEchoRefStruct);
const LaserEchoRefStructArray = StructType({
  data: LaserEchoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LaserEchoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LaserEchoRefStruct(other._refObject.toObject());
      this._wrapperFields.echoes = Float32Wrapper.createArray();
      this._wrapperFields.echoes.copy(other._wrapperFields.echoes);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LaserEchoRefStruct();
    this._wrapperFields.echoes = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new LaserEchoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LaserEchoArrayWrapper;
  }
  static get ArrayType() {
    return LaserEchoArrayWrapper;
  }
  static get refObjectArrayType() {
    return LaserEchoRefStructArray
  }
  static get refObjectType() {
    return LaserEchoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.echoes.freeze(own, checkConsistency);
    this._refObject.echoes = this._wrapperFields.echoes.refObject;
    if (own) {
      this._wrapperFields.echoes.fill(Float32Array.from([]));
      this._wrapperFields.echoes.freeze(own, checkConsistency);
      this._refObject.echoes = this._wrapperFields.echoes.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.echoes.copyRefObject(refObject.echoes);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.echoes.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.echoes);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.echoes, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    LaserEchoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'LaserEcho'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get echoes() {
    return this._wrapperFields['echoes'].data;
  }
  set echoes(value) {
    this._wrapperFields['echoes'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new LaserEchoRefStruct(refObject.toObject());
    this._wrapperFields.echoes.copyRefObject(this._refObject.echoes);
  }
  copy(other) {
    this._refObject = new LaserEchoRefStruct(other._refObject.toObject());
    this._wrapperFields.echoes.copy(other._wrapperFields.echoes);
  }
  static get classType() {
    return LaserEchoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"echoes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"LaserEcho","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LaserEcho"};
  }
  hasMember(name) {
    let memberNames = ["echoes"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LaserEchoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LaserEchoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LaserEchoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LaserEchoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LaserEchoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LaserEchoRefArray(size);
    this._refObject = new LaserEchoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LaserEchoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LaserEchoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LaserEchoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LaserEchoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LaserEchoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LaserEchoArrayWrapper;
  }
}
module.exports = LaserEchoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "echoes",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "LaserEcho",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LaserEcho"
}
*/
