// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__NavSatStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const NavSatStatusRefStruct = StructType({
  status: primitiveTypes.int8,
  service: primitiveTypes.uint16,
});
const NavSatStatusRefArray = ArrayType(NavSatStatusRefStruct);
const NavSatStatusRefStructArray = StructType({
  data: NavSatStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NavSatStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NavSatStatusRefStruct(other._refObject.toObject());
      this._statusIntialized = true;
      this._serviceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NavSatStatusRefStruct();
    this._statusIntialized = false;
    this._serviceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new NavSatStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NavSatStatusArrayWrapper;
  }
  static get ArrayType() {
    return NavSatStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return NavSatStatusRefStructArray
  }
  static get refObjectType() {
    return NavSatStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._statusIntialized) {
        throw new TypeError('Invalid argument: status in NavSatStatus');
      }
      if (!this._serviceIntialized) {
        throw new TypeError('Invalid argument: service in NavSatStatus');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._statusIntialized = true;
    this._refObject.status = refObject.status;
    this._serviceIntialized = true;
    this._refObject.service = refObject.service;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    NavSatStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'NavSatStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status() {
    if (!this._statusIntialized) {
      return undefined;
    }
    return this._refObject.status;
  }
  set status(value) {
    this._statusIntialized = true;
    this._refObject.status = value;
  }
  get service() {
    if (!this._serviceIntialized) {
      return undefined;
    }
    return this._refObject.service;
  }
  set service(value) {
    this._serviceIntialized = true;
    this._refObject.service = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NavSatStatusRefStruct(refObject.toObject());
    this._statusIntialized = true;
    this._serviceIntialized = true;
  }
  copy(other) {
    this._refObject = new NavSatStatusRefStruct(other._refObject.toObject());
    this._statusIntialized = true;
    this._serviceIntialized = true;
  }
  static get classType() {
    return NavSatStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"STATUS_NO_FIX","value":-1},{"type":"int8","name":"STATUS_FIX","value":0},{"type":"int8","name":"STATUS_SBAS_FIX","value":1},{"type":"int8","name":"STATUS_GBAS_FIX","value":2},{"type":"uint16","name":"SERVICE_GPS","value":1},{"type":"uint16","name":"SERVICE_GLONASS","value":2},{"type":"uint16","name":"SERVICE_COMPASS","value":4},{"type":"uint16","name":"SERVICE_GALILEO","value":8}],"fields":[{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"service","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"NavSatStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NavSatStatus"};
  }
  hasMember(name) {
    let memberNames = ["status","service"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NavSatStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NavSatStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NavSatStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavSatStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavSatStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NavSatStatusRefArray(size);
    this._refObject = new NavSatStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NavSatStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NavSatStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NavSatStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NavSatStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NavSatStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NavSatStatusArrayWrapper;
  }
}
// Define constants (8 in total)
Object.defineProperty(NavSatStatusWrapper, "STATUS_NO_FIX", {value: -1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatStatusWrapper, "STATUS_FIX", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatStatusWrapper, "STATUS_SBAS_FIX", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatStatusWrapper, "STATUS_GBAS_FIX", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatStatusWrapper, "SERVICE_GPS", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatStatusWrapper, "SERVICE_GLONASS", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatStatusWrapper, "SERVICE_COMPASS", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatStatusWrapper, "SERVICE_GALILEO", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = NavSatStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "int8",
      "name": "STATUS_NO_FIX",
      "value": -1
    },
    {
      "type": "int8",
      "name": "STATUS_FIX",
      "value": 0
    },
    {
      "type": "int8",
      "name": "STATUS_SBAS_FIX",
      "value": 1
    },
    {
      "type": "int8",
      "name": "STATUS_GBAS_FIX",
      "value": 2
    },
    {
      "type": "uint16",
      "name": "SERVICE_GPS",
      "value": 1
    },
    {
      "type": "uint16",
      "name": "SERVICE_GLONASS",
      "value": 2
    },
    {
      "type": "uint16",
      "name": "SERVICE_COMPASS",
      "value": 4
    },
    {
      "type": "uint16",
      "name": "SERVICE_GALILEO",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "service",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "NavSatStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NavSatStatus"
}
*/
