// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__RegionOfInterest.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RegionOfInterestRefStruct = StructType({
  x_offset: primitiveTypes.uint32,
  y_offset: primitiveTypes.uint32,
  height: primitiveTypes.uint32,
  width: primitiveTypes.uint32,
  do_rectify: primitiveTypes.bool,
});
const RegionOfInterestRefArray = ArrayType(RegionOfInterestRefStruct);
const RegionOfInterestRefStructArray = StructType({
  data: RegionOfInterestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RegionOfInterestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RegionOfInterestRefStruct(other._refObject.toObject());
      this._x_offsetIntialized = true;
      this._y_offsetIntialized = true;
      this._heightIntialized = true;
      this._widthIntialized = true;
      this._do_rectifyIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RegionOfInterestRefStruct();
    this._x_offsetIntialized = false;
    this._y_offsetIntialized = false;
    this._heightIntialized = false;
    this._widthIntialized = false;
    this._do_rectifyIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new RegionOfInterestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RegionOfInterestArrayWrapper;
  }
  static get ArrayType() {
    return RegionOfInterestArrayWrapper;
  }
  static get refObjectArrayType() {
    return RegionOfInterestRefStructArray
  }
  static get refObjectType() {
    return RegionOfInterestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._x_offsetIntialized) {
        throw new TypeError('Invalid argument: x_offset in RegionOfInterest');
      }
      if (!this._y_offsetIntialized) {
        throw new TypeError('Invalid argument: y_offset in RegionOfInterest');
      }
      if (!this._heightIntialized) {
        throw new TypeError('Invalid argument: height in RegionOfInterest');
      }
      if (!this._widthIntialized) {
        throw new TypeError('Invalid argument: width in RegionOfInterest');
      }
      if (!this._do_rectifyIntialized) {
        throw new TypeError('Invalid argument: do_rectify in RegionOfInterest');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._x_offsetIntialized = true;
    this._refObject.x_offset = refObject.x_offset;
    this._y_offsetIntialized = true;
    this._refObject.y_offset = refObject.y_offset;
    this._heightIntialized = true;
    this._refObject.height = refObject.height;
    this._widthIntialized = true;
    this._refObject.width = refObject.width;
    this._do_rectifyIntialized = true;
    this._refObject.do_rectify = refObject.do_rectify;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    RegionOfInterestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'RegionOfInterest'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get x_offset() {
    if (!this._x_offsetIntialized) {
      return undefined;
    }
    return this._refObject.x_offset;
  }
  set x_offset(value) {
    this._x_offsetIntialized = true;
    this._refObject.x_offset = value;
  }
  get y_offset() {
    if (!this._y_offsetIntialized) {
      return undefined;
    }
    return this._refObject.y_offset;
  }
  set y_offset(value) {
    this._y_offsetIntialized = true;
    this._refObject.y_offset = value;
  }
  get height() {
    if (!this._heightIntialized) {
      return undefined;
    }
    return this._refObject.height;
  }
  set height(value) {
    this._heightIntialized = true;
    this._refObject.height = value;
  }
  get width() {
    if (!this._widthIntialized) {
      return undefined;
    }
    return this._refObject.width;
  }
  set width(value) {
    this._widthIntialized = true;
    this._refObject.width = value;
  }
  get do_rectify() {
    if (!this._do_rectifyIntialized) {
      return undefined;
    }
    return this._refObject.do_rectify;
  }
  set do_rectify(value) {
    this._do_rectifyIntialized = true;
    this._refObject.do_rectify = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RegionOfInterestRefStruct(refObject.toObject());
    this._x_offsetIntialized = true;
    this._y_offsetIntialized = true;
    this._heightIntialized = true;
    this._widthIntialized = true;
    this._do_rectifyIntialized = true;
  }
  copy(other) {
    this._refObject = new RegionOfInterestRefStruct(other._refObject.toObject());
    this._x_offsetIntialized = true;
    this._y_offsetIntialized = true;
    this._heightIntialized = true;
    this._widthIntialized = true;
    this._do_rectifyIntialized = true;
  }
  static get classType() {
    return RegionOfInterestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"x_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"do_rectify","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"RegionOfInterest","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RegionOfInterest"};
  }
  hasMember(name) {
    let memberNames = ["x_offset","y_offset","height","width","do_rectify"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RegionOfInterestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RegionOfInterestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RegionOfInterestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RegionOfInterestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RegionOfInterestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RegionOfInterestRefArray(size);
    this._refObject = new RegionOfInterestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RegionOfInterestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RegionOfInterestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RegionOfInterestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RegionOfInterestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RegionOfInterestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RegionOfInterestArrayWrapper;
  }
}
module.exports = RegionOfInterestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "x_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "do_rectify",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "RegionOfInterest",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RegionOfInterest"
}
*/
