// rclnodejs/generated/statistics_msgs/statistics_msgs__msg__StatisticDataPoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StatisticDataPointRefStruct = StructType({
  data_type: primitiveTypes.uint8,
  data: primitiveTypes.float64,
});
const StatisticDataPointRefArray = ArrayType(StatisticDataPointRefStruct);
const StatisticDataPointRefStructArray = StructType({
  data: StatisticDataPointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StatisticDataPointWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StatisticDataPointRefStruct(other._refObject.toObject());
      this._data_typeIntialized = true;
      this._dataIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StatisticDataPointRefStruct();
    this._data_typeIntialized = false;
    this._dataIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new StatisticDataPointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StatisticDataPointArrayWrapper;
  }
  static get ArrayType() {
    return StatisticDataPointArrayWrapper;
  }
  static get refObjectArrayType() {
    return StatisticDataPointRefStructArray
  }
  static get refObjectType() {
    return StatisticDataPointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._data_typeIntialized) {
        throw new TypeError('Invalid argument: data_type in StatisticDataPoint');
      }
      if (!this._dataIntialized) {
        throw new TypeError('Invalid argument: data in StatisticDataPoint');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._data_typeIntialized = true;
    this._refObject.data_type = refObject.data_type;
    this._dataIntialized = true;
    this._refObject.data = refObject.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    StatisticDataPointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'statistics_msgs', subFolder: 'msg', interfaceName: 'StatisticDataPoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get data_type() {
    if (!this._data_typeIntialized) {
      return undefined;
    }
    return this._refObject.data_type;
  }
  set data_type(value) {
    this._data_typeIntialized = true;
    this._refObject.data_type = value;
  }
  get data() {
    if (!this._dataIntialized) {
      return undefined;
    }
    return this._refObject.data;
  }
  set data(value) {
    this._dataIntialized = true;
    this._refObject.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StatisticDataPointRefStruct(refObject.toObject());
    this._data_typeIntialized = true;
    this._dataIntialized = true;
  }
  copy(other) {
    this._refObject = new StatisticDataPointRefStruct(other._refObject.toObject());
    this._data_typeIntialized = true;
    this._dataIntialized = true;
  }
  static get classType() {
    return StatisticDataPointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"data_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"statistics_msgs","type":"StatisticDataPoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StatisticDataPoint"};
  }
  hasMember(name) {
    let memberNames = ["data_type","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StatisticDataPointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StatisticDataPointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StatisticDataPointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatisticDataPointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatisticDataPointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StatisticDataPointRefArray(size);
    this._refObject = new StatisticDataPointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StatisticDataPointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StatisticDataPointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StatisticDataPointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StatisticDataPointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StatisticDataPointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StatisticDataPointArrayWrapper;
  }
}
module.exports = StatisticDataPointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "data_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "statistics_msgs",
    "type": "StatisticDataPoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StatisticDataPoint"
}
*/
