// rclnodejs/generated/statistics_msgs/statistics_msgs__msg__StatisticDataType.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StatisticDataTypeRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const StatisticDataTypeRefArray = ArrayType(StatisticDataTypeRefStruct);
const StatisticDataTypeRefStructArray = StructType({
  data: StatisticDataTypeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StatisticDataTypeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StatisticDataTypeRefStruct(other._refObject.toObject());
      this.__dummyIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StatisticDataTypeRefStruct();
    this.__dummyIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new StatisticDataTypeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StatisticDataTypeArrayWrapper;
  }
  static get ArrayType() {
    return StatisticDataTypeArrayWrapper;
  }
  static get refObjectArrayType() {
    return StatisticDataTypeRefStructArray
  }
  static get refObjectType() {
    return StatisticDataTypeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this.__dummyIntialized) {
        throw new TypeError('Invalid argument: _dummy in StatisticDataType');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this.__dummyIntialized = true;
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    StatisticDataTypeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'statistics_msgs', subFolder: 'msg', interfaceName: 'StatisticDataType'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    if (!this.__dummyIntialized) {
      return undefined;
    }
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this.__dummyIntialized = true;
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StatisticDataTypeRefStruct(refObject.toObject());
    this.__dummyIntialized = true;
  }
  copy(other) {
    this._refObject = new StatisticDataTypeRefStruct(other._refObject.toObject());
    this.__dummyIntialized = true;
  }
  static get classType() {
    return StatisticDataTypeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"STATISTICS_DATA_TYPE_UNINITIALIZED","value":0},{"type":"uint8","name":"STATISTICS_DATA_TYPE_AVERAGE","value":1},{"type":"uint8","name":"STATISTICS_DATA_TYPE_MINIMUM","value":2},{"type":"uint8","name":"STATISTICS_DATA_TYPE_MAXIMUM","value":3},{"type":"uint8","name":"STATISTICS_DATA_TYPE_STDDEV","value":4},{"type":"uint8","name":"STATISTICS_DATA_TYPE_SAMPLE_COUNT","value":5}],"fields":[],"baseType":{"pkgName":"statistics_msgs","type":"StatisticDataType","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StatisticDataType"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StatisticDataTypeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StatisticDataTypeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StatisticDataTypeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatisticDataTypeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatisticDataTypeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StatisticDataTypeRefArray(size);
    this._refObject = new StatisticDataTypeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StatisticDataTypeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StatisticDataTypeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StatisticDataTypeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StatisticDataTypeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StatisticDataTypeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StatisticDataTypeArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(StatisticDataTypeWrapper, "STATISTICS_DATA_TYPE_UNINITIALIZED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatisticDataTypeWrapper, "STATISTICS_DATA_TYPE_AVERAGE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatisticDataTypeWrapper, "STATISTICS_DATA_TYPE_MINIMUM", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatisticDataTypeWrapper, "STATISTICS_DATA_TYPE_MAXIMUM", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatisticDataTypeWrapper, "STATISTICS_DATA_TYPE_STDDEV", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatisticDataTypeWrapper, "STATISTICS_DATA_TYPE_SAMPLE_COUNT", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = StatisticDataTypeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "STATISTICS_DATA_TYPE_UNINITIALIZED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "STATISTICS_DATA_TYPE_AVERAGE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "STATISTICS_DATA_TYPE_MINIMUM",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "STATISTICS_DATA_TYPE_MAXIMUM",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "STATISTICS_DATA_TYPE_STDDEV",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "STATISTICS_DATA_TYPE_SAMPLE_COUNT",
      "value": 5
    }
  ],
  "fields": [],
  "baseType": {
    "pkgName": "statistics_msgs",
    "type": "StatisticDataType",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StatisticDataType"
}
*/
