// rclnodejs/generated/action_msgs/action_msgs__srv__CancelGoal_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GoalInfoWrapper = require('../../generated/action_msgs/action_msgs__msg__GoalInfo.js');
const CancelGoal_RequestRefStruct = StructType({
  goal_info: GoalInfoWrapper.refObjectType,
});
const CancelGoal_RequestRefArray = ArrayType(CancelGoal_RequestRefStruct);
const CancelGoal_RequestRefStructArray = StructType({
  data: CancelGoal_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CancelGoal_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CancelGoal_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.goal_info =  new GoalInfoWrapper(other._wrapperFields.goal_info);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CancelGoal_RequestRefStruct();
    this._wrapperFields.goal_info = new GoalInfoWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CancelGoal_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CancelGoal_RequestArrayWrapper;
  }
  static get ArrayType() {
    return CancelGoal_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CancelGoal_RequestRefStructArray
  }
  static get refObjectType() {
    return CancelGoal_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.goal_info.freeze(own, checkConsistency);
    this._refObject.goal_info = this._wrapperFields.goal_info.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.goal_info.copyRefObject(refObject.goal_info);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GoalInfoWrapper.freeStruct(refObject.goal_info);
  }
  static destoryRawROS(msg) {
    CancelGoal_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'action_msgs', subFolder: 'srv', interfaceName: 'CancelGoal_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get goal_info() {
    return this._wrapperFields.goal_info;
  }
  set goal_info(value) {
    if (value instanceof GoalInfoWrapper) {
      this._wrapperFields.goal_info.copy(value);
    } else {
      this._wrapperFields.goal_info.copy(new GoalInfoWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new CancelGoal_RequestRefStruct(refObject.toObject());
    this._wrapperFields.goal_info.copyRefObject(this._refObject.goal_info);
  }
  copy(other) {
    this._refObject = new CancelGoal_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.goal_info.copy(other._wrapperFields.goal_info);
  }
  static get classType() {
    return CancelGoal_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"goal_info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"action_msgs","type":"GoalInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"action_msgs","type":"CancelGoal_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CancelGoal_Request"};
  }
  hasMember(name) {
    let memberNames = ["goal_info"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CancelGoal_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CancelGoal_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CancelGoal_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CancelGoal_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CancelGoal_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CancelGoal_RequestRefArray(size);
    this._refObject = new CancelGoal_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CancelGoal_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CancelGoal_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CancelGoal_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CancelGoal_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CancelGoal_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CancelGoal_RequestArrayWrapper;
  }
}
module.exports = CancelGoal_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "goal_info",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "action_msgs",
        "type": "GoalInfo",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "action_msgs",
    "type": "CancelGoal_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CancelGoal_Request"
}
*/
