// rclnodejs/generated/rmw_dds_common/rmw_dds_common__msg__NodeEntitiesInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GidWrapper = require('../../generated/rmw_dds_common/rmw_dds_common__msg__Gid.js');
const NodeEntitiesInfoRefStruct = StructType({
  node_namespace: primitiveTypes.string,
  node_name: primitiveTypes.string,
  reader_gid_seq: GidWrapper.refObjectArrayType,
  writer_gid_seq: GidWrapper.refObjectArrayType,
});
const NodeEntitiesInfoRefArray = ArrayType(NodeEntitiesInfoRefStruct);
const NodeEntitiesInfoRefStructArray = StructType({
  data: NodeEntitiesInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NodeEntitiesInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NodeEntitiesInfoRefStruct(other._refObject.toObject());
      this._node_namespaceIntialized = true;
      this._wrapperFields.node_namespace =  new StringWrapper(other._wrapperFields.node_namespace);
      this._node_nameIntialized = true;
      this._wrapperFields.node_name =  new StringWrapper(other._wrapperFields.node_name);
      this._wrapperFields.reader_gid_seq = GidWrapper.createArray();
      this._wrapperFields.reader_gid_seq.copy(other._wrapperFields.reader_gid_seq);
      this._wrapperFields.writer_gid_seq = GidWrapper.createArray();
      this._wrapperFields.writer_gid_seq.copy(other._wrapperFields.writer_gid_seq);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NodeEntitiesInfoRefStruct();
    this._node_namespaceIntialized = false;
    this._wrapperFields.node_namespace = new StringWrapper();
    this._node_nameIntialized = false;
    this._wrapperFields.node_name = new StringWrapper();
    this._wrapperFields.reader_gid_seq = GidWrapper.createArray();
    this._wrapperFields.writer_gid_seq = GidWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new NodeEntitiesInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NodeEntitiesInfoArrayWrapper;
  }
  static get ArrayType() {
    return NodeEntitiesInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return NodeEntitiesInfoRefStructArray
  }
  static get refObjectType() {
    return NodeEntitiesInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._node_namespaceIntialized) {
        throw new TypeError('Invalid argument: node_namespace in NodeEntitiesInfo');
      }
      if (!this._node_nameIntialized) {
        throw new TypeError('Invalid argument: node_name in NodeEntitiesInfo');
      }
    }
    if (own) {
      this._wrapperFields.node_namespace.freeze(own, checkConsistency);
    }
    this._refObject.node_namespace = this._wrapperFields.node_namespace.refObject;
    if (own) {
      this._wrapperFields.node_name.freeze(own, checkConsistency);
    }
    this._refObject.node_name = this._wrapperFields.node_name.refObject;
    this._wrapperFields.reader_gid_seq.freeze(own, checkConsistency);
    this._refObject.reader_gid_seq = this._wrapperFields.reader_gid_seq.refObject;
    this._wrapperFields.writer_gid_seq.freeze(own, checkConsistency);
    this._refObject.writer_gid_seq = this._wrapperFields.writer_gid_seq.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._node_namespaceIntialized = true;
    this._wrapperFields.node_namespace.data = refObject.node_namespace.data;
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.data = refObject.node_name.data;
    this._wrapperFields.reader_gid_seq.copyRefObject(refObject.reader_gid_seq);
    this._wrapperFields.writer_gid_seq.copyRefObject(refObject.writer_gid_seq);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.node_namespace);
    StringWrapper.freeStruct(refObject.node_name);
    if (refObject.reader_gid_seq.size != 0) {
      GidWrapper.ArrayType.freeArray(refObject.reader_gid_seq);
      if (GidWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.reader_gid_seq, GidWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.writer_gid_seq.size != 0) {
      GidWrapper.ArrayType.freeArray(refObject.writer_gid_seq);
      if (GidWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.writer_gid_seq, GidWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    NodeEntitiesInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rmw_dds_common', subFolder: 'msg', interfaceName: 'NodeEntitiesInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get node_namespace() {
    if (!this._node_namespaceIntialized) {
      return undefined;
    }
    return this._wrapperFields.node_namespace.data;
  }
  set node_namespace(value) {
    this._node_namespaceIntialized = true;
    this._wrapperFields.node_namespace.data = value;
  }
  get node_name() {
    if (!this._node_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.node_name.data;
  }
  set node_name(value) {
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.data = value;
  }
  get reader_gid_seq() {
    return this._wrapperFields.reader_gid_seq;
  }
  set reader_gid_seq(value) {
    this._wrapperFields.reader_gid_seq.fill(value);
  }
  get writer_gid_seq() {
    return this._wrapperFields.writer_gid_seq;
  }
  set writer_gid_seq(value) {
    this._wrapperFields.writer_gid_seq.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new NodeEntitiesInfoRefStruct(refObject.toObject());
    this._node_namespaceIntialized = true;
    this._wrapperFields.node_namespace.copyRefObject(this._refObject.node_namespace);
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.copyRefObject(this._refObject.node_name);
    this._wrapperFields.reader_gid_seq.copyRefObject(this._refObject.reader_gid_seq);
    this._wrapperFields.writer_gid_seq.copyRefObject(this._refObject.writer_gid_seq);
  }
  copy(other) {
    this._refObject = new NodeEntitiesInfoRefStruct(other._refObject.toObject());
    this._node_namespaceIntialized = true;
    this._wrapperFields.node_namespace.copy(other._wrapperFields.node_namespace);
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.copy(other._wrapperFields.node_name);
    this._wrapperFields.reader_gid_seq.copy(other._wrapperFields.reader_gid_seq);
    this._wrapperFields.writer_gid_seq.copy(other._wrapperFields.writer_gid_seq);
  }
  static get classType() {
    return NodeEntitiesInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"node_namespace","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":256,"isPrimitiveType":true},"default_value":null},{"name":"node_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":256,"isPrimitiveType":true},"default_value":null},{"name":"reader_gid_seq","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rmw_dds_common","type":"Gid","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"writer_gid_seq","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rmw_dds_common","type":"Gid","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rmw_dds_common","type":"NodeEntitiesInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NodeEntitiesInfo"};
  }
  hasMember(name) {
    let memberNames = ["node_namespace","node_name","reader_gid_seq","writer_gid_seq"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NodeEntitiesInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NodeEntitiesInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NodeEntitiesInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NodeEntitiesInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NodeEntitiesInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NodeEntitiesInfoRefArray(size);
    this._refObject = new NodeEntitiesInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NodeEntitiesInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NodeEntitiesInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NodeEntitiesInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NodeEntitiesInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NodeEntitiesInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NodeEntitiesInfoArrayWrapper;
  }
}
module.exports = NodeEntitiesInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "node_namespace",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 256,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "node_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 256,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reader_gid_seq",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rmw_dds_common",
        "type": "Gid",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "writer_gid_seq",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rmw_dds_common",
        "type": "Gid",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rmw_dds_common",
    "type": "NodeEntitiesInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NodeEntitiesInfo"
}
*/
