// rclnodejs/generated/rmw_dds_common/rmw_dds_common__msg__ParticipantEntitiesInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GidWrapper = require('../../generated/rmw_dds_common/rmw_dds_common__msg__Gid.js');
const NodeEntitiesInfoWrapper = require('../../generated/rmw_dds_common/rmw_dds_common__msg__NodeEntitiesInfo.js');
const ParticipantEntitiesInfoRefStruct = StructType({
  gid: GidWrapper.refObjectType,
  node_entities_info_seq: NodeEntitiesInfoWrapper.refObjectArrayType,
});
const ParticipantEntitiesInfoRefArray = ArrayType(ParticipantEntitiesInfoRefStruct);
const ParticipantEntitiesInfoRefStructArray = StructType({
  data: ParticipantEntitiesInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ParticipantEntitiesInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParticipantEntitiesInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.gid =  new GidWrapper(other._wrapperFields.gid);
      this._wrapperFields.node_entities_info_seq = NodeEntitiesInfoWrapper.createArray();
      this._wrapperFields.node_entities_info_seq.copy(other._wrapperFields.node_entities_info_seq);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ParticipantEntitiesInfoRefStruct();
    this._wrapperFields.gid = new GidWrapper();
    this._wrapperFields.node_entities_info_seq = NodeEntitiesInfoWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ParticipantEntitiesInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParticipantEntitiesInfoArrayWrapper;
  }
  static get ArrayType() {
    return ParticipantEntitiesInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParticipantEntitiesInfoRefStructArray
  }
  static get refObjectType() {
    return ParticipantEntitiesInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.gid.freeze(own, checkConsistency);
    this._refObject.gid = this._wrapperFields.gid.refObject;
    this._wrapperFields.node_entities_info_seq.freeze(own, checkConsistency);
    this._refObject.node_entities_info_seq = this._wrapperFields.node_entities_info_seq.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.gid.copyRefObject(refObject.gid);
    this._wrapperFields.node_entities_info_seq.copyRefObject(refObject.node_entities_info_seq);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GidWrapper.freeStruct(refObject.gid);
    if (refObject.node_entities_info_seq.size != 0) {
      NodeEntitiesInfoWrapper.ArrayType.freeArray(refObject.node_entities_info_seq);
      if (NodeEntitiesInfoWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.node_entities_info_seq, NodeEntitiesInfoWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ParticipantEntitiesInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rmw_dds_common', subFolder: 'msg', interfaceName: 'ParticipantEntitiesInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get gid() {
    return this._wrapperFields.gid;
  }
  set gid(value) {
    if (value instanceof GidWrapper) {
      this._wrapperFields.gid.copy(value);
    } else {
      this._wrapperFields.gid.copy(new GidWrapper(value));
    }
  }
  get node_entities_info_seq() {
    return this._wrapperFields.node_entities_info_seq;
  }
  set node_entities_info_seq(value) {
    this._wrapperFields.node_entities_info_seq.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ParticipantEntitiesInfoRefStruct(refObject.toObject());
    this._wrapperFields.gid.copyRefObject(this._refObject.gid);
    this._wrapperFields.node_entities_info_seq.copyRefObject(this._refObject.node_entities_info_seq);
  }
  copy(other) {
    this._refObject = new ParticipantEntitiesInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.gid.copy(other._wrapperFields.gid);
    this._wrapperFields.node_entities_info_seq.copy(other._wrapperFields.node_entities_info_seq);
  }
  static get classType() {
    return ParticipantEntitiesInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"gid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"rmw_dds_common","type":"Gid","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"node_entities_info_seq","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rmw_dds_common","type":"NodeEntitiesInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rmw_dds_common","type":"ParticipantEntitiesInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParticipantEntitiesInfo"};
  }
  hasMember(name) {
    let memberNames = ["gid","node_entities_info_seq"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ParticipantEntitiesInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParticipantEntitiesInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParticipantEntitiesInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParticipantEntitiesInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParticipantEntitiesInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ParticipantEntitiesInfoRefArray(size);
    this._refObject = new ParticipantEntitiesInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParticipantEntitiesInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParticipantEntitiesInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParticipantEntitiesInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParticipantEntitiesInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParticipantEntitiesInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParticipantEntitiesInfoArrayWrapper;
  }
}
module.exports = ParticipantEntitiesInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "gid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "rmw_dds_common",
        "type": "Gid",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "node_entities_info_seq",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rmw_dds_common",
        "type": "NodeEntitiesInfo",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rmw_dds_common",
    "type": "ParticipantEntitiesInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ParticipantEntitiesInfo"
}
*/
