// rclnodejs/generated/std_msgs/std_msgs__msg__MultiArrayDimension.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MultiArrayDimensionRefStruct = StructType({
  label: primitiveTypes.string,
  size: primitiveTypes.uint32,
  stride: primitiveTypes.uint32,
});
const MultiArrayDimensionRefArray = ArrayType(MultiArrayDimensionRefStruct);
const MultiArrayDimensionRefStructArray = StructType({
  data: MultiArrayDimensionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MultiArrayDimensionWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MultiArrayDimensionRefStruct(other._refObject.toObject());
      this._labelIntialized = true;
      this._wrapperFields.label =  new StringWrapper(other._wrapperFields.label);
      this._sizeIntialized = true;
      this._strideIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MultiArrayDimensionRefStruct();
    this._labelIntialized = false;
    this._wrapperFields.label = new StringWrapper();
    this._sizeIntialized = false;
    this._strideIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MultiArrayDimensionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MultiArrayDimensionArrayWrapper;
  }
  static get ArrayType() {
    return MultiArrayDimensionArrayWrapper;
  }
  static get refObjectArrayType() {
    return MultiArrayDimensionRefStructArray
  }
  static get refObjectType() {
    return MultiArrayDimensionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._labelIntialized) {
        throw new TypeError('Invalid argument: label in MultiArrayDimension');
      }
      if (!this._sizeIntialized) {
        throw new TypeError('Invalid argument: size in MultiArrayDimension');
      }
      if (!this._strideIntialized) {
        throw new TypeError('Invalid argument: stride in MultiArrayDimension');
      }
    }
    if (own) {
      this._wrapperFields.label.freeze(own, checkConsistency);
    }
    this._refObject.label = this._wrapperFields.label.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._labelIntialized = true;
    this._wrapperFields.label.data = refObject.label.data;
    this._sizeIntialized = true;
    this._refObject.size = refObject.size;
    this._strideIntialized = true;
    this._refObject.stride = refObject.stride;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.label);
  }
  static destoryRawROS(msg) {
    MultiArrayDimensionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'std_msgs', subFolder: 'msg', interfaceName: 'MultiArrayDimension'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get label() {
    if (!this._labelIntialized) {
      return undefined;
    }
    return this._wrapperFields.label.data;
  }
  set label(value) {
    this._labelIntialized = true;
    this._wrapperFields.label.data = value;
  }
  get size() {
    if (!this._sizeIntialized) {
      return undefined;
    }
    return this._refObject.size;
  }
  set size(value) {
    this._sizeIntialized = true;
    this._refObject.size = value;
  }
  get stride() {
    if (!this._strideIntialized) {
      return undefined;
    }
    return this._refObject.stride;
  }
  set stride(value) {
    this._strideIntialized = true;
    this._refObject.stride = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MultiArrayDimensionRefStruct(refObject.toObject());
    this._labelIntialized = true;
    this._wrapperFields.label.copyRefObject(this._refObject.label);
    this._sizeIntialized = true;
    this._strideIntialized = true;
  }
  copy(other) {
    this._refObject = new MultiArrayDimensionRefStruct(other._refObject.toObject());
    this._labelIntialized = true;
    this._wrapperFields.label.copy(other._wrapperFields.label);
    this._sizeIntialized = true;
    this._strideIntialized = true;
  }
  static get classType() {
    return MultiArrayDimensionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"label","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stride","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"std_msgs","type":"MultiArrayDimension","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MultiArrayDimension"};
  }
  hasMember(name) {
    let memberNames = ["label","size","stride"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MultiArrayDimensionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MultiArrayDimensionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MultiArrayDimensionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiArrayDimensionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiArrayDimensionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MultiArrayDimensionRefArray(size);
    this._refObject = new MultiArrayDimensionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MultiArrayDimensionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MultiArrayDimensionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MultiArrayDimensionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MultiArrayDimensionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MultiArrayDimensionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MultiArrayDimensionArrayWrapper;
  }
}
module.exports = MultiArrayDimensionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "label",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stride",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "std_msgs",
    "type": "MultiArrayDimension",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MultiArrayDimension"
}
*/
