// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__NavSatFix.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const NavSatStatusWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__NavSatStatus.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const NavSatFixRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  status: NavSatStatusWrapper.refObjectType,
  latitude: primitiveTypes.float64,
  longitude: primitiveTypes.float64,
  altitude: primitiveTypes.float64,
  position_covariance: ArrayType(primitiveTypes.float64, 9),
  position_covariance_type: primitiveTypes.uint8,
});
const NavSatFixRefArray = ArrayType(NavSatFixRefStruct);
const NavSatFixRefStructArray = StructType({
  data: NavSatFixRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NavSatFixWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NavSatFixRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.status =  new NavSatStatusWrapper(other._wrapperFields.status);
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
      this._altitudeIntialized = true;
      this._wrapperFields.position_covariance = Float64Wrapper.createArray();
      this._wrapperFields.position_covariance.copy(other._wrapperFields.position_covariance);
      this._position_covariance_typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NavSatFixRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.status = new NavSatStatusWrapper();
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
    this._altitudeIntialized = false;
    this._wrapperFields.position_covariance = Float64Wrapper.createArray();
    this._position_covariance_typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new NavSatFixWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NavSatFixArrayWrapper;
  }
  static get ArrayType() {
    return NavSatFixArrayWrapper;
  }
  static get refObjectArrayType() {
    return NavSatFixRefStructArray
  }
  static get refObjectType() {
    return NavSatFixRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in NavSatFix');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in NavSatFix');
      }
      if (!this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in NavSatFix');
      }
      if (!this._position_covariance_typeIntialized) {
        throw new TypeError('Invalid argument: position_covariance_type in NavSatFix');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
    for (let i = 0; i < this._wrapperFields.position_covariance.data.length; i++) {
      this._refObject.position_covariance[i] = this._wrapperFields.position_covariance.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.status.copyRefObject(refObject.status);
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
    this._refObject.position_covariance = refObject.position_covariance;
    this._wrapperFields.position_covariance.fill(refObject.position_covariance.toArray());
    this._position_covariance_typeIntialized = true;
    this._refObject.position_covariance_type = refObject.position_covariance_type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    NavSatStatusWrapper.freeStruct(refObject.status);
  }
  static destoryRawROS(msg) {
    NavSatFixWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'NavSatFix'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    if (value instanceof NavSatStatusWrapper) {
      this._wrapperFields.status.copy(value);
    } else {
      this._wrapperFields.status.copy(new NavSatStatusWrapper(value));
    }
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  get position_covariance() {
    return this._wrapperFields['position_covariance'].data;
  }
  set position_covariance(value) {
    if (value.length !== 9) {
      throw new RangeError('The length of the array must be 9.');
    }
    this._wrapperFields['position_covariance'].fill(value);
  }
  get position_covariance_type() {
    if (!this._position_covariance_typeIntialized) {
      return undefined;
    }
    return this._refObject.position_covariance_type;
  }
  set position_covariance_type(value) {
    this._position_covariance_typeIntialized = true;
    this._refObject.position_covariance_type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NavSatFixRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.status.copyRefObject(this._refObject.status);
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._wrapperFields.position_covariance.fill(refObject.position_covariance.toArray());
    this._position_covariance_typeIntialized = true;
  }
  copy(other) {
    this._refObject = new NavSatFixRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.status.copy(other._wrapperFields.status);
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._wrapperFields.position_covariance.copy(other._wrapperFields.position_covariance);
    this._position_covariance_typeIntialized = true;
  }
  static get classType() {
    return NavSatFixWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"COVARIANCE_TYPE_UNKNOWN","value":0},{"type":"uint8","name":"COVARIANCE_TYPE_APPROXIMATED","value":1},{"type":"uint8","name":"COVARIANCE_TYPE_DIAGONAL_KNOWN","value":2},{"type":"uint8","name":"COVARIANCE_TYPE_KNOWN","value":3}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"NavSatStatus","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_covariance","type":{"isArray":true,"arraySize":9,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_covariance_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"NavSatFix","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NavSatFix"};
  }
  hasMember(name) {
    let memberNames = ["header","status","latitude","longitude","altitude","position_covariance","position_covariance_type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NavSatFixArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NavSatFixWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NavSatFixWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavSatFixArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavSatFixArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NavSatFixRefArray(size);
    this._refObject = new NavSatFixRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NavSatFixWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NavSatFixArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NavSatFixArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NavSatFixWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NavSatFixWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NavSatFixArrayWrapper;
  }
}
// Define constants (4 in total)
Object.defineProperty(NavSatFixWrapper, "COVARIANCE_TYPE_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatFixWrapper, "COVARIANCE_TYPE_APPROXIMATED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatFixWrapper, "COVARIANCE_TYPE_DIAGONAL_KNOWN", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(NavSatFixWrapper, "COVARIANCE_TYPE_KNOWN", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = NavSatFixWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "COVARIANCE_TYPE_UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "COVARIANCE_TYPE_APPROXIMATED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "COVARIANCE_TYPE_DIAGONAL_KNOWN",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "COVARIANCE_TYPE_KNOWN",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "NavSatStatus",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position_covariance",
      "type": {
        "isArray": true,
        "arraySize": 9,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position_covariance_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "NavSatFix",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NavSatFix"
}
*/
