// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__JoyFeedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const JoyFeedbackRefStruct = StructType({
  type: primitiveTypes.uint8,
  id: primitiveTypes.uint8,
  intensity: primitiveTypes.float32,
});
const JoyFeedbackRefArray = ArrayType(JoyFeedbackRefStruct);
const JoyFeedbackRefStructArray = StructType({
  data: JoyFeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class JoyFeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new JoyFeedbackRefStruct(other._refObject.toObject());
      this._typeIntialized = true;
      this._idIntialized = true;
      this._intensityIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new JoyFeedbackRefStruct();
    this._typeIntialized = false;
    this._idIntialized = false;
    this._intensityIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new JoyFeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JoyFeedbackArrayWrapper;
  }
  static get ArrayType() {
    return JoyFeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return JoyFeedbackRefStructArray
  }
  static get refObjectType() {
    return JoyFeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in JoyFeedback');
      }
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in JoyFeedback');
      }
      if (!this._intensityIntialized) {
        throw new TypeError('Invalid argument: intensity in JoyFeedback');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._intensityIntialized = true;
    this._refObject.intensity = refObject.intensity;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    JoyFeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'JoyFeedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get intensity() {
    if (!this._intensityIntialized) {
      return undefined;
    }
    return this._refObject.intensity;
  }
  set intensity(value) {
    this._intensityIntialized = true;
    this._refObject.intensity = value;
  }
  copyRefObject(refObject) {
    this._refObject = new JoyFeedbackRefStruct(refObject.toObject());
    this._typeIntialized = true;
    this._idIntialized = true;
    this._intensityIntialized = true;
  }
  copy(other) {
    this._refObject = new JoyFeedbackRefStruct(other._refObject.toObject());
    this._typeIntialized = true;
    this._idIntialized = true;
    this._intensityIntialized = true;
  }
  static get classType() {
    return JoyFeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"TYPE_LED","value":0},{"type":"uint8","name":"TYPE_RUMBLE","value":1},{"type":"uint8","name":"TYPE_BUZZER","value":2}],"fields":[{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"intensity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"JoyFeedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JoyFeedback"};
  }
  hasMember(name) {
    let memberNames = ["type","id","intensity"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class JoyFeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JoyFeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JoyFeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JoyFeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JoyFeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new JoyFeedbackRefArray(size);
    this._refObject = new JoyFeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JoyFeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JoyFeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JoyFeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JoyFeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JoyFeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JoyFeedbackArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(JoyFeedbackWrapper, "TYPE_LED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(JoyFeedbackWrapper, "TYPE_RUMBLE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(JoyFeedbackWrapper, "TYPE_BUZZER", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = JoyFeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "TYPE_LED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "TYPE_RUMBLE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "TYPE_BUZZER",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "intensity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "JoyFeedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "JoyFeedback"
}
*/
