// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__PointCloud.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Point32Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point32.js');
const ChannelFloat32Wrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__ChannelFloat32.js');
const PointCloudRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  points: Point32Wrapper.refObjectArrayType,
  channels: ChannelFloat32Wrapper.refObjectArrayType,
});
const PointCloudRefArray = ArrayType(PointCloudRefStruct);
const PointCloudRefStructArray = StructType({
  data: PointCloudRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PointCloudWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PointCloudRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.points = Point32Wrapper.createArray();
      this._wrapperFields.points.copy(other._wrapperFields.points);
      this._wrapperFields.channels = ChannelFloat32Wrapper.createArray();
      this._wrapperFields.channels.copy(other._wrapperFields.channels);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PointCloudRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.points = Point32Wrapper.createArray();
    this._wrapperFields.channels = ChannelFloat32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new PointCloudWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PointCloudArrayWrapper;
  }
  static get ArrayType() {
    return PointCloudArrayWrapper;
  }
  static get refObjectArrayType() {
    return PointCloudRefStructArray
  }
  static get refObjectType() {
    return PointCloudRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.points.freeze(own, checkConsistency);
    this._refObject.points = this._wrapperFields.points.refObject;
    this._wrapperFields.channels.freeze(own, checkConsistency);
    this._refObject.channels = this._wrapperFields.channels.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.points.copyRefObject(refObject.points);
    this._wrapperFields.channels.copyRefObject(refObject.channels);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.points.size != 0) {
      Point32Wrapper.ArrayType.freeArray(refObject.points);
      if (Point32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.points, Point32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.channels.size != 0) {
      ChannelFloat32Wrapper.ArrayType.freeArray(refObject.channels);
      if (ChannelFloat32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.channels, ChannelFloat32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    PointCloudWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'PointCloud'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get points() {
    return this._wrapperFields.points;
  }
  set points(value) {
    this._wrapperFields.points.fill(value);
  }
  get channels() {
    return this._wrapperFields.channels;
  }
  set channels(value) {
    this._wrapperFields.channels.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new PointCloudRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.points.copyRefObject(this._refObject.points);
    this._wrapperFields.channels.copyRefObject(this._refObject.channels);
  }
  copy(other) {
    this._refObject = new PointCloudRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.points.copy(other._wrapperFields.points);
    this._wrapperFields.channels.copy(other._wrapperFields.channels);
  }
  static get classType() {
    return PointCloudWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"points","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Point32","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"channels","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"sensor_msgs","type":"ChannelFloat32","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"PointCloud","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PointCloud"};
  }
  hasMember(name) {
    let memberNames = ["header","points","channels"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PointCloudArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PointCloudWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PointCloudWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointCloudArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointCloudArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PointCloudRefArray(size);
    this._refObject = new PointCloudRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PointCloudWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PointCloudArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PointCloudArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PointCloudWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PointCloudWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PointCloudArrayWrapper;
  }
}
module.exports = PointCloudWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "points",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Point32",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "channels",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "sensor_msgs",
        "type": "ChannelFloat32",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "PointCloud",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PointCloud"
}
*/
