// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__JoyFeedbackArray.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const JoyFeedbackWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__JoyFeedback.js');
const JoyFeedbackArrayRefStruct = StructType({
  array: JoyFeedbackWrapper.refObjectArrayType,
});
const JoyFeedbackArrayRefArray = ArrayType(JoyFeedbackArrayRefStruct);
const JoyFeedbackArrayRefStructArray = StructType({
  data: JoyFeedbackArrayRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class JoyFeedbackArrayWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new JoyFeedbackArrayRefStruct(other._refObject.toObject());
      this._wrapperFields.array = JoyFeedbackWrapper.createArray();
      this._wrapperFields.array.copy(other._wrapperFields.array);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new JoyFeedbackArrayRefStruct();
    this._wrapperFields.array = JoyFeedbackWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new JoyFeedbackArrayWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JoyFeedbackArrayArrayWrapper;
  }
  static get ArrayType() {
    return JoyFeedbackArrayArrayWrapper;
  }
  static get refObjectArrayType() {
    return JoyFeedbackArrayRefStructArray
  }
  static get refObjectType() {
    return JoyFeedbackArrayRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.array.freeze(own, checkConsistency);
    this._refObject.array = this._wrapperFields.array.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.array.copyRefObject(refObject.array);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.array.size != 0) {
      JoyFeedbackWrapper.ArrayType.freeArray(refObject.array);
      if (JoyFeedbackWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.array, JoyFeedbackWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    JoyFeedbackArrayWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'JoyFeedbackArray'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get array() {
    return this._wrapperFields.array;
  }
  set array(value) {
    this._wrapperFields.array.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new JoyFeedbackArrayRefStruct(refObject.toObject());
    this._wrapperFields.array.copyRefObject(this._refObject.array);
  }
  copy(other) {
    this._refObject = new JoyFeedbackArrayRefStruct(other._refObject.toObject());
    this._wrapperFields.array.copy(other._wrapperFields.array);
  }
  static get classType() {
    return JoyFeedbackArrayWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"array","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"sensor_msgs","type":"JoyFeedback","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"JoyFeedbackArray","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JoyFeedbackArray"};
  }
  hasMember(name) {
    let memberNames = ["array"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class JoyFeedbackArrayArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JoyFeedbackArrayWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JoyFeedbackArrayWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JoyFeedbackArrayArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JoyFeedbackArrayArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new JoyFeedbackArrayRefArray(size);
    this._refObject = new JoyFeedbackArrayRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JoyFeedbackArrayWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JoyFeedbackArrayArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JoyFeedbackArrayArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JoyFeedbackArrayWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JoyFeedbackArrayWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JoyFeedbackArrayArrayWrapper;
  }
}
module.exports = JoyFeedbackArrayWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "array",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "sensor_msgs",
        "type": "JoyFeedback",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "JoyFeedbackArray",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "JoyFeedbackArray"
}
*/
