// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__MultiDOFJointState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TransformWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Transform.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const WrenchWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Wrench.js');
const MultiDOFJointStateRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  joint_names: StringWrapper.refObjectArrayType,
  transforms: TransformWrapper.refObjectArrayType,
  twist: TwistWrapper.refObjectArrayType,
  wrench: WrenchWrapper.refObjectArrayType,
});
const MultiDOFJointStateRefArray = ArrayType(MultiDOFJointStateRefStruct);
const MultiDOFJointStateRefStructArray = StructType({
  data: MultiDOFJointStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MultiDOFJointStateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._joint_namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MultiDOFJointStateRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.joint_names = StringWrapper.createArray();
      this._wrapperFields.joint_names.copy(other._wrapperFields.joint_names);
      this.joint_names = other.joint_names;
      this._wrapperFields.transforms = TransformWrapper.createArray();
      this._wrapperFields.transforms.copy(other._wrapperFields.transforms);
      this._wrapperFields.twist = TwistWrapper.createArray();
      this._wrapperFields.twist.copy(other._wrapperFields.twist);
      this._wrapperFields.wrench = WrenchWrapper.createArray();
      this._wrapperFields.wrench.copy(other._wrapperFields.wrench);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MultiDOFJointStateRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.joint_names = StringWrapper.createArray();
    this._wrapperFields.transforms = TransformWrapper.createArray();
    this._wrapperFields.twist = TwistWrapper.createArray();
    this._wrapperFields.wrench = WrenchWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MultiDOFJointStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MultiDOFJointStateArrayWrapper;
  }
  static get ArrayType() {
    return MultiDOFJointStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return MultiDOFJointStateRefStructArray
  }
  static get refObjectType() {
    return MultiDOFJointStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (!own) {
      this._wrapperFields.joint_names.fill(this._joint_namesArray);
      this._wrapperFields.joint_names.freeze(own, checkConsistency);
      this._refObject.joint_names = this._wrapperFields.joint_names.refObject;
    } else {
      this._wrapperFields.joint_names.fill([]);
      this._wrapperFields.joint_names.freeze(own, checkConsistency);
      this._refObject.joint_names = this._wrapperFields.joint_names.refObject;
    }
    this._wrapperFields.transforms.freeze(own, checkConsistency);
    this._refObject.transforms = this._wrapperFields.transforms.refObject;
    this._wrapperFields.twist.freeze(own, checkConsistency);
    this._refObject.twist = this._wrapperFields.twist.refObject;
    this._wrapperFields.wrench.freeze(own, checkConsistency);
    this._refObject.wrench = this._wrapperFields.wrench.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    refObject.joint_names.data.length = refObject.joint_names.size;
    for (let index = 0; index < refObject.joint_names.size; index++) {
      this._joint_namesArray[index] = refObject.joint_names.data[index].data;
    }
    this._wrapperFields.transforms.copyRefObject(refObject.transforms);
    this._wrapperFields.twist.copyRefObject(refObject.twist);
    this._wrapperFields.wrench.copyRefObject(refObject.wrench);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.joint_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.joint_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.joint_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.transforms.size != 0) {
      TransformWrapper.ArrayType.freeArray(refObject.transforms);
      if (TransformWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.transforms, TransformWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.twist.size != 0) {
      TwistWrapper.ArrayType.freeArray(refObject.twist);
      if (TwistWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.twist, TwistWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.wrench.size != 0) {
      WrenchWrapper.ArrayType.freeArray(refObject.wrench);
      if (WrenchWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.wrench, WrenchWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MultiDOFJointStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'MultiDOFJointState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get joint_names() {
    return this._joint_namesArray;
  }
  set joint_names(value) {
    this._joint_namesArray = value;
  }
  get transforms() {
    return this._wrapperFields.transforms;
  }
  set transforms(value) {
    this._wrapperFields.transforms.fill(value);
  }
  get twist() {
    return this._wrapperFields.twist;
  }
  set twist(value) {
    this._wrapperFields.twist.fill(value);
  }
  get wrench() {
    return this._wrapperFields.wrench;
  }
  set wrench(value) {
    this._wrapperFields.wrench.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MultiDOFJointStateRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    refObject.joint_names.data.length = refObject.joint_names.size;
    for (let index = 0; index < refObject.joint_names.size; index++) {
      this._joint_namesArray[index] = refObject.joint_names.data[index].data;
    }
    this._wrapperFields.transforms.copyRefObject(this._refObject.transforms);
    this._wrapperFields.twist.copyRefObject(this._refObject.twist);
    this._wrapperFields.wrench.copyRefObject(this._refObject.wrench);
  }
  copy(other) {
    this._refObject = new MultiDOFJointStateRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._joint_namesArray = other._joint_namesArray.slice();
    this._wrapperFields.transforms.copy(other._wrapperFields.transforms);
    this._wrapperFields.twist.copy(other._wrapperFields.twist);
    this._wrapperFields.wrench.copy(other._wrapperFields.wrench);
  }
  static get classType() {
    return MultiDOFJointStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"joint_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"transforms","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Transform","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"twist","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"wrench","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"MultiDOFJointState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MultiDOFJointState"};
  }
  hasMember(name) {
    let memberNames = ["header","joint_names","transforms","twist","wrench"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MultiDOFJointStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MultiDOFJointStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MultiDOFJointStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiDOFJointStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiDOFJointStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MultiDOFJointStateRefArray(size);
    this._refObject = new MultiDOFJointStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MultiDOFJointStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MultiDOFJointStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MultiDOFJointStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MultiDOFJointStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MultiDOFJointStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MultiDOFJointStateArrayWrapper;
  }
}
module.exports = MultiDOFJointStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "joint_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "transforms",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Transform",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "twist",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Twist",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "wrench",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Wrench",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "MultiDOFJointState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MultiDOFJointState"
}
*/
