// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__MagneticField.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const MagneticFieldRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  magnetic_field: Vector3Wrapper.refObjectType,
  magnetic_field_covariance: ArrayType(primitiveTypes.float64, 9),
});
const MagneticFieldRefArray = ArrayType(MagneticFieldRefStruct);
const MagneticFieldRefStructArray = StructType({
  data: MagneticFieldRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MagneticFieldWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MagneticFieldRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.magnetic_field =  new Vector3Wrapper(other._wrapperFields.magnetic_field);
      this._wrapperFields.magnetic_field_covariance = Float64Wrapper.createArray();
      this._wrapperFields.magnetic_field_covariance.copy(other._wrapperFields.magnetic_field_covariance);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MagneticFieldRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.magnetic_field = new Vector3Wrapper();
    this._wrapperFields.magnetic_field_covariance = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MagneticFieldWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MagneticFieldArrayWrapper;
  }
  static get ArrayType() {
    return MagneticFieldArrayWrapper;
  }
  static get refObjectArrayType() {
    return MagneticFieldRefStructArray
  }
  static get refObjectType() {
    return MagneticFieldRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.magnetic_field.freeze(own, checkConsistency);
    this._refObject.magnetic_field = this._wrapperFields.magnetic_field.refObject;
    for (let i = 0; i < this._wrapperFields.magnetic_field_covariance.data.length; i++) {
      this._refObject.magnetic_field_covariance[i] = this._wrapperFields.magnetic_field_covariance.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.magnetic_field.copyRefObject(refObject.magnetic_field);
    this._refObject.magnetic_field_covariance = refObject.magnetic_field_covariance;
    this._wrapperFields.magnetic_field_covariance.fill(refObject.magnetic_field_covariance.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    Vector3Wrapper.freeStruct(refObject.magnetic_field);
  }
  static destoryRawROS(msg) {
    MagneticFieldWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'MagneticField'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get magnetic_field() {
    return this._wrapperFields.magnetic_field;
  }
  set magnetic_field(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.magnetic_field.copy(value);
    } else {
      this._wrapperFields.magnetic_field.copy(new Vector3Wrapper(value));
    }
  }
  get magnetic_field_covariance() {
    return this._wrapperFields['magnetic_field_covariance'].data;
  }
  set magnetic_field_covariance(value) {
    if (value.length !== 9) {
      throw new RangeError('The length of the array must be 9.');
    }
    this._wrapperFields['magnetic_field_covariance'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MagneticFieldRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.magnetic_field.copyRefObject(this._refObject.magnetic_field);
    this._wrapperFields.magnetic_field_covariance.fill(refObject.magnetic_field_covariance.toArray());
  }
  copy(other) {
    this._refObject = new MagneticFieldRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.magnetic_field.copy(other._wrapperFields.magnetic_field);
    this._wrapperFields.magnetic_field_covariance.copy(other._wrapperFields.magnetic_field_covariance);
  }
  static get classType() {
    return MagneticFieldWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"magnetic_field","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"magnetic_field_covariance","type":{"isArray":true,"arraySize":9,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"MagneticField","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MagneticField"};
  }
  hasMember(name) {
    let memberNames = ["header","magnetic_field","magnetic_field_covariance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MagneticFieldArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MagneticFieldWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MagneticFieldWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MagneticFieldArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MagneticFieldArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MagneticFieldRefArray(size);
    this._refObject = new MagneticFieldRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MagneticFieldWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MagneticFieldArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MagneticFieldArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MagneticFieldWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MagneticFieldWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MagneticFieldArrayWrapper;
  }
}
module.exports = MagneticFieldWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "magnetic_field",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "magnetic_field_covariance",
      "type": {
        "isArray": true,
        "arraySize": 9,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "MagneticField",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MagneticField"
}
*/
