// rclnodejs/generated/sensor_msgs/sensor_msgs__msg__LaserScan.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const LaserScanRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  angle_min: primitiveTypes.float32,
  angle_max: primitiveTypes.float32,
  angle_increment: primitiveTypes.float32,
  time_increment: primitiveTypes.float32,
  scan_time: primitiveTypes.float32,
  range_min: primitiveTypes.float32,
  range_max: primitiveTypes.float32,
  ranges: Float32Wrapper.refObjectArrayType,
  intensities: Float32Wrapper.refObjectArrayType,
});
const LaserScanRefArray = ArrayType(LaserScanRefStruct);
const LaserScanRefStructArray = StructType({
  data: LaserScanRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LaserScanWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LaserScanRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._angle_minIntialized = true;
      this._angle_maxIntialized = true;
      this._angle_incrementIntialized = true;
      this._time_incrementIntialized = true;
      this._scan_timeIntialized = true;
      this._range_minIntialized = true;
      this._range_maxIntialized = true;
      this._wrapperFields.ranges = Float32Wrapper.createArray();
      this._wrapperFields.ranges.copy(other._wrapperFields.ranges);
      this._wrapperFields.intensities = Float32Wrapper.createArray();
      this._wrapperFields.intensities.copy(other._wrapperFields.intensities);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LaserScanRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._angle_minIntialized = false;
    this._angle_maxIntialized = false;
    this._angle_incrementIntialized = false;
    this._time_incrementIntialized = false;
    this._scan_timeIntialized = false;
    this._range_minIntialized = false;
    this._range_maxIntialized = false;
    this._wrapperFields.ranges = Float32Wrapper.createArray();
    this._wrapperFields.intensities = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new LaserScanWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LaserScanArrayWrapper;
  }
  static get ArrayType() {
    return LaserScanArrayWrapper;
  }
  static get refObjectArrayType() {
    return LaserScanRefStructArray
  }
  static get refObjectType() {
    return LaserScanRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._angle_minIntialized) {
        throw new TypeError('Invalid argument: angle_min in LaserScan');
      }
      if (!this._angle_maxIntialized) {
        throw new TypeError('Invalid argument: angle_max in LaserScan');
      }
      if (!this._angle_incrementIntialized) {
        throw new TypeError('Invalid argument: angle_increment in LaserScan');
      }
      if (!this._time_incrementIntialized) {
        throw new TypeError('Invalid argument: time_increment in LaserScan');
      }
      if (!this._scan_timeIntialized) {
        throw new TypeError('Invalid argument: scan_time in LaserScan');
      }
      if (!this._range_minIntialized) {
        throw new TypeError('Invalid argument: range_min in LaserScan');
      }
      if (!this._range_maxIntialized) {
        throw new TypeError('Invalid argument: range_max in LaserScan');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.ranges.freeze(own, checkConsistency);
    this._refObject.ranges = this._wrapperFields.ranges.refObject;
    if (own) {
      this._wrapperFields.ranges.fill(Float32Array.from([]));
      this._wrapperFields.ranges.freeze(own, checkConsistency);
      this._refObject.ranges = this._wrapperFields.ranges.refObject;
    }
    this._wrapperFields.intensities.freeze(own, checkConsistency);
    this._refObject.intensities = this._wrapperFields.intensities.refObject;
    if (own) {
      this._wrapperFields.intensities.fill(Float32Array.from([]));
      this._wrapperFields.intensities.freeze(own, checkConsistency);
      this._refObject.intensities = this._wrapperFields.intensities.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._angle_minIntialized = true;
    this._refObject.angle_min = refObject.angle_min;
    this._angle_maxIntialized = true;
    this._refObject.angle_max = refObject.angle_max;
    this._angle_incrementIntialized = true;
    this._refObject.angle_increment = refObject.angle_increment;
    this._time_incrementIntialized = true;
    this._refObject.time_increment = refObject.time_increment;
    this._scan_timeIntialized = true;
    this._refObject.scan_time = refObject.scan_time;
    this._range_minIntialized = true;
    this._refObject.range_min = refObject.range_min;
    this._range_maxIntialized = true;
    this._refObject.range_max = refObject.range_max;
    this._wrapperFields.ranges.copyRefObject(refObject.ranges);
    this._wrapperFields.intensities.copyRefObject(refObject.intensities);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.ranges.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.ranges);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.ranges, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.intensities.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.intensities);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.intensities, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    LaserScanWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'msg', interfaceName: 'LaserScan'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get angle_min() {
    if (!this._angle_minIntialized) {
      return undefined;
    }
    return this._refObject.angle_min;
  }
  set angle_min(value) {
    this._angle_minIntialized = true;
    this._refObject.angle_min = value;
  }
  get angle_max() {
    if (!this._angle_maxIntialized) {
      return undefined;
    }
    return this._refObject.angle_max;
  }
  set angle_max(value) {
    this._angle_maxIntialized = true;
    this._refObject.angle_max = value;
  }
  get angle_increment() {
    if (!this._angle_incrementIntialized) {
      return undefined;
    }
    return this._refObject.angle_increment;
  }
  set angle_increment(value) {
    this._angle_incrementIntialized = true;
    this._refObject.angle_increment = value;
  }
  get time_increment() {
    if (!this._time_incrementIntialized) {
      return undefined;
    }
    return this._refObject.time_increment;
  }
  set time_increment(value) {
    this._time_incrementIntialized = true;
    this._refObject.time_increment = value;
  }
  get scan_time() {
    if (!this._scan_timeIntialized) {
      return undefined;
    }
    return this._refObject.scan_time;
  }
  set scan_time(value) {
    this._scan_timeIntialized = true;
    this._refObject.scan_time = value;
  }
  get range_min() {
    if (!this._range_minIntialized) {
      return undefined;
    }
    return this._refObject.range_min;
  }
  set range_min(value) {
    this._range_minIntialized = true;
    this._refObject.range_min = value;
  }
  get range_max() {
    if (!this._range_maxIntialized) {
      return undefined;
    }
    return this._refObject.range_max;
  }
  set range_max(value) {
    this._range_maxIntialized = true;
    this._refObject.range_max = value;
  }
  get ranges() {
    return this._wrapperFields['ranges'].data;
  }
  set ranges(value) {
    this._wrapperFields['ranges'].fill(value);
  }
  get intensities() {
    return this._wrapperFields['intensities'].data;
  }
  set intensities(value) {
    this._wrapperFields['intensities'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new LaserScanRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._angle_minIntialized = true;
    this._angle_maxIntialized = true;
    this._angle_incrementIntialized = true;
    this._time_incrementIntialized = true;
    this._scan_timeIntialized = true;
    this._range_minIntialized = true;
    this._range_maxIntialized = true;
    this._wrapperFields.ranges.copyRefObject(this._refObject.ranges);
    this._wrapperFields.intensities.copyRefObject(this._refObject.intensities);
  }
  copy(other) {
    this._refObject = new LaserScanRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._angle_minIntialized = true;
    this._angle_maxIntialized = true;
    this._angle_incrementIntialized = true;
    this._time_incrementIntialized = true;
    this._scan_timeIntialized = true;
    this._range_minIntialized = true;
    this._range_maxIntialized = true;
    this._wrapperFields.ranges.copy(other._wrapperFields.ranges);
    this._wrapperFields.intensities.copy(other._wrapperFields.intensities);
  }
  static get classType() {
    return LaserScanWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"angle_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angle_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angle_increment","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_increment","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scan_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"range_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"range_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ranges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"intensities","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"LaserScan","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LaserScan"};
  }
  hasMember(name) {
    let memberNames = ["header","angle_min","angle_max","angle_increment","time_increment","scan_time","range_min","range_max","ranges","intensities"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LaserScanArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LaserScanWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LaserScanWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LaserScanArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LaserScanArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LaserScanRefArray(size);
    this._refObject = new LaserScanRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LaserScanWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LaserScanArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LaserScanArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LaserScanWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LaserScanWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LaserScanArrayWrapper;
  }
}
module.exports = LaserScanWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "angle_min",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angle_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angle_increment",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_increment",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "scan_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "range_min",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "range_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ranges",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "intensities",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "LaserScan",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LaserScan"
}
*/
