// rclnodejs/generated/sensor_msgs/sensor_msgs__srv__SetCameraInfo_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CameraInfoWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__CameraInfo.js');
const SetCameraInfo_RequestRefStruct = StructType({
  camera_info: CameraInfoWrapper.refObjectType,
});
const SetCameraInfo_RequestRefArray = ArrayType(SetCameraInfo_RequestRefStruct);
const SetCameraInfo_RequestRefStructArray = StructType({
  data: SetCameraInfo_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetCameraInfo_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetCameraInfo_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.camera_info =  new CameraInfoWrapper(other._wrapperFields.camera_info);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetCameraInfo_RequestRefStruct();
    this._wrapperFields.camera_info = new CameraInfoWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetCameraInfo_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetCameraInfo_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetCameraInfo_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetCameraInfo_RequestRefStructArray
  }
  static get refObjectType() {
    return SetCameraInfo_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.camera_info.freeze(own, checkConsistency);
    this._refObject.camera_info = this._wrapperFields.camera_info.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.camera_info.copyRefObject(refObject.camera_info);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    CameraInfoWrapper.freeStruct(refObject.camera_info);
  }
  static destoryRawROS(msg) {
    SetCameraInfo_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sensor_msgs', subFolder: 'srv', interfaceName: 'SetCameraInfo_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get camera_info() {
    return this._wrapperFields.camera_info;
  }
  set camera_info(value) {
    if (value instanceof CameraInfoWrapper) {
      this._wrapperFields.camera_info.copy(value);
    } else {
      this._wrapperFields.camera_info.copy(new CameraInfoWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetCameraInfo_RequestRefStruct(refObject.toObject());
    this._wrapperFields.camera_info.copyRefObject(this._refObject.camera_info);
  }
  copy(other) {
    this._refObject = new SetCameraInfo_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.camera_info.copy(other._wrapperFields.camera_info);
  }
  static get classType() {
    return SetCameraInfo_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"camera_info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"CameraInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"sensor_msgs","type":"SetCameraInfo_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetCameraInfo_Request"};
  }
  hasMember(name) {
    let memberNames = ["camera_info"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetCameraInfo_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetCameraInfo_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetCameraInfo_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetCameraInfo_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetCameraInfo_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetCameraInfo_RequestRefArray(size);
    this._refObject = new SetCameraInfo_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetCameraInfo_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetCameraInfo_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetCameraInfo_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetCameraInfo_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetCameraInfo_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetCameraInfo_RequestArrayWrapper;
  }
}
module.exports = SetCameraInfo_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "camera_info",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "CameraInfo",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sensor_msgs",
    "type": "SetCameraInfo_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetCameraInfo_Request"
}
*/
