// rclnodejs/generated/diagnostic_msgs/diagnostic_msgs__srv__SelfTest_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const DiagnosticStatusWrapper = require('../../generated/diagnostic_msgs/diagnostic_msgs__msg__DiagnosticStatus.js');
const SelfTest_ResponseRefStruct = StructType({
  id: primitiveTypes.string,
  passed: primitiveTypes.byte,
  status: DiagnosticStatusWrapper.refObjectArrayType,
});
const SelfTest_ResponseRefArray = ArrayType(SelfTest_ResponseRefStruct);
const SelfTest_ResponseRefStructArray = StructType({
  data: SelfTest_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SelfTest_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SelfTest_ResponseRefStruct(other._refObject.toObject());
      this._idIntialized = true;
      this._wrapperFields.id =  new StringWrapper(other._wrapperFields.id);
      this._passedIntialized = true;
      this._wrapperFields.status = DiagnosticStatusWrapper.createArray();
      this._wrapperFields.status.copy(other._wrapperFields.status);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SelfTest_ResponseRefStruct();
    this._idIntialized = false;
    this._wrapperFields.id = new StringWrapper();
    this._passedIntialized = false;
    this._wrapperFields.status = DiagnosticStatusWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SelfTest_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SelfTest_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return SelfTest_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SelfTest_ResponseRefStructArray
  }
  static get refObjectType() {
    return SelfTest_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in SelfTest_Response');
      }
      if (!this._passedIntialized) {
        throw new TypeError('Invalid argument: passed in SelfTest_Response');
      }
    }
    if (own) {
      this._wrapperFields.id.freeze(own, checkConsistency);
    }
    this._refObject.id = this._wrapperFields.id.refObject;
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._idIntialized = true;
    this._wrapperFields.id.data = refObject.id.data;
    this._passedIntialized = true;
    this._refObject.passed = refObject.passed;
    this._wrapperFields.status.copyRefObject(refObject.status);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.id);
    if (refObject.status.size != 0) {
      DiagnosticStatusWrapper.ArrayType.freeArray(refObject.status);
      if (DiagnosticStatusWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.status, DiagnosticStatusWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    SelfTest_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'diagnostic_msgs', subFolder: 'srv', interfaceName: 'SelfTest_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._wrapperFields.id.data;
  }
  set id(value) {
    this._idIntialized = true;
    this._wrapperFields.id.data = value;
  }
  get passed() {
    if (!this._passedIntialized) {
      return undefined;
    }
    return this._refObject.passed;
  }
  set passed(value) {
    this._passedIntialized = true;
    this._refObject.passed = value;
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    this._wrapperFields.status.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SelfTest_ResponseRefStruct(refObject.toObject());
    this._idIntialized = true;
    this._wrapperFields.id.copyRefObject(this._refObject.id);
    this._passedIntialized = true;
    this._wrapperFields.status.copyRefObject(this._refObject.status);
  }
  copy(other) {
    this._refObject = new SelfTest_ResponseRefStruct(other._refObject.toObject());
    this._idIntialized = true;
    this._wrapperFields.id.copy(other._wrapperFields.id);
    this._passedIntialized = true;
    this._wrapperFields.status.copy(other._wrapperFields.status);
  }
  static get classType() {
    return SelfTest_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"passed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"byte","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"diagnostic_msgs","type":"DiagnosticStatus","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"diagnostic_msgs","type":"SelfTest_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SelfTest_Response"};
  }
  hasMember(name) {
    let memberNames = ["id","passed","status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SelfTest_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SelfTest_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SelfTest_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SelfTest_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SelfTest_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SelfTest_ResponseRefArray(size);
    this._refObject = new SelfTest_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SelfTest_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SelfTest_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SelfTest_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SelfTest_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SelfTest_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SelfTest_ResponseArrayWrapper;
  }
}
module.exports = SelfTest_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "passed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "byte",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "diagnostic_msgs",
        "type": "DiagnosticStatus",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "diagnostic_msgs",
    "type": "SelfTest_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SelfTest_Response"
}
*/
