// rclnodejs/generated/geometry_msgs/geometry_msgs__msg__Accel.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const AccelRefStruct = StructType({
  linear: Vector3Wrapper.refObjectType,
  angular: Vector3Wrapper.refObjectType,
});
const AccelRefArray = ArrayType(AccelRefStruct);
const AccelRefStructArray = StructType({
  data: AccelRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AccelWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AccelRefStruct(other._refObject.toObject());
      this._wrapperFields.linear =  new Vector3Wrapper(other._wrapperFields.linear);
      this._wrapperFields.angular =  new Vector3Wrapper(other._wrapperFields.angular);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AccelRefStruct();
    this._wrapperFields.linear = new Vector3Wrapper();
    this._wrapperFields.angular = new Vector3Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new AccelWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AccelArrayWrapper;
  }
  static get ArrayType() {
    return AccelArrayWrapper;
  }
  static get refObjectArrayType() {
    return AccelRefStructArray
  }
  static get refObjectType() {
    return AccelRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.linear.freeze(own, checkConsistency);
    this._refObject.linear = this._wrapperFields.linear.refObject;
    this._wrapperFields.angular.freeze(own, checkConsistency);
    this._refObject.angular = this._wrapperFields.angular.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.linear.copyRefObject(refObject.linear);
    this._wrapperFields.angular.copyRefObject(refObject.angular);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Vector3Wrapper.freeStruct(refObject.linear);
    Vector3Wrapper.freeStruct(refObject.angular);
  }
  static destoryRawROS(msg) {
    AccelWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geometry_msgs', subFolder: 'msg', interfaceName: 'Accel'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get linear() {
    return this._wrapperFields.linear;
  }
  set linear(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.linear.copy(value);
    } else {
      this._wrapperFields.linear.copy(new Vector3Wrapper(value));
    }
  }
  get angular() {
    return this._wrapperFields.angular;
  }
  set angular(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.angular.copy(value);
    } else {
      this._wrapperFields.angular.copy(new Vector3Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new AccelRefStruct(refObject.toObject());
    this._wrapperFields.linear.copyRefObject(this._refObject.linear);
    this._wrapperFields.angular.copyRefObject(this._refObject.angular);
  }
  copy(other) {
    this._refObject = new AccelRefStruct(other._refObject.toObject());
    this._wrapperFields.linear.copy(other._wrapperFields.linear);
    this._wrapperFields.angular.copy(other._wrapperFields.angular);
  }
  static get classType() {
    return AccelWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"linear","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"angular","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geometry_msgs","type":"Accel","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Accel"};
  }
  hasMember(name) {
    let memberNames = ["linear","angular"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AccelArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AccelWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AccelWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AccelArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AccelArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AccelRefArray(size);
    this._refObject = new AccelRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AccelWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AccelArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AccelArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AccelWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AccelWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AccelArrayWrapper;
  }
}
module.exports = AccelWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "linear",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "angular",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geometry_msgs",
    "type": "Accel",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Accel"
}
*/
