// rclnodejs/generated/geometry_msgs/geometry_msgs__msg__Wrench.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const WrenchRefStruct = StructType({
  force: Vector3Wrapper.refObjectType,
  torque: Vector3Wrapper.refObjectType,
});
const WrenchRefArray = ArrayType(WrenchRefStruct);
const WrenchRefStructArray = StructType({
  data: WrenchRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WrenchWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WrenchRefStruct(other._refObject.toObject());
      this._wrapperFields.force =  new Vector3Wrapper(other._wrapperFields.force);
      this._wrapperFields.torque =  new Vector3Wrapper(other._wrapperFields.torque);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WrenchRefStruct();
    this._wrapperFields.force = new Vector3Wrapper();
    this._wrapperFields.torque = new Vector3Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new WrenchWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WrenchArrayWrapper;
  }
  static get ArrayType() {
    return WrenchArrayWrapper;
  }
  static get refObjectArrayType() {
    return WrenchRefStructArray
  }
  static get refObjectType() {
    return WrenchRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.force.freeze(own, checkConsistency);
    this._refObject.force = this._wrapperFields.force.refObject;
    this._wrapperFields.torque.freeze(own, checkConsistency);
    this._refObject.torque = this._wrapperFields.torque.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.force.copyRefObject(refObject.force);
    this._wrapperFields.torque.copyRefObject(refObject.torque);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Vector3Wrapper.freeStruct(refObject.force);
    Vector3Wrapper.freeStruct(refObject.torque);
  }
  static destoryRawROS(msg) {
    WrenchWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geometry_msgs', subFolder: 'msg', interfaceName: 'Wrench'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get force() {
    return this._wrapperFields.force;
  }
  set force(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.force.copy(value);
    } else {
      this._wrapperFields.force.copy(new Vector3Wrapper(value));
    }
  }
  get torque() {
    return this._wrapperFields.torque;
  }
  set torque(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.torque.copy(value);
    } else {
      this._wrapperFields.torque.copy(new Vector3Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new WrenchRefStruct(refObject.toObject());
    this._wrapperFields.force.copyRefObject(this._refObject.force);
    this._wrapperFields.torque.copyRefObject(this._refObject.torque);
  }
  copy(other) {
    this._refObject = new WrenchRefStruct(other._refObject.toObject());
    this._wrapperFields.force.copy(other._wrapperFields.force);
    this._wrapperFields.torque.copy(other._wrapperFields.torque);
  }
  static get classType() {
    return WrenchWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"force","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"torque","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Wrench"};
  }
  hasMember(name) {
    let memberNames = ["force","torque"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WrenchArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WrenchWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WrenchWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WrenchArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WrenchArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WrenchRefArray(size);
    this._refObject = new WrenchRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WrenchWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WrenchArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WrenchArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WrenchWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WrenchWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WrenchArrayWrapper;
  }
}
module.exports = WrenchWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "force",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "torque",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geometry_msgs",
    "type": "Wrench",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Wrench"
}
*/
