// rclnodejs/generated/geometry_msgs/geometry_msgs__msg__TransformStamped.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TransformWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Transform.js');
const TransformStampedRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  child_frame_id: primitiveTypes.string,
  transform: TransformWrapper.refObjectType,
});
const TransformStampedRefArray = ArrayType(TransformStampedRefStruct);
const TransformStampedRefStructArray = StructType({
  data: TransformStampedRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TransformStampedWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TransformStampedRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._child_frame_idIntialized = true;
      this._wrapperFields.child_frame_id =  new StringWrapper(other._wrapperFields.child_frame_id);
      this._wrapperFields.transform =  new TransformWrapper(other._wrapperFields.transform);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TransformStampedRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._child_frame_idIntialized = false;
    this._wrapperFields.child_frame_id = new StringWrapper();
    this._wrapperFields.transform = new TransformWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new TransformStampedWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TransformStampedArrayWrapper;
  }
  static get ArrayType() {
    return TransformStampedArrayWrapper;
  }
  static get refObjectArrayType() {
    return TransformStampedRefStructArray
  }
  static get refObjectType() {
    return TransformStampedRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._child_frame_idIntialized) {
        throw new TypeError('Invalid argument: child_frame_id in TransformStamped');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.child_frame_id.freeze(own, checkConsistency);
    }
    this._refObject.child_frame_id = this._wrapperFields.child_frame_id.refObject;
    this._wrapperFields.transform.freeze(own, checkConsistency);
    this._refObject.transform = this._wrapperFields.transform.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._child_frame_idIntialized = true;
    this._wrapperFields.child_frame_id.data = refObject.child_frame_id.data;
    this._wrapperFields.transform.copyRefObject(refObject.transform);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.child_frame_id);
    TransformWrapper.freeStruct(refObject.transform);
  }
  static destoryRawROS(msg) {
    TransformStampedWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geometry_msgs', subFolder: 'msg', interfaceName: 'TransformStamped'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get child_frame_id() {
    if (!this._child_frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.child_frame_id.data;
  }
  set child_frame_id(value) {
    this._child_frame_idIntialized = true;
    this._wrapperFields.child_frame_id.data = value;
  }
  get transform() {
    return this._wrapperFields.transform;
  }
  set transform(value) {
    if (value instanceof TransformWrapper) {
      this._wrapperFields.transform.copy(value);
    } else {
      this._wrapperFields.transform.copy(new TransformWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new TransformStampedRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._child_frame_idIntialized = true;
    this._wrapperFields.child_frame_id.copyRefObject(this._refObject.child_frame_id);
    this._wrapperFields.transform.copyRefObject(this._refObject.transform);
  }
  copy(other) {
    this._refObject = new TransformStampedRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._child_frame_idIntialized = true;
    this._wrapperFields.child_frame_id.copy(other._wrapperFields.child_frame_id);
    this._wrapperFields.transform.copy(other._wrapperFields.transform);
  }
  static get classType() {
    return TransformStampedWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"child_frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"transform","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Transform","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geometry_msgs","type":"TransformStamped","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TransformStamped"};
  }
  hasMember(name) {
    let memberNames = ["header","child_frame_id","transform"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TransformStampedArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TransformStampedWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TransformStampedWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TransformStampedArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TransformStampedArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TransformStampedRefArray(size);
    this._refObject = new TransformStampedRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TransformStampedWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TransformStampedArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TransformStampedArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TransformStampedWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TransformStampedWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TransformStampedArrayWrapper;
  }
}
module.exports = TransformStampedWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "child_frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "transform",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Transform",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geometry_msgs",
    "type": "TransformStamped",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TransformStamped"
}
*/
