// rclnodejs/generated/lifecycle_msgs/lifecycle_msgs__msg__TransitionDescription.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TransitionWrapper = require('../../generated/lifecycle_msgs/lifecycle_msgs__msg__Transition.js');
const StateWrapper = require('../../generated/lifecycle_msgs/lifecycle_msgs__msg__State.js');
const TransitionDescriptionRefStruct = StructType({
  transition: TransitionWrapper.refObjectType,
  start_state: StateWrapper.refObjectType,
  goal_state: StateWrapper.refObjectType,
});
const TransitionDescriptionRefArray = ArrayType(TransitionDescriptionRefStruct);
const TransitionDescriptionRefStructArray = StructType({
  data: TransitionDescriptionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TransitionDescriptionWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TransitionDescriptionRefStruct(other._refObject.toObject());
      this._wrapperFields.transition =  new TransitionWrapper(other._wrapperFields.transition);
      this._wrapperFields.start_state =  new StateWrapper(other._wrapperFields.start_state);
      this._wrapperFields.goal_state =  new StateWrapper(other._wrapperFields.goal_state);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TransitionDescriptionRefStruct();
    this._wrapperFields.transition = new TransitionWrapper();
    this._wrapperFields.start_state = new StateWrapper();
    this._wrapperFields.goal_state = new StateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new TransitionDescriptionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TransitionDescriptionArrayWrapper;
  }
  static get ArrayType() {
    return TransitionDescriptionArrayWrapper;
  }
  static get refObjectArrayType() {
    return TransitionDescriptionRefStructArray
  }
  static get refObjectType() {
    return TransitionDescriptionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.transition.freeze(own, checkConsistency);
    this._refObject.transition = this._wrapperFields.transition.refObject;
    this._wrapperFields.start_state.freeze(own, checkConsistency);
    this._refObject.start_state = this._wrapperFields.start_state.refObject;
    this._wrapperFields.goal_state.freeze(own, checkConsistency);
    this._refObject.goal_state = this._wrapperFields.goal_state.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.transition.copyRefObject(refObject.transition);
    this._wrapperFields.start_state.copyRefObject(refObject.start_state);
    this._wrapperFields.goal_state.copyRefObject(refObject.goal_state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TransitionWrapper.freeStruct(refObject.transition);
    StateWrapper.freeStruct(refObject.start_state);
    StateWrapper.freeStruct(refObject.goal_state);
  }
  static destoryRawROS(msg) {
    TransitionDescriptionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'lifecycle_msgs', subFolder: 'msg', interfaceName: 'TransitionDescription'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get transition() {
    return this._wrapperFields.transition;
  }
  set transition(value) {
    if (value instanceof TransitionWrapper) {
      this._wrapperFields.transition.copy(value);
    } else {
      this._wrapperFields.transition.copy(new TransitionWrapper(value));
    }
  }
  get start_state() {
    return this._wrapperFields.start_state;
  }
  set start_state(value) {
    if (value instanceof StateWrapper) {
      this._wrapperFields.start_state.copy(value);
    } else {
      this._wrapperFields.start_state.copy(new StateWrapper(value));
    }
  }
  get goal_state() {
    return this._wrapperFields.goal_state;
  }
  set goal_state(value) {
    if (value instanceof StateWrapper) {
      this._wrapperFields.goal_state.copy(value);
    } else {
      this._wrapperFields.goal_state.copy(new StateWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new TransitionDescriptionRefStruct(refObject.toObject());
    this._wrapperFields.transition.copyRefObject(this._refObject.transition);
    this._wrapperFields.start_state.copyRefObject(this._refObject.start_state);
    this._wrapperFields.goal_state.copyRefObject(this._refObject.goal_state);
  }
  copy(other) {
    this._refObject = new TransitionDescriptionRefStruct(other._refObject.toObject());
    this._wrapperFields.transition.copy(other._wrapperFields.transition);
    this._wrapperFields.start_state.copy(other._wrapperFields.start_state);
    this._wrapperFields.goal_state.copy(other._wrapperFields.goal_state);
  }
  static get classType() {
    return TransitionDescriptionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"transition","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"lifecycle_msgs","type":"Transition","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"lifecycle_msgs","type":"State","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"lifecycle_msgs","type":"State","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"lifecycle_msgs","type":"TransitionDescription","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TransitionDescription"};
  }
  hasMember(name) {
    let memberNames = ["transition","start_state","goal_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TransitionDescriptionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TransitionDescriptionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TransitionDescriptionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TransitionDescriptionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TransitionDescriptionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TransitionDescriptionRefArray(size);
    this._refObject = new TransitionDescriptionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TransitionDescriptionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TransitionDescriptionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TransitionDescriptionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TransitionDescriptionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TransitionDescriptionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TransitionDescriptionArrayWrapper;
  }
}
module.exports = TransitionDescriptionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "transition",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "lifecycle_msgs",
        "type": "Transition",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "start_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "lifecycle_msgs",
        "type": "State",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "goal_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "lifecycle_msgs",
        "type": "State",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "lifecycle_msgs",
    "type": "TransitionDescription",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TransitionDescription"
}
*/
