// rclnodejs/generated/visualization_msgs/visualization_msgs__srv__GetInteractiveMarkers_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const InteractiveMarkerWrapper = require('../../generated/visualization_msgs/visualization_msgs__msg__InteractiveMarker.js');
const GetInteractiveMarkers_ResponseRefStruct = StructType({
  sequence_number: primitiveTypes.uint64,
  markers: InteractiveMarkerWrapper.refObjectArrayType,
});
const GetInteractiveMarkers_ResponseRefArray = ArrayType(GetInteractiveMarkers_ResponseRefStruct);
const GetInteractiveMarkers_ResponseRefStructArray = StructType({
  data: GetInteractiveMarkers_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetInteractiveMarkers_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetInteractiveMarkers_ResponseRefStruct(other._refObject.toObject());
      this._sequence_numberIntialized = true;
      this._wrapperFields.markers = InteractiveMarkerWrapper.createArray();
      this._wrapperFields.markers.copy(other._wrapperFields.markers);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetInteractiveMarkers_ResponseRefStruct();
    this._sequence_numberIntialized = false;
    this._wrapperFields.markers = InteractiveMarkerWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetInteractiveMarkers_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetInteractiveMarkers_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetInteractiveMarkers_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetInteractiveMarkers_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetInteractiveMarkers_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._sequence_numberIntialized) {
        throw new TypeError('Invalid argument: sequence_number in GetInteractiveMarkers_Response');
      }
    }
    this._wrapperFields.markers.freeze(own, checkConsistency);
    this._refObject.markers = this._wrapperFields.markers.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._sequence_numberIntialized = true;
    this._refObject.sequence_number = refObject.sequence_number;
    this._wrapperFields.markers.copyRefObject(refObject.markers);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.markers.size != 0) {
      InteractiveMarkerWrapper.ArrayType.freeArray(refObject.markers);
      if (InteractiveMarkerWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.markers, InteractiveMarkerWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GetInteractiveMarkers_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'visualization_msgs', subFolder: 'srv', interfaceName: 'GetInteractiveMarkers_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get sequence_number() {
    if (!this._sequence_numberIntialized) {
      return undefined;
    }
    return this._refObject.sequence_number;
  }
  set sequence_number(value) {
    this._sequence_numberIntialized = true;
    this._refObject.sequence_number = value;
  }
  get markers() {
    return this._wrapperFields.markers;
  }
  set markers(value) {
    this._wrapperFields.markers.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetInteractiveMarkers_ResponseRefStruct(refObject.toObject());
    this._sequence_numberIntialized = true;
    this._wrapperFields.markers.copyRefObject(this._refObject.markers);
  }
  copy(other) {
    this._refObject = new GetInteractiveMarkers_ResponseRefStruct(other._refObject.toObject());
    this._sequence_numberIntialized = true;
    this._wrapperFields.markers.copy(other._wrapperFields.markers);
  }
  static get classType() {
    return GetInteractiveMarkers_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"sequence_number","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"markers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"visualization_msgs","type":"InteractiveMarker","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"visualization_msgs","type":"GetInteractiveMarkers_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetInteractiveMarkers_Response"};
  }
  hasMember(name) {
    let memberNames = ["sequence_number","markers"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetInteractiveMarkers_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetInteractiveMarkers_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetInteractiveMarkers_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetInteractiveMarkers_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetInteractiveMarkers_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetInteractiveMarkers_ResponseRefArray(size);
    this._refObject = new GetInteractiveMarkers_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetInteractiveMarkers_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetInteractiveMarkers_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetInteractiveMarkers_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetInteractiveMarkers_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetInteractiveMarkers_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetInteractiveMarkers_ResponseArrayWrapper;
  }
}
module.exports = GetInteractiveMarkers_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "sequence_number",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "markers",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "visualization_msgs",
        "type": "InteractiveMarker",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "visualization_msgs",
    "type": "GetInteractiveMarkers_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetInteractiveMarkers_Response"
}
*/
