// rclnodejs/generated/trajectory_msgs/trajectory_msgs__msg__JointTrajectoryPoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const JointTrajectoryPointRefStruct = StructType({
  positions: Float64Wrapper.refObjectArrayType,
  velocities: Float64Wrapper.refObjectArrayType,
  accelerations: Float64Wrapper.refObjectArrayType,
  effort: Float64Wrapper.refObjectArrayType,
  time_from_start: DurationWrapper.refObjectType,
});
const JointTrajectoryPointRefArray = ArrayType(JointTrajectoryPointRefStruct);
const JointTrajectoryPointRefStructArray = StructType({
  data: JointTrajectoryPointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class JointTrajectoryPointWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new JointTrajectoryPointRefStruct(other._refObject.toObject());
      this._wrapperFields.positions = Float64Wrapper.createArray();
      this._wrapperFields.positions.copy(other._wrapperFields.positions);
      this._wrapperFields.velocities = Float64Wrapper.createArray();
      this._wrapperFields.velocities.copy(other._wrapperFields.velocities);
      this._wrapperFields.accelerations = Float64Wrapper.createArray();
      this._wrapperFields.accelerations.copy(other._wrapperFields.accelerations);
      this._wrapperFields.effort = Float64Wrapper.createArray();
      this._wrapperFields.effort.copy(other._wrapperFields.effort);
      this._wrapperFields.time_from_start =  new DurationWrapper(other._wrapperFields.time_from_start);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new JointTrajectoryPointRefStruct();
    this._wrapperFields.positions = Float64Wrapper.createArray();
    this._wrapperFields.velocities = Float64Wrapper.createArray();
    this._wrapperFields.accelerations = Float64Wrapper.createArray();
    this._wrapperFields.effort = Float64Wrapper.createArray();
    this._wrapperFields.time_from_start = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new JointTrajectoryPointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JointTrajectoryPointArrayWrapper;
  }
  static get ArrayType() {
    return JointTrajectoryPointArrayWrapper;
  }
  static get refObjectArrayType() {
    return JointTrajectoryPointRefStructArray
  }
  static get refObjectType() {
    return JointTrajectoryPointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.positions.freeze(own, checkConsistency);
    this._refObject.positions = this._wrapperFields.positions.refObject;
    if (own) {
      this._wrapperFields.positions.fill(Float64Array.from([]));
      this._wrapperFields.positions.freeze(own, checkConsistency);
      this._refObject.positions = this._wrapperFields.positions.refObject;
    }
    this._wrapperFields.velocities.freeze(own, checkConsistency);
    this._refObject.velocities = this._wrapperFields.velocities.refObject;
    if (own) {
      this._wrapperFields.velocities.fill(Float64Array.from([]));
      this._wrapperFields.velocities.freeze(own, checkConsistency);
      this._refObject.velocities = this._wrapperFields.velocities.refObject;
    }
    this._wrapperFields.accelerations.freeze(own, checkConsistency);
    this._refObject.accelerations = this._wrapperFields.accelerations.refObject;
    if (own) {
      this._wrapperFields.accelerations.fill(Float64Array.from([]));
      this._wrapperFields.accelerations.freeze(own, checkConsistency);
      this._refObject.accelerations = this._wrapperFields.accelerations.refObject;
    }
    this._wrapperFields.effort.freeze(own, checkConsistency);
    this._refObject.effort = this._wrapperFields.effort.refObject;
    if (own) {
      this._wrapperFields.effort.fill(Float64Array.from([]));
      this._wrapperFields.effort.freeze(own, checkConsistency);
      this._refObject.effort = this._wrapperFields.effort.refObject;
    }
    this._wrapperFields.time_from_start.freeze(own, checkConsistency);
    this._refObject.time_from_start = this._wrapperFields.time_from_start.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.positions.copyRefObject(refObject.positions);
    this._wrapperFields.velocities.copyRefObject(refObject.velocities);
    this._wrapperFields.accelerations.copyRefObject(refObject.accelerations);
    this._wrapperFields.effort.copyRefObject(refObject.effort);
    this._wrapperFields.time_from_start.copyRefObject(refObject.time_from_start);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.positions.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.positions);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.positions, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.velocities.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.velocities);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.velocities, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.accelerations.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.accelerations);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.accelerations, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.effort.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.effort);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.effort, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    DurationWrapper.freeStruct(refObject.time_from_start);
  }
  static destoryRawROS(msg) {
    JointTrajectoryPointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trajectory_msgs', subFolder: 'msg', interfaceName: 'JointTrajectoryPoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get positions() {
    return this._wrapperFields['positions'].data;
  }
  set positions(value) {
    this._wrapperFields['positions'].fill(value);
  }
  get velocities() {
    return this._wrapperFields['velocities'].data;
  }
  set velocities(value) {
    this._wrapperFields['velocities'].fill(value);
  }
  get accelerations() {
    return this._wrapperFields['accelerations'].data;
  }
  set accelerations(value) {
    this._wrapperFields['accelerations'].fill(value);
  }
  get effort() {
    return this._wrapperFields['effort'].data;
  }
  set effort(value) {
    this._wrapperFields['effort'].fill(value);
  }
  get time_from_start() {
    return this._wrapperFields.time_from_start;
  }
  set time_from_start(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.time_from_start.copy(value);
    } else {
      this._wrapperFields.time_from_start.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new JointTrajectoryPointRefStruct(refObject.toObject());
    this._wrapperFields.positions.copyRefObject(this._refObject.positions);
    this._wrapperFields.velocities.copyRefObject(this._refObject.velocities);
    this._wrapperFields.accelerations.copyRefObject(this._refObject.accelerations);
    this._wrapperFields.effort.copyRefObject(this._refObject.effort);
    this._wrapperFields.time_from_start.copyRefObject(this._refObject.time_from_start);
  }
  copy(other) {
    this._refObject = new JointTrajectoryPointRefStruct(other._refObject.toObject());
    this._wrapperFields.positions.copy(other._wrapperFields.positions);
    this._wrapperFields.velocities.copy(other._wrapperFields.velocities);
    this._wrapperFields.accelerations.copy(other._wrapperFields.accelerations);
    this._wrapperFields.effort.copy(other._wrapperFields.effort);
    this._wrapperFields.time_from_start.copy(other._wrapperFields.time_from_start);
  }
  static get classType() {
    return JointTrajectoryPointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"positions","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocities","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accelerations","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"effort","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_from_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"trajectory_msgs","type":"JointTrajectoryPoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JointTrajectoryPoint"};
  }
  hasMember(name) {
    let memberNames = ["positions","velocities","accelerations","effort","time_from_start"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class JointTrajectoryPointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JointTrajectoryPointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JointTrajectoryPointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointTrajectoryPointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointTrajectoryPointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new JointTrajectoryPointRefArray(size);
    this._refObject = new JointTrajectoryPointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JointTrajectoryPointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JointTrajectoryPointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JointTrajectoryPointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JointTrajectoryPointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JointTrajectoryPointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JointTrajectoryPointArrayWrapper;
  }
}
module.exports = JointTrajectoryPointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "positions",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocities",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accelerations",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "effort",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_from_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trajectory_msgs",
    "type": "JointTrajectoryPoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "JointTrajectoryPoint"
}
*/
