// rclnodejs/generated/nav_msgs/nav_msgs__srv__SetMap_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const OccupancyGridWrapper = require('../../generated/nav_msgs/nav_msgs__msg__OccupancyGrid.js');
const PoseWithCovarianceStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseWithCovarianceStamped.js');
const SetMap_RequestRefStruct = StructType({
  map: OccupancyGridWrapper.refObjectType,
  initial_pose: PoseWithCovarianceStampedWrapper.refObjectType,
});
const SetMap_RequestRefArray = ArrayType(SetMap_RequestRefStruct);
const SetMap_RequestRefStructArray = StructType({
  data: SetMap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetMap_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetMap_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.map =  new OccupancyGridWrapper(other._wrapperFields.map);
      this._wrapperFields.initial_pose =  new PoseWithCovarianceStampedWrapper(other._wrapperFields.initial_pose);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetMap_RequestRefStruct();
    this._wrapperFields.map = new OccupancyGridWrapper();
    this._wrapperFields.initial_pose = new PoseWithCovarianceStampedWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetMap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetMap_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetMap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetMap_RequestRefStructArray
  }
  static get refObjectType() {
    return SetMap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.map.freeze(own, checkConsistency);
    this._refObject.map = this._wrapperFields.map.refObject;
    this._wrapperFields.initial_pose.freeze(own, checkConsistency);
    this._refObject.initial_pose = this._wrapperFields.initial_pose.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map.copyRefObject(refObject.map);
    this._wrapperFields.initial_pose.copyRefObject(refObject.initial_pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    OccupancyGridWrapper.freeStruct(refObject.map);
    PoseWithCovarianceStampedWrapper.freeStruct(refObject.initial_pose);
  }
  static destoryRawROS(msg) {
    SetMap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'srv', interfaceName: 'SetMap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map() {
    return this._wrapperFields.map;
  }
  set map(value) {
    if (value instanceof OccupancyGridWrapper) {
      this._wrapperFields.map.copy(value);
    } else {
      this._wrapperFields.map.copy(new OccupancyGridWrapper(value));
    }
  }
  get initial_pose() {
    return this._wrapperFields.initial_pose;
  }
  set initial_pose(value) {
    if (value instanceof PoseWithCovarianceStampedWrapper) {
      this._wrapperFields.initial_pose.copy(value);
    } else {
      this._wrapperFields.initial_pose.copy(new PoseWithCovarianceStampedWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetMap_RequestRefStruct(refObject.toObject());
    this._wrapperFields.map.copyRefObject(this._refObject.map);
    this._wrapperFields.initial_pose.copyRefObject(this._refObject.initial_pose);
  }
  copy(other) {
    this._refObject = new SetMap_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.map.copy(other._wrapperFields.map);
    this._wrapperFields.initial_pose.copy(other._wrapperFields.initial_pose);
  }
  static get classType() {
    return SetMap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"OccupancyGrid","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"initial_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseWithCovarianceStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"SetMap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetMap_Request"};
  }
  hasMember(name) {
    let memberNames = ["map","initial_pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetMap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetMap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetMap_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMap_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetMap_RequestRefArray(size);
    this._refObject = new SetMap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetMap_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetMap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetMap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetMap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetMap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetMap_RequestArrayWrapper;
  }
}
module.exports = SetMap_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "map",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_msgs",
        "type": "OccupancyGrid",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "initial_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseWithCovarianceStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav_msgs",
    "type": "SetMap_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetMap_Request"
}
*/
