// rclnodejs/generated/nav_msgs/nav_msgs__msg__GridCells.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const GridCellsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  cell_width: primitiveTypes.float32,
  cell_height: primitiveTypes.float32,
  cells: PointWrapper.refObjectArrayType,
});
const GridCellsRefArray = ArrayType(GridCellsRefStruct);
const GridCellsRefStructArray = StructType({
  data: GridCellsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GridCellsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GridCellsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._cell_widthIntialized = true;
      this._cell_heightIntialized = true;
      this._wrapperFields.cells = PointWrapper.createArray();
      this._wrapperFields.cells.copy(other._wrapperFields.cells);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GridCellsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._cell_widthIntialized = false;
    this._cell_heightIntialized = false;
    this._wrapperFields.cells = PointWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GridCellsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GridCellsArrayWrapper;
  }
  static get ArrayType() {
    return GridCellsArrayWrapper;
  }
  static get refObjectArrayType() {
    return GridCellsRefStructArray
  }
  static get refObjectType() {
    return GridCellsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._cell_widthIntialized) {
        throw new TypeError('Invalid argument: cell_width in GridCells');
      }
      if (!this._cell_heightIntialized) {
        throw new TypeError('Invalid argument: cell_height in GridCells');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.cells.freeze(own, checkConsistency);
    this._refObject.cells = this._wrapperFields.cells.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._cell_widthIntialized = true;
    this._refObject.cell_width = refObject.cell_width;
    this._cell_heightIntialized = true;
    this._refObject.cell_height = refObject.cell_height;
    this._wrapperFields.cells.copyRefObject(refObject.cells);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.cells.size != 0) {
      PointWrapper.ArrayType.freeArray(refObject.cells);
      if (PointWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.cells, PointWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GridCellsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'msg', interfaceName: 'GridCells'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get cell_width() {
    if (!this._cell_widthIntialized) {
      return undefined;
    }
    return this._refObject.cell_width;
  }
  set cell_width(value) {
    this._cell_widthIntialized = true;
    this._refObject.cell_width = value;
  }
  get cell_height() {
    if (!this._cell_heightIntialized) {
      return undefined;
    }
    return this._refObject.cell_height;
  }
  set cell_height(value) {
    this._cell_heightIntialized = true;
    this._refObject.cell_height = value;
  }
  get cells() {
    return this._wrapperFields.cells;
  }
  set cells(value) {
    this._wrapperFields.cells.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GridCellsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._cell_widthIntialized = true;
    this._cell_heightIntialized = true;
    this._wrapperFields.cells.copyRefObject(this._refObject.cells);
  }
  copy(other) {
    this._refObject = new GridCellsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._cell_widthIntialized = true;
    this._cell_heightIntialized = true;
    this._wrapperFields.cells.copy(other._wrapperFields.cells);
  }
  static get classType() {
    return GridCellsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"cell_width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cell_height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cells","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"GridCells","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GridCells"};
  }
  hasMember(name) {
    let memberNames = ["header","cell_width","cell_height","cells"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GridCellsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GridCellsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GridCellsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GridCellsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GridCellsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GridCellsRefArray(size);
    this._refObject = new GridCellsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GridCellsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GridCellsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GridCellsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GridCellsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GridCellsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GridCellsArrayWrapper;
  }
}
module.exports = GridCellsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "cell_width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cell_height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cells",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav_msgs",
    "type": "GridCells",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GridCells"
}
*/
