// rclnodejs/generated/diagnostic_msgs/diagnostic_msgs__msg__DiagnosticArray.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const DiagnosticStatusWrapper = require('../../generated/diagnostic_msgs/diagnostic_msgs__msg__DiagnosticStatus.js');
const DiagnosticArrayRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  status: DiagnosticStatusWrapper.refObjectArrayType,
});
const DiagnosticArrayRefArray = ArrayType(DiagnosticArrayRefStruct);
const DiagnosticArrayRefStructArray = StructType({
  data: DiagnosticArrayRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DiagnosticArrayWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DiagnosticArrayRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.status = DiagnosticStatusWrapper.createArray();
      this._wrapperFields.status.copy(other._wrapperFields.status);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DiagnosticArrayRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.status = DiagnosticStatusWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new DiagnosticArrayWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DiagnosticArrayArrayWrapper;
  }
  static get ArrayType() {
    return DiagnosticArrayArrayWrapper;
  }
  static get refObjectArrayType() {
    return DiagnosticArrayRefStructArray
  }
  static get refObjectType() {
    return DiagnosticArrayRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.status.copyRefObject(refObject.status);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.status.size != 0) {
      DiagnosticStatusWrapper.ArrayType.freeArray(refObject.status);
      if (DiagnosticStatusWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.status, DiagnosticStatusWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    DiagnosticArrayWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'diagnostic_msgs', subFolder: 'msg', interfaceName: 'DiagnosticArray'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    this._wrapperFields.status.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new DiagnosticArrayRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.status.copyRefObject(this._refObject.status);
  }
  copy(other) {
    this._refObject = new DiagnosticArrayRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.status.copy(other._wrapperFields.status);
  }
  static get classType() {
    return DiagnosticArrayWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"status","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"diagnostic_msgs","type":"DiagnosticStatus","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"diagnostic_msgs","type":"DiagnosticArray","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DiagnosticArray"};
  }
  hasMember(name) {
    let memberNames = ["header","status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DiagnosticArrayArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DiagnosticArrayWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DiagnosticArrayWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DiagnosticArrayArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DiagnosticArrayArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DiagnosticArrayRefArray(size);
    this._refObject = new DiagnosticArrayRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DiagnosticArrayWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DiagnosticArrayArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DiagnosticArrayArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DiagnosticArrayWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DiagnosticArrayWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DiagnosticArrayArrayWrapper;
  }
}
module.exports = DiagnosticArrayWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "diagnostic_msgs",
        "type": "DiagnosticStatus",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "diagnostic_msgs",
    "type": "DiagnosticArray",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DiagnosticArray"
}
*/
