// rclnodejs/generated/action_msgs/action_msgs__msg__GoalInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const GoalInfoRefStruct = StructType({
  goal_id: UUIDWrapper.refObjectType,
  stamp: TimeWrapper.refObjectType,
});
const GoalInfoRefArray = ArrayType(GoalInfoRefStruct);
const GoalInfoRefStructArray = StructType({
  data: GoalInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GoalInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GoalInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.goal_id =  new UUIDWrapper(other._wrapperFields.goal_id);
      this._wrapperFields.stamp =  new TimeWrapper(other._wrapperFields.stamp);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GoalInfoRefStruct();
    this._wrapperFields.goal_id = new UUIDWrapper();
    this._wrapperFields.stamp = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GoalInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GoalInfoArrayWrapper;
  }
  static get ArrayType() {
    return GoalInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return GoalInfoRefStructArray
  }
  static get refObjectType() {
    return GoalInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.goal_id.freeze(own, checkConsistency);
    this._refObject.goal_id = this._wrapperFields.goal_id.refObject;
    this._wrapperFields.stamp.freeze(own, checkConsistency);
    this._refObject.stamp = this._wrapperFields.stamp.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.goal_id.copyRefObject(refObject.goal_id);
    this._wrapperFields.stamp.copyRefObject(refObject.stamp);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    UUIDWrapper.freeStruct(refObject.goal_id);
    TimeWrapper.freeStruct(refObject.stamp);
  }
  static destoryRawROS(msg) {
    GoalInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'action_msgs', subFolder: 'msg', interfaceName: 'GoalInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get goal_id() {
    return this._wrapperFields.goal_id;
  }
  set goal_id(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.goal_id.copy(value);
    } else {
      this._wrapperFields.goal_id.copy(new UUIDWrapper(value));
    }
  }
  get stamp() {
    return this._wrapperFields.stamp;
  }
  set stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.stamp.copy(value);
    } else {
      this._wrapperFields.stamp.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GoalInfoRefStruct(refObject.toObject());
    this._wrapperFields.goal_id.copyRefObject(this._refObject.goal_id);
    this._wrapperFields.stamp.copyRefObject(this._refObject.stamp);
  }
  copy(other) {
    this._refObject = new GoalInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.goal_id.copy(other._wrapperFields.goal_id);
    this._wrapperFields.stamp.copy(other._wrapperFields.stamp);
  }
  static get classType() {
    return GoalInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"goal_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"action_msgs","type":"GoalInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GoalInfo"};
  }
  hasMember(name) {
    let memberNames = ["goal_id","stamp"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GoalInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GoalInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GoalInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GoalInfoRefArray(size);
    this._refObject = new GoalInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GoalInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GoalInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GoalInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GoalInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GoalInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GoalInfoArrayWrapper;
  }
}
module.exports = GoalInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "goal_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "stamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "action_msgs",
    "type": "GoalInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GoalInfo"
}
*/
