// rclnodejs/generated/action_msgs/action_msgs__msg__GoalStatusArray.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GoalStatusWrapper = require('../../generated/action_msgs/action_msgs__msg__GoalStatus.js');
const GoalStatusArrayRefStruct = StructType({
  status_list: GoalStatusWrapper.refObjectArrayType,
});
const GoalStatusArrayRefArray = ArrayType(GoalStatusArrayRefStruct);
const GoalStatusArrayRefStructArray = StructType({
  data: GoalStatusArrayRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GoalStatusArrayWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GoalStatusArrayRefStruct(other._refObject.toObject());
      this._wrapperFields.status_list = GoalStatusWrapper.createArray();
      this._wrapperFields.status_list.copy(other._wrapperFields.status_list);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GoalStatusArrayRefStruct();
    this._wrapperFields.status_list = GoalStatusWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GoalStatusArrayWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GoalStatusArrayArrayWrapper;
  }
  static get ArrayType() {
    return GoalStatusArrayArrayWrapper;
  }
  static get refObjectArrayType() {
    return GoalStatusArrayRefStructArray
  }
  static get refObjectType() {
    return GoalStatusArrayRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.status_list.freeze(own, checkConsistency);
    this._refObject.status_list = this._wrapperFields.status_list.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.status_list.copyRefObject(refObject.status_list);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.status_list.size != 0) {
      GoalStatusWrapper.ArrayType.freeArray(refObject.status_list);
      if (GoalStatusWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.status_list, GoalStatusWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GoalStatusArrayWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'action_msgs', subFolder: 'msg', interfaceName: 'GoalStatusArray'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status_list() {
    return this._wrapperFields.status_list;
  }
  set status_list(value) {
    this._wrapperFields.status_list.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GoalStatusArrayRefStruct(refObject.toObject());
    this._wrapperFields.status_list.copyRefObject(this._refObject.status_list);
  }
  copy(other) {
    this._refObject = new GoalStatusArrayRefStruct(other._refObject.toObject());
    this._wrapperFields.status_list.copy(other._wrapperFields.status_list);
  }
  static get classType() {
    return GoalStatusArrayWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"status_list","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"action_msgs","type":"GoalStatus","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"action_msgs","type":"GoalStatusArray","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GoalStatusArray"};
  }
  hasMember(name) {
    let memberNames = ["status_list"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GoalStatusArrayArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GoalStatusArrayWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GoalStatusArrayWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalStatusArrayArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalStatusArrayArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GoalStatusArrayRefArray(size);
    this._refObject = new GoalStatusArrayRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GoalStatusArrayWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GoalStatusArrayArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GoalStatusArrayArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GoalStatusArrayWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GoalStatusArrayWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GoalStatusArrayArrayWrapper;
  }
}
module.exports = GoalStatusArrayWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "status_list",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "action_msgs",
        "type": "GoalStatus",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "action_msgs",
    "type": "GoalStatusArray",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GoalStatusArray"
}
*/
