// rclnodejs/generated/action_msgs/action_msgs__msg__GoalStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GoalInfoWrapper = require('../../generated/action_msgs/action_msgs__msg__GoalInfo.js');
const GoalStatusRefStruct = StructType({
  goal_info: GoalInfoWrapper.refObjectType,
  status: primitiveTypes.int8,
});
const GoalStatusRefArray = ArrayType(GoalStatusRefStruct);
const GoalStatusRefStructArray = StructType({
  data: GoalStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GoalStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GoalStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.goal_info =  new GoalInfoWrapper(other._wrapperFields.goal_info);
      this._statusIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GoalStatusRefStruct();
    this._wrapperFields.goal_info = new GoalInfoWrapper();
    this._statusIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GoalStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GoalStatusArrayWrapper;
  }
  static get ArrayType() {
    return GoalStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return GoalStatusRefStructArray
  }
  static get refObjectType() {
    return GoalStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._statusIntialized) {
        throw new TypeError('Invalid argument: status in GoalStatus');
      }
    }
    this._wrapperFields.goal_info.freeze(own, checkConsistency);
    this._refObject.goal_info = this._wrapperFields.goal_info.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.goal_info.copyRefObject(refObject.goal_info);
    this._statusIntialized = true;
    this._refObject.status = refObject.status;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GoalInfoWrapper.freeStruct(refObject.goal_info);
  }
  static destoryRawROS(msg) {
    GoalStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'action_msgs', subFolder: 'msg', interfaceName: 'GoalStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get goal_info() {
    return this._wrapperFields.goal_info;
  }
  set goal_info(value) {
    if (value instanceof GoalInfoWrapper) {
      this._wrapperFields.goal_info.copy(value);
    } else {
      this._wrapperFields.goal_info.copy(new GoalInfoWrapper(value));
    }
  }
  get status() {
    if (!this._statusIntialized) {
      return undefined;
    }
    return this._refObject.status;
  }
  set status(value) {
    this._statusIntialized = true;
    this._refObject.status = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GoalStatusRefStruct(refObject.toObject());
    this._wrapperFields.goal_info.copyRefObject(this._refObject.goal_info);
    this._statusIntialized = true;
  }
  copy(other) {
    this._refObject = new GoalStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.goal_info.copy(other._wrapperFields.goal_info);
    this._statusIntialized = true;
  }
  static get classType() {
    return GoalStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"STATUS_UNKNOWN","value":0},{"type":"int8","name":"STATUS_ACCEPTED","value":1},{"type":"int8","name":"STATUS_EXECUTING","value":2},{"type":"int8","name":"STATUS_CANCELING","value":3},{"type":"int8","name":"STATUS_SUCCEEDED","value":4},{"type":"int8","name":"STATUS_CANCELED","value":5},{"type":"int8","name":"STATUS_ABORTED","value":6}],"fields":[{"name":"goal_info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"action_msgs","type":"GoalInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"action_msgs","type":"GoalStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GoalStatus"};
  }
  hasMember(name) {
    let memberNames = ["goal_info","status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GoalStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GoalStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GoalStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GoalStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GoalStatusRefArray(size);
    this._refObject = new GoalStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GoalStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GoalStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GoalStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GoalStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GoalStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GoalStatusArrayWrapper;
  }
}
// Define constants (7 in total)
Object.defineProperty(GoalStatusWrapper, "STATUS_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GoalStatusWrapper, "STATUS_ACCEPTED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GoalStatusWrapper, "STATUS_EXECUTING", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GoalStatusWrapper, "STATUS_CANCELING", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GoalStatusWrapper, "STATUS_SUCCEEDED", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GoalStatusWrapper, "STATUS_CANCELED", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GoalStatusWrapper, "STATUS_ABORTED", {value: 6, writable: false, enumerable: true, configurable: true});
module.exports = GoalStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "int8",
      "name": "STATUS_UNKNOWN",
      "value": 0
    },
    {
      "type": "int8",
      "name": "STATUS_ACCEPTED",
      "value": 1
    },
    {
      "type": "int8",
      "name": "STATUS_EXECUTING",
      "value": 2
    },
    {
      "type": "int8",
      "name": "STATUS_CANCELING",
      "value": 3
    },
    {
      "type": "int8",
      "name": "STATUS_SUCCEEDED",
      "value": 4
    },
    {
      "type": "int8",
      "name": "STATUS_CANCELED",
      "value": 5
    },
    {
      "type": "int8",
      "name": "STATUS_ABORTED",
      "value": 6
    }
  ],
  "fields": [
    {
      "name": "goal_info",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "action_msgs",
        "type": "GoalInfo",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "action_msgs",
    "type": "GoalStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GoalStatus"
}
*/
