// rclnodejs/generated/geometry_msgs/geometry_msgs__msg__PoseWithCovariance.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const PoseWithCovarianceRefStruct = StructType({
  pose: PoseWrapper.refObjectType,
  covariance: ArrayType(primitiveTypes.float64, 36),
});
const PoseWithCovarianceRefArray = ArrayType(PoseWithCovarianceRefStruct);
const PoseWithCovarianceRefStructArray = StructType({
  data: PoseWithCovarianceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PoseWithCovarianceWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PoseWithCovarianceRefStruct(other._refObject.toObject());
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._wrapperFields.covariance = Float64Wrapper.createArray();
      this._wrapperFields.covariance.copy(other._wrapperFields.covariance);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PoseWithCovarianceRefStruct();
    this._wrapperFields.pose = new PoseWrapper();
    this._wrapperFields.covariance = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new PoseWithCovarianceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PoseWithCovarianceArrayWrapper;
  }
  static get ArrayType() {
    return PoseWithCovarianceArrayWrapper;
  }
  static get refObjectArrayType() {
    return PoseWithCovarianceRefStructArray
  }
  static get refObjectType() {
    return PoseWithCovarianceRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    for (let i = 0; i < this._wrapperFields.covariance.data.length; i++) {
      this._refObject.covariance[i] = this._wrapperFields.covariance.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._refObject.covariance = refObject.covariance;
    this._wrapperFields.covariance.fill(refObject.covariance.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseWrapper.freeStruct(refObject.pose);
  }
  static destoryRawROS(msg) {
    PoseWithCovarianceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geometry_msgs', subFolder: 'msg', interfaceName: 'PoseWithCovariance'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get covariance() {
    return this._wrapperFields['covariance'].data;
  }
  set covariance(value) {
    if (value.length !== 36) {
      throw new RangeError('The length of the array must be 36.');
    }
    this._wrapperFields['covariance'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new PoseWithCovarianceRefStruct(refObject.toObject());
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._wrapperFields.covariance.fill(refObject.covariance.toArray());
  }
  copy(other) {
    this._refObject = new PoseWithCovarianceRefStruct(other._refObject.toObject());
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._wrapperFields.covariance.copy(other._wrapperFields.covariance);
  }
  static get classType() {
    return PoseWithCovarianceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"covariance","type":{"isArray":true,"arraySize":36,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"geometry_msgs","type":"PoseWithCovariance","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PoseWithCovariance"};
  }
  hasMember(name) {
    let memberNames = ["pose","covariance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PoseWithCovarianceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PoseWithCovarianceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PoseWithCovarianceWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PoseWithCovarianceArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PoseWithCovarianceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PoseWithCovarianceRefArray(size);
    this._refObject = new PoseWithCovarianceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PoseWithCovarianceWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PoseWithCovarianceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PoseWithCovarianceArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PoseWithCovarianceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PoseWithCovarianceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PoseWithCovarianceArrayWrapper;
  }
}
module.exports = PoseWithCovarianceWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "covariance",
      "type": {
        "isArray": true,
        "arraySize": 36,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geometry_msgs",
    "type": "PoseWithCovariance",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PoseWithCovariance"
}
*/
