// rclnodejs/generated/geometry_msgs/geometry_msgs__msg__TwistWithCovariance.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const TwistWithCovarianceRefStruct = StructType({
  twist: TwistWrapper.refObjectType,
  covariance: ArrayType(primitiveTypes.float64, 36),
});
const TwistWithCovarianceRefArray = ArrayType(TwistWithCovarianceRefStruct);
const TwistWithCovarianceRefStructArray = StructType({
  data: TwistWithCovarianceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TwistWithCovarianceWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TwistWithCovarianceRefStruct(other._refObject.toObject());
      this._wrapperFields.twist =  new TwistWrapper(other._wrapperFields.twist);
      this._wrapperFields.covariance = Float64Wrapper.createArray();
      this._wrapperFields.covariance.copy(other._wrapperFields.covariance);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TwistWithCovarianceRefStruct();
    this._wrapperFields.twist = new TwistWrapper();
    this._wrapperFields.covariance = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new TwistWithCovarianceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TwistWithCovarianceArrayWrapper;
  }
  static get ArrayType() {
    return TwistWithCovarianceArrayWrapper;
  }
  static get refObjectArrayType() {
    return TwistWithCovarianceRefStructArray
  }
  static get refObjectType() {
    return TwistWithCovarianceRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.twist.freeze(own, checkConsistency);
    this._refObject.twist = this._wrapperFields.twist.refObject;
    for (let i = 0; i < this._wrapperFields.covariance.data.length; i++) {
      this._refObject.covariance[i] = this._wrapperFields.covariance.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.twist.copyRefObject(refObject.twist);
    this._refObject.covariance = refObject.covariance;
    this._wrapperFields.covariance.fill(refObject.covariance.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TwistWrapper.freeStruct(refObject.twist);
  }
  static destoryRawROS(msg) {
    TwistWithCovarianceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geometry_msgs', subFolder: 'msg', interfaceName: 'TwistWithCovariance'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get twist() {
    return this._wrapperFields.twist;
  }
  set twist(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.twist.copy(value);
    } else {
      this._wrapperFields.twist.copy(new TwistWrapper(value));
    }
  }
  get covariance() {
    return this._wrapperFields['covariance'].data;
  }
  set covariance(value) {
    if (value.length !== 36) {
      throw new RangeError('The length of the array must be 36.');
    }
    this._wrapperFields['covariance'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new TwistWithCovarianceRefStruct(refObject.toObject());
    this._wrapperFields.twist.copyRefObject(this._refObject.twist);
    this._wrapperFields.covariance.fill(refObject.covariance.toArray());
  }
  copy(other) {
    this._refObject = new TwistWithCovarianceRefStruct(other._refObject.toObject());
    this._wrapperFields.twist.copy(other._wrapperFields.twist);
    this._wrapperFields.covariance.copy(other._wrapperFields.covariance);
  }
  static get classType() {
    return TwistWithCovarianceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"twist","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"covariance","type":{"isArray":true,"arraySize":36,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"geometry_msgs","type":"TwistWithCovariance","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TwistWithCovariance"};
  }
  hasMember(name) {
    let memberNames = ["twist","covariance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TwistWithCovarianceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TwistWithCovarianceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TwistWithCovarianceWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TwistWithCovarianceArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TwistWithCovarianceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TwistWithCovarianceRefArray(size);
    this._refObject = new TwistWithCovarianceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TwistWithCovarianceWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TwistWithCovarianceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TwistWithCovarianceArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TwistWithCovarianceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TwistWithCovarianceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TwistWithCovarianceArrayWrapper;
  }
}
module.exports = TwistWithCovarianceWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "twist",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Twist",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "covariance",
      "type": {
        "isArray": true,
        "arraySize": 36,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geometry_msgs",
    "type": "TwistWithCovariance",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TwistWithCovariance"
}
*/
