// rclnodejs/generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const UUIDRefStruct = StructType({
  uuid: ArrayType(primitiveTypes.uint8, 16),
});
const UUIDRefArray = ArrayType(UUIDRefStruct);
const UUIDRefStructArray = StructType({
  data: UUIDRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UUIDWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UUIDRefStruct(other._refObject.toObject());
      this._wrapperFields.uuid = UInt8Wrapper.createArray();
      this._wrapperFields.uuid.copy(other._wrapperFields.uuid);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UUIDRefStruct();
    this._wrapperFields.uuid = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new UUIDWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UUIDArrayWrapper;
  }
  static get ArrayType() {
    return UUIDArrayWrapper;
  }
  static get refObjectArrayType() {
    return UUIDRefStructArray
  }
  static get refObjectType() {
    return UUIDRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    for (let i = 0; i < this._wrapperFields.uuid.data.length; i++) {
      this._refObject.uuid[i] = this._wrapperFields.uuid.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.uuid = refObject.uuid;
    this._wrapperFields.uuid.fill(refObject.uuid.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    UUIDWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'unique_identifier_msgs', subFolder: 'msg', interfaceName: 'UUID'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get uuid() {
    return this._wrapperFields['uuid'].data;
  }
  set uuid(value) {
    if (value.length !== 16) {
      throw new RangeError('The length of the array must be 16.');
    }
    this._wrapperFields['uuid'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new UUIDRefStruct(refObject.toObject());
    this._wrapperFields.uuid.fill(refObject.uuid.toArray());
  }
  copy(other) {
    this._refObject = new UUIDRefStruct(other._refObject.toObject());
    this._wrapperFields.uuid.copy(other._wrapperFields.uuid);
  }
  static get classType() {
    return UUIDWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"uuid","type":{"isArray":true,"arraySize":16,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UUID"};
  }
  hasMember(name) {
    let memberNames = ["uuid"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UUIDArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UUIDWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UUIDWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UUIDArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UUIDArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UUIDRefArray(size);
    this._refObject = new UUIDRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UUIDWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UUIDArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UUIDArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UUIDWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UUIDWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UUIDArrayWrapper;
  }
}
module.exports = UUIDWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "uuid",
      "type": {
        "isArray": true,
        "arraySize": 16,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "unique_identifier_msgs",
    "type": "UUID",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UUID"
}
*/
