// rclnodejs/generated/lifecycle_msgs/lifecycle_msgs__msg__State.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StateRefStruct = StructType({
  id: primitiveTypes.uint8,
  label: primitiveTypes.string,
});
const StateRefArray = ArrayType(StateRefStruct);
const StateRefStructArray = StructType({
  data: StateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StateRefStruct(other._refObject.toObject());
      this._idIntialized = true;
      this._labelIntialized = true;
      this._wrapperFields.label =  new StringWrapper(other._wrapperFields.label);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StateRefStruct();
    this._idIntialized = false;
    this._labelIntialized = false;
    this._wrapperFields.label = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new StateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StateArrayWrapper;
  }
  static get ArrayType() {
    return StateArrayWrapper;
  }
  static get refObjectArrayType() {
    return StateRefStructArray
  }
  static get refObjectType() {
    return StateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in State');
      }
      if (!this._labelIntialized) {
        throw new TypeError('Invalid argument: label in State');
      }
    }
    if (own) {
      this._wrapperFields.label.freeze(own, checkConsistency);
    }
    this._refObject.label = this._wrapperFields.label.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._labelIntialized = true;
    this._wrapperFields.label.data = refObject.label.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.label);
  }
  static destoryRawROS(msg) {
    StateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'lifecycle_msgs', subFolder: 'msg', interfaceName: 'State'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get label() {
    if (!this._labelIntialized) {
      return undefined;
    }
    return this._wrapperFields.label.data;
  }
  set label(value) {
    this._labelIntialized = true;
    this._wrapperFields.label.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StateRefStruct(refObject.toObject());
    this._idIntialized = true;
    this._labelIntialized = true;
    this._wrapperFields.label.copyRefObject(this._refObject.label);
  }
  copy(other) {
    this._refObject = new StateRefStruct(other._refObject.toObject());
    this._idIntialized = true;
    this._labelIntialized = true;
    this._wrapperFields.label.copy(other._wrapperFields.label);
  }
  static get classType() {
    return StateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"PRIMARY_STATE_UNKNOWN","value":0},{"type":"uint8","name":"PRIMARY_STATE_UNCONFIGURED","value":1},{"type":"uint8","name":"PRIMARY_STATE_INACTIVE","value":2},{"type":"uint8","name":"PRIMARY_STATE_ACTIVE","value":3},{"type":"uint8","name":"PRIMARY_STATE_FINALIZED","value":4},{"type":"uint8","name":"TRANSITION_STATE_CONFIGURING","value":10},{"type":"uint8","name":"TRANSITION_STATE_CLEANINGUP","value":11},{"type":"uint8","name":"TRANSITION_STATE_SHUTTINGDOWN","value":12},{"type":"uint8","name":"TRANSITION_STATE_ACTIVATING","value":13},{"type":"uint8","name":"TRANSITION_STATE_DEACTIVATING","value":14},{"type":"uint8","name":"TRANSITION_STATE_ERRORPROCESSING","value":15}],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"label","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"lifecycle_msgs","type":"State","stringUpperBound":null,"isPrimitiveType":false},"msgName":"State"};
  }
  hasMember(name) {
    let memberNames = ["id","label"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StateRefArray(size);
    this._refObject = new StateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StateArrayWrapper;
  }
}
// Define constants (11 in total)
Object.defineProperty(StateWrapper, "PRIMARY_STATE_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "PRIMARY_STATE_UNCONFIGURED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "PRIMARY_STATE_INACTIVE", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "PRIMARY_STATE_ACTIVE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "PRIMARY_STATE_FINALIZED", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "TRANSITION_STATE_CONFIGURING", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "TRANSITION_STATE_CLEANINGUP", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "TRANSITION_STATE_SHUTTINGDOWN", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "TRANSITION_STATE_ACTIVATING", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "TRANSITION_STATE_DEACTIVATING", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StateWrapper, "TRANSITION_STATE_ERRORPROCESSING", {value: 15, writable: false, enumerable: true, configurable: true});
module.exports = StateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "PRIMARY_STATE_UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "PRIMARY_STATE_UNCONFIGURED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "PRIMARY_STATE_INACTIVE",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "PRIMARY_STATE_ACTIVE",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "PRIMARY_STATE_FINALIZED",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "TRANSITION_STATE_CONFIGURING",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "TRANSITION_STATE_CLEANINGUP",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "TRANSITION_STATE_SHUTTINGDOWN",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "TRANSITION_STATE_ACTIVATING",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "TRANSITION_STATE_DEACTIVATING",
      "value": 14
    },
    {
      "type": "uint8",
      "name": "TRANSITION_STATE_ERRORPROCESSING",
      "value": 15
    }
  ],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "label",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "lifecycle_msgs",
    "type": "State",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "State"
}
*/
