// rclnodejs/generated/lifecycle_msgs/lifecycle_msgs__srv__ChangeState_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TransitionWrapper = require('../../generated/lifecycle_msgs/lifecycle_msgs__msg__Transition.js');
const ChangeState_RequestRefStruct = StructType({
  transition: TransitionWrapper.refObjectType,
});
const ChangeState_RequestRefArray = ArrayType(ChangeState_RequestRefStruct);
const ChangeState_RequestRefStructArray = StructType({
  data: ChangeState_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ChangeState_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ChangeState_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.transition =  new TransitionWrapper(other._wrapperFields.transition);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ChangeState_RequestRefStruct();
    this._wrapperFields.transition = new TransitionWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ChangeState_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ChangeState_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ChangeState_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ChangeState_RequestRefStructArray
  }
  static get refObjectType() {
    return ChangeState_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.transition.freeze(own, checkConsistency);
    this._refObject.transition = this._wrapperFields.transition.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.transition.copyRefObject(refObject.transition);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TransitionWrapper.freeStruct(refObject.transition);
  }
  static destoryRawROS(msg) {
    ChangeState_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'lifecycle_msgs', subFolder: 'srv', interfaceName: 'ChangeState_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get transition() {
    return this._wrapperFields.transition;
  }
  set transition(value) {
    if (value instanceof TransitionWrapper) {
      this._wrapperFields.transition.copy(value);
    } else {
      this._wrapperFields.transition.copy(new TransitionWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ChangeState_RequestRefStruct(refObject.toObject());
    this._wrapperFields.transition.copyRefObject(this._refObject.transition);
  }
  copy(other) {
    this._refObject = new ChangeState_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.transition.copy(other._wrapperFields.transition);
  }
  static get classType() {
    return ChangeState_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"transition","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"lifecycle_msgs","type":"Transition","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"lifecycle_msgs","type":"ChangeState_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ChangeState_Request"};
  }
  hasMember(name) {
    let memberNames = ["transition"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ChangeState_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ChangeState_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ChangeState_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChangeState_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChangeState_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ChangeState_RequestRefArray(size);
    this._refObject = new ChangeState_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ChangeState_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ChangeState_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ChangeState_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ChangeState_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ChangeState_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ChangeState_RequestArrayWrapper;
  }
}
module.exports = ChangeState_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "transition",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "lifecycle_msgs",
        "type": "Transition",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "lifecycle_msgs",
    "type": "ChangeState_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ChangeState_Request"
}
*/
